<!DOCTYPE html>

<html lang="en" data-content_root="../../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Using the Data Imputator on a Real Public Dataset, Herold et al. (2020) &#8212; MIMIC 0.1.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="../../../_static/alabaster.css?v=12dfc556" />
    <link rel="stylesheet" type="text/css" href="../../../_static/nbsphinx-code-cells.css" />
    <script src="../../../_static/documentation_options.js?v=01f34227"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Simulate some time course data from the gMLV" href="../../gMLV/examples-sim-gMLV.html" />
    <link rel="prev" title="&lt;no title&gt;" href="examples_impute_data.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="Using-the-Data-Imputator-on-a-Real-Public-Dataset,-Herold-et-al.-(2020)">
<h1>Using the Data Imputator on a Real Public Dataset, Herold et al. (2020)<a class="headerlink" href="#Using-the-Data-Imputator-on-a-Real-Public-Dataset,-Herold-et-al.-(2020)" title="Link to this heading">¶</a></h1>
<p>In this example, we will be using the GP imputator on a dataset available at <a class="reference external" href="https://doi.org/10.1038/s41467-020-19006-2">Herold et al., 2020</a> to imputate missing data. Then, we are going to use the VAR inference method to infer the causal relationships between the variables in the time series.</p>
<p>First, we will import the neccessary libraries and load the dataset.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>


<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>


<span class="kn">from</span> <span class="nn">mimic.data_imputation.impute_GP</span> <span class="kn">import</span> <span class="n">GPImputer</span>
<span class="kn">from</span> <span class="nn">mimic.model_infer.infer_VAR_bayes</span> <span class="kn">import</span> <span class="o">*</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING:tensorflow:From c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\tf_keras\src\losses.py:2976: The name tf.losses.sparse_softmax_cross_entropy is deprecated. Please use tf.compat.v1.losses.sparse_softmax_cross_entropy instead.

WARNING:tensorflow:From c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\tensorflow_probability\python\internal\backend\numpy\_utils.py:48: The name tf.logging.TaskLevelStatusMessage is deprecated. Please use tf.compat.v1.logging.TaskLevelStatusMessage instead.

WARNING:tensorflow:From c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\tensorflow_probability\python\internal\backend\numpy\_utils.py:48: The name tf.control_flow_v2_enabled is deprecated. Please use tf.compat.v1.control_flow_v2_enabled instead.

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
WARNING (pytensor.tensor.blas): Using NumPy C-API based implementation for BLAS functions.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Source Data/fig3ab.tsv&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>genus</th>
      <th>date</th>
      <th>val</th>
      <th>type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Acidimicrobium</td>
      <td>2011-03-21</td>
      <td>8.322484</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Acidimicrobium</td>
      <td>2011-03-29</td>
      <td>7.664534</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Acidimicrobium</td>
      <td>2011-04-05</td>
      <td>7.871188</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Acidimicrobium</td>
      <td>2011-04-14</td>
      <td>6.695900</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Acidimicrobium</td>
      <td>2011-04-21</td>
      <td>6.695177</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Acidimicrobium</td>
      <td>2011-04-29</td>
      <td>6.304652</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>6</th>
      <td>Acidimicrobium</td>
      <td>2011-05-06</td>
      <td>7.462918</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Acidimicrobium</td>
      <td>2011-05-13</td>
      <td>7.742366</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Acidimicrobium</td>
      <td>2011-05-20</td>
      <td>6.627358</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>9</th>
      <td>Acidimicrobium</td>
      <td>2011-05-27</td>
      <td>7.425880</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>10</th>
      <td>Acidimicrobium</td>
      <td>2011-06-03</td>
      <td>7.890684</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>11</th>
      <td>Acidimicrobium</td>
      <td>2011-06-09</td>
      <td>6.306718</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>12</th>
      <td>Acidimicrobium</td>
      <td>2011-06-17</td>
      <td>6.607237</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>13</th>
      <td>Acidimicrobium</td>
      <td>2011-06-24</td>
      <td>8.888402</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>14</th>
      <td>Acidimicrobium</td>
      <td>2011-07-01</td>
      <td>9.086751</td>
      <td>MG</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
<section id="Separate-the-data-into-MG-and-MT-datasets">
<h1>Separate the data into MG and MT datasets<a class="headerlink" href="#Separate-the-data-into-MG-and-MT-datasets" title="Link to this heading">¶</a></h1>
<p>The data is separated into two datasets, MG and MT. The MG dataset contains the meta-genomics data, while the MT dataset contains the meta-transcriptomics. We will be using the MG dataset in this example.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># separate the data into MG and MT datasets</span>

<span class="n">dataMG</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;MG&#39;</span><span class="p">]</span>

<span class="n">dataMG</span><span class="o">.</span><span class="n">info</span><span class="p">()</span>

<span class="n">dataMT</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;MT&#39;</span><span class="p">]</span>

<span class="n">dataMT</span><span class="o">.</span><span class="n">info</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>We will proceed with some data wrangling to prepare the data for the imputation process.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Data cleaning</span>
<span class="c1"># Rename Nan in the genus column to &#39;Other&#39;</span>

<span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="s1">&#39;Other&#39;</span><span class="p">)</span>

<span class="c1"># Drop the &#39;type&#39; column</span>
<span class="n">dataMG</span> <span class="o">=</span> <span class="n">dataMG</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">])</span>

<span class="c1"># Change relative Date columns to days</span>
<span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_datetime</span><span class="p">(</span><span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">days</span>

<span class="c1"># Change the &#39;genus&#39; column to a categorical column</span>
<span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>


<span class="n">dataMG</span><span class="o">.</span><span class="n">info</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the data, relative abundance (value) vs. days (date) for each genus</span>


<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataMG</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Microbial Growth Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Pivot the data so each &#39;genus&#39; is a column</span>
<span class="n">dataMG_pivot</span> <span class="o">=</span> <span class="n">dataMG</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Create a stacked area chart</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">stackplot</span><span class="p">(</span><span class="n">dataMG_pivot</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">dataMG_pivot</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">dataMG_pivot</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Microbial Growth Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Do the same data cleaning for the MT data set</span>

<span class="c1"># Rename Nan in the genus column to &#39;Other&#39;</span>

<span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="s1">&#39;Other&#39;</span><span class="p">)</span>

<span class="c1"># Drop the &#39;type&#39; column</span>
<span class="n">dataMT</span> <span class="o">=</span> <span class="n">dataMT</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">])</span>

<span class="c1"># Change relative Date columns to days</span>
<span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_datetime</span><span class="p">(</span><span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">days</span>

<span class="c1"># Change the &#39;genus&#39; column to a categorical column</span>
<span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the data, relative abundance (value) vs. days (date) for each genus</span>


<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataMT</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Microbial Meta Transcriptomics Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Pivot the data so each &#39;genus&#39; is a column</span>
<span class="n">dataMT_pivot</span> <span class="o">=</span> <span class="n">dataMT</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Pivot the MG data so each &#39;genus&#39; is a column</span>
<span class="n">dataMG_pivot</span> <span class="o">=</span> <span class="n">dataMG</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metabolites</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Source Data/fig4.tsv&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>This dataset contains several columns, each representing different attributes related to microbial activity and metabolite concentrations. Here is a detailed explanation of each column:</p>
<ul class="simple">
<li><p><strong>date</strong>: The date when the sample was collected.</p></li>
<li><p><strong>L1</strong>: The functional category or pathway associated with the data point.</p></li>
<li><p><strong>mp_tot</strong>: Total metaproteomic count for that date and pathway.</p></li>
<li><p><strong>mt_tot</strong>: Total metatranscriptomic count for that date and pathway.</p></li>
<li><p><strong>L1sum_MP</strong>: Sum of metaproteomic counts for that specific functional category (L1).</p></li>
<li><p><strong>L1sum_MT</strong>: Sum of metatranscriptomic counts for that specific functional category (L1).</p></li>
<li><p><strong>L1rel_MP</strong>: Relative abundance of metaproteomic counts within the functional category.</p></li>
<li><p><strong>L1rel_MT</strong>: Relative abundance of metatranscriptomic counts within the functional category.</p></li>
<li><p><strong>MP_std_L1</strong>: Standard deviation of metaproteomic counts for that functional category.</p></li>
<li><p><strong>MT_std_L1</strong>: Standard deviation of metatranscriptomic counts for that functional category.</p></li>
<li><p><strong>cor</strong>: Correlation value between metaproteomic and metatranscriptomic data for that pathway.</p></li>
<li><p><strong>newL1</strong>: A reformatted version of the functional category name, often including a correlation coefficient (r value) for easier readability in the figure.</p></li>
</ul>
<p>We will be using the <code class="docutils literal notranslate"><span class="pre">L1rel_MP</span></code> and <code class="docutils literal notranslate"><span class="pre">L1rel_MT</span></code> columns in this example to use it with the VAR inference method to infer the causal relationships between the variables in the time series with the meta transcriptomics (MT) and meta genomics (MG) data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Letr&#39;s create 2 dataframes, one for the L1rel_MP and the other for the L1rel_MT data, using the date, L1 and L2tel values</span>

<span class="n">dataL1_MP</span> <span class="o">=</span> <span class="n">metabolites</span><span class="p">[[</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="s1">&#39;L1rel_MP&#39;</span><span class="p">]]</span>

<span class="n">dataL1_MT</span> <span class="o">=</span> <span class="n">metabolites</span><span class="p">[[</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="s1">&#39;L1rel_MT&#39;</span><span class="p">]]</span>
<span class="n">dataL1_MP</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now let&#39;s visualize the data</span>

<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;L1rel_MP&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataL1_MP</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Metaproteomic Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Same for the metatranscriptomic data</span>

<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;L1rel_MT&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataL1_MT</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Metatranscriptomic Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># And do the same as we did for the MG and MT data sets</span>

<span class="c1"># Rename Nan in the L1 column to &#39;Other&#39;</span>

<span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="s1">&#39;Other&#39;</span><span class="p">)</span>

<span class="c1"># Change relative Date columns to days</span>
<span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_datetime</span><span class="p">(</span><span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">days</span>

<span class="c1"># Change the &#39;L1&#39; column to a categorical column</span>
<span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># do the same for the L1rel_MT data set</span>

<span class="c1"># Rename Nan in the L1 column to &#39;Other&#39;</span>

<span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="s1">&#39;Other&#39;</span><span class="p">)</span>

<span class="c1"># Change relative Date columns to days</span>
<span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_datetime</span><span class="p">(</span><span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">days</span>

<span class="c1"># Change the &#39;L1&#39; column to a categorical column</span>
<span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>

<span class="c1"># Plot the data, relative abundance (value) vs. days (date) for each L1</span>


<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;L1rel_MP&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataL1_MP</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Metaproteomic Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># And now pivot the data so each &#39;L1&#39; is a column for each data set</span>

<span class="n">dataL1_MP_pivot</span> <span class="o">=</span> <span class="n">dataL1_MP</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;L1rel_MP&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="n">dataL1_MT_pivot</span> <span class="o">=</span> <span class="n">dataL1_MT</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;L1rel_MT&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now let&#39;s try to infer the var model for both the L1rel_MP or L1rel_MT data sets with the MG or MT data sets</span>
<span class="c1"># So we will have 4 combinations to try</span>
<span class="c1"># 1. L1rel_MP with MG</span>
<span class="c1"># 2. L1rel_MP with MT</span>
<span class="c1"># 3. L1rel_MT with MG</span>
<span class="c1"># 4. L1rel_MT with MT</span>

<span class="n">infer12</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dataMG_pivot</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">dataL1_MP_pivot</span><span class="p">)</span>

<span class="n">infer12</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer12</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer13</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dataMT_pivot</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">dataL1_MP_pivot</span><span class="p">)</span>

<span class="n">infer13</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer13</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer14</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dataMG_pivot</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">dataL1_MT_pivot</span><span class="p">)</span>

<span class="n">infer14</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer14</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer15</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dataMT_pivot</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">dataL1_MT_pivot</span><span class="p">)</span>

<span class="n">infer15</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer15</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dataL1_MP_pivot</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">dataL1_MP_pivot</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Step 1: Determine the minimum and maximum index</span>
<span class="n">min_index</span> <span class="o">=</span> <span class="n">dataL1_MP_pivot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<span class="n">max_index</span> <span class="o">=</span> <span class="n">dataL1_MP_pivot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>

<span class="c1"># Step 2: Create a new index range from min_index to max_index</span>
<span class="n">new_index</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">min_index</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">max_index</span><span class="p">)</span> <span class="o">+</span>
                  <span class="mi">1</span><span class="p">)</span>  <span class="c1"># Ensure it&#39;s an integer range</span>

<span class="c1"># Step 3: Re-index the DataFrame</span>
<span class="n">dataL1_MP_complete</span> <span class="o">=</span> <span class="n">dataL1_MP_pivot</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">new_index</span><span class="p">)</span>

<span class="n">dataL1_MP_complete</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reset the index</span>
<span class="n">complete_dataMP_reset</span> <span class="o">=</span> <span class="n">dataL1_MP_complete</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

<span class="c1"># Make index as float</span>
<span class="n">complete_dataMP_reset</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">complete_dataMP_reset</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">complete_dataMP_reset</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># do the imputation for each column (metabolite) in the dataset</span>
<span class="c1"># Copy the original dataset</span>
<span class="n">new_dataset_complete</span> <span class="o">=</span> <span class="n">complete_dataMP_reset</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="k">for</span> <span class="n">metabolite</span> <span class="ow">in</span> <span class="n">dataL1_MP_complete</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="c1"># Impute the missing values for the current column</span>
    <span class="n">imputed_data</span> <span class="o">=</span> <span class="n">GPImputer</span><span class="p">()</span><span class="o">.</span><span class="n">impute_missing_values</span><span class="p">(</span>
        <span class="n">complete_dataMP_reset</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">metabolite</span><span class="p">],</span> <span class="n">metabolite</span>
    <span class="p">)</span>

    <span class="c1"># Update the column in the new dataset with the imputed data</span>
    <span class="n">new_dataset_complete</span><span class="p">[</span><span class="n">metabolite</span><span class="p">]</span> <span class="o">=</span> <span class="n">imputed_data</span><span class="p">[</span><span class="n">metabolite</span><span class="p">]</span>

<span class="c1"># Save the dataset (new_dataset_complete) to a csv file</span>
<span class="n">new_dataset_complete</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;MP_metabolite_imputed_data.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Let&#39;s do the same for the MT dataset</span>

<span class="c1"># Step 1: Determine the minimum and maximum index</span>
<span class="n">min_index</span> <span class="o">=</span> <span class="n">dataL1_MT_pivot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<span class="n">max_index</span> <span class="o">=</span> <span class="n">dataL1_MT_pivot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>

<span class="c1"># Step 2: Create a new index range from min_index to max_index</span>
<span class="n">new_index</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">min_index</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">max_index</span><span class="p">)</span> <span class="o">+</span>
                  <span class="mi">1</span><span class="p">)</span>  <span class="c1"># Ensure it&#39;s an integer range</span>

<span class="c1"># Step 3: Re-index the DataFrame</span>
<span class="n">dataL1_MT_complete</span> <span class="o">=</span> <span class="n">dataL1_MT_pivot</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">new_index</span><span class="p">)</span>

<span class="n">dataL1_MT_complete</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reset the index</span>
<span class="n">complete_dataMT_reset</span> <span class="o">=</span> <span class="n">dataL1_MT_complete</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

<span class="c1"># Make index as float</span>
<span class="n">complete_dataMT_reset</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">complete_dataMT_reset</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># do the imputation for each column (metabolite) in the dataset</span>
<span class="c1"># Copy the original dataset</span>
<span class="n">new_dataset_complete</span> <span class="o">=</span> <span class="n">complete_dataMT_reset</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="k">for</span> <span class="n">metabolite</span> <span class="ow">in</span> <span class="n">dataL1_MT_complete</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="c1"># Impute the missing values for the current column</span>
    <span class="n">imputed_data</span> <span class="o">=</span> <span class="n">GPImputer</span><span class="p">()</span><span class="o">.</span><span class="n">impute_missing_values</span><span class="p">(</span>
        <span class="n">complete_dataMT_reset</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">metabolite</span><span class="p">],</span> <span class="n">metabolite</span>
    <span class="p">)</span>

    <span class="c1"># Update the column in the new dataset with the imputed data</span>
    <span class="n">new_dataset_complete</span><span class="p">[</span><span class="n">metabolite</span><span class="p">]</span> <span class="o">=</span> <span class="n">imputed_data</span><span class="p">[</span><span class="n">metabolite</span><span class="p">]</span>

<span class="c1"># Save the dataset (new_dataset_complete) to a csv file</span>
<span class="n">new_dataset_complete</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;MT_metabolite_imputed_data.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>So now, let’s try doing the inference with the VAR method and the imputed data to see if it can infer the causal relationships between the variables in the time series.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reload the imputed data into a new DataFrame</span>

<span class="n">imputed_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;imputed_data.csv&#39;</span><span class="p">)</span>

<span class="c1"># remove the &#39;days&#39; column</span>
<span class="n">imputed_data</span> <span class="o">=</span> <span class="n">imputed_data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;days&#39;</span><span class="p">])</span>
<span class="n">imputed_data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Acidimicrobium</th>
      <th>Acinetobacter</th>
      <th>Albidiferax</th>
      <th>Candidatus Microthrix</th>
      <th>Chitinophaga pinensis</th>
      <th>Dechloromonas</th>
      <th>Haliscomenobacter</th>
      <th>Intrasporangium</th>
      <th>Leptospira</th>
      <th>Other</th>
      <th>Xanthomonas</th>
      <th>mean abundance &lt; 2%</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>8.322484</td>
      <td>0.791870</td>
      <td>6.383491</td>
      <td>26.180838</td>
      <td>7.055585</td>
      <td>4.998287</td>
      <td>15.142620</td>
      <td>11.445072</td>
      <td>4.742982</td>
      <td>2.357318</td>
      <td>0.853658</td>
      <td>11.725795</td>
    </tr>
    <tr>
      <th>1</th>
      <td>8.265177</td>
      <td>0.763703</td>
      <td>6.362575</td>
      <td>25.593890</td>
      <td>7.081694</td>
      <td>5.072711</td>
      <td>15.397948</td>
      <td>11.546582</td>
      <td>4.791269</td>
      <td>2.382206</td>
      <td>0.864697</td>
      <td>11.836536</td>
    </tr>
    <tr>
      <th>2</th>
      <td>8.180701</td>
      <td>0.781881</td>
      <td>6.330330</td>
      <td>25.226861</td>
      <td>7.127157</td>
      <td>5.147886</td>
      <td>15.663537</td>
      <td>11.636934</td>
      <td>4.833619</td>
      <td>2.399779</td>
      <td>0.879132</td>
      <td>11.947293</td>
    </tr>
    <tr>
      <th>3</th>
      <td>8.078737</td>
      <td>0.801890</td>
      <td>6.290243</td>
      <td>24.849371</td>
      <td>7.185821</td>
      <td>5.219273</td>
      <td>15.926969</td>
      <td>11.710476</td>
      <td>4.868031</td>
      <td>2.418784</td>
      <td>0.894873</td>
      <td>12.058067</td>
    </tr>
    <tr>
      <th>4</th>
      <td>7.969807</td>
      <td>0.823758</td>
      <td>6.246734</td>
      <td>24.505012</td>
      <td>7.249498</td>
      <td>5.280679</td>
      <td>16.172482</td>
      <td>11.759176</td>
      <td>4.892380</td>
      <td>2.438263</td>
      <td>0.909303</td>
      <td>12.168858</td>
    </tr>
    <tr>
      <th>5</th>
      <td>7.863614</td>
      <td>0.847488</td>
      <td>6.204391</td>
      <td>24.245813</td>
      <td>7.309111</td>
      <td>5.324508</td>
      <td>16.382314</td>
      <td>11.772278</td>
      <td>4.904550</td>
      <td>2.457029</td>
      <td>0.919595</td>
      <td>12.279667</td>
    </tr>
    <tr>
      <th>6</th>
      <td>7.770084</td>
      <td>0.873055</td>
      <td>6.167656</td>
      <td>24.127242</td>
      <td>7.355842</td>
      <td>5.342460</td>
      <td>16.538718</td>
      <td>11.737040</td>
      <td>4.902619</td>
      <td>2.473734</td>
      <td>0.923138</td>
      <td>12.390494</td>
    </tr>
    <tr>
      <th>7</th>
      <td>7.700314</td>
      <td>0.900405</td>
      <td>6.140528</td>
      <td>24.199845</td>
      <td>7.382293</td>
      <td>5.326927</td>
      <td>16.626579</td>
      <td>11.641157</td>
      <td>4.885061</td>
      <td>2.487002</td>
      <td>0.918012</td>
      <td>12.501338</td>
    </tr>
    <tr>
      <th>8</th>
      <td>7.664534</td>
      <td>0.911291</td>
      <td>6.125822</td>
      <td>23.641355</td>
      <td>7.383681</td>
      <td>5.273044</td>
      <td>16.636265</td>
      <td>11.476739</td>
      <td>4.850959</td>
      <td>2.520630</td>
      <td>0.903478</td>
      <td>12.612201</td>
    </tr>
    <tr>
      <th>9</th>
      <td>7.666801</td>
      <td>0.960073</td>
      <td>6.124129</td>
      <td>25.024940</td>
      <td>7.358852</td>
      <td>5.180566</td>
      <td>16.565955</td>
      <td>11.244474</td>
      <td>4.800196</td>
      <td>2.499159</td>
      <td>0.880339</td>
      <td>12.321359</td>
    </tr>
    <tr>
      <th>10</th>
      <td>7.701597</td>
      <td>0.992116</td>
      <td>6.133189</td>
      <td>25.753165</td>
      <td>7.310707</td>
      <td>5.055128</td>
      <td>16.422573</td>
      <td>10.955857</td>
      <td>4.733611</td>
      <td>2.497398</td>
      <td>0.851104</td>
      <td>12.030537</td>
    </tr>
    <tr>
      <th>11</th>
      <td>7.756540</td>
      <td>1.025393</td>
      <td>6.148177</td>
      <td>26.621835</td>
      <td>7.245802</td>
      <td>4.907436</td>
      <td>16.220878</td>
      <td>10.632238</td>
      <td>4.653122</td>
      <td>2.490776</td>
      <td>0.819790</td>
      <td>11.739731</td>
    </tr>
    <tr>
      <th>12</th>
      <td>7.816900</td>
      <td>1.059680</td>
      <td>6.162504</td>
      <td>27.550701</td>
      <td>7.173270</td>
      <td>4.751605</td>
      <td>15.981051</td>
      <td>10.301403</td>
      <td>4.561808</td>
      <td>2.479674</td>
      <td>0.791448</td>
      <td>11.448942</td>
    </tr>
    <tr>
      <th>13</th>
      <td>7.867041</td>
      <td>1.094721</td>
      <td>6.168573</td>
      <td>28.452806</td>
      <td>7.103389</td>
      <td>4.602845</td>
      <td>15.725567</td>
      <td>9.993155</td>
      <td>4.463977</td>
      <td>2.464200</td>
      <td>0.771427</td>
      <td>11.158170</td>
    </tr>
    <tr>
      <th>14</th>
      <td>7.890203</td>
      <td>1.130229</td>
      <td>6.158709</td>
      <td>29.247651</td>
      <td>7.046066</td>
      <td>4.475146</td>
      <td>15.476159</td>
      <td>9.734647</td>
      <td>4.365213</td>
      <td>2.444171</td>
      <td>0.764505</td>
      <td>10.867413</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">imputed_MP_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;MP_metabolite_imputed_data.csv&#39;</span><span class="p">)</span>
<span class="n">imputed_MP_data</span> <span class="o">=</span> <span class="n">imputed_MP_data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">imputed_MP_data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>01_Fermentation</th>
      <th>02_Homoacetogenesis</th>
      <th>03_Superpathway of thiosulfate metabolism (Desulfovibrio sulfodismutans)</th>
      <th>04_Utililization of sugar, conversion of pentose to EMP pathway intermediates</th>
      <th>05_Fatty acid oxidation</th>
      <th>06_Amino acid utilization biosynthesis metabolism</th>
      <th>07_Nucleic acid metabolism</th>
      <th>08_Hydrocarbon degradation</th>
      <th>09_Carbohydrate Active enzyme - CAZy</th>
      <th>10_TCA cycle</th>
      <th>...</th>
      <th>12_Transporters</th>
      <th>13_Hydrogen metabolism</th>
      <th>14_Methanogenesis</th>
      <th>15_Methylotrophy</th>
      <th>16_Embden Meyerhof-Parnos (EMP)</th>
      <th>17_Gluconeogenesis</th>
      <th>18_Sulfur compounds metabolism</th>
      <th>19_Saccharide and derivated synthesis</th>
      <th>20_Hydrolysis of polymers</th>
      <th>21_Cellular response to stress</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.096899</td>
      <td>0.009890</td>
      <td>0.000579</td>
      <td>0.000535</td>
      <td>0.025483</td>
      <td>0.122784</td>
      <td>0.119531</td>
      <td>0.005524</td>
      <td>0.003921</td>
      <td>0.045398</td>
      <td>...</td>
      <td>0.015593</td>
      <td>0.000846</td>
      <td>0.003520</td>
      <td>0.009267</td>
      <td>0.030473</td>
      <td>0.016261</td>
      <td>0.005346</td>
      <td>0.009267</td>
      <td>0.006193</td>
      <td>0.037334</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.104379</td>
      <td>0.010097</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.026447</td>
      <td>0.125917</td>
      <td>0.111970</td>
      <td>0.004937</td>
      <td>0.003630</td>
      <td>0.046199</td>
      <td>...</td>
      <td>0.015868</td>
      <td>0.000653</td>
      <td>0.003944</td>
      <td>0.011636</td>
      <td>0.030150</td>
      <td>0.016665</td>
      <td>0.004634</td>
      <td>0.008585</td>
      <td>0.006081</td>
      <td>0.036379</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.107954</td>
      <td>0.010104</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.027555</td>
      <td>0.125917</td>
      <td>0.111924</td>
      <td>0.004942</td>
      <td>0.003630</td>
      <td>0.046254</td>
      <td>...</td>
      <td>0.015999</td>
      <td>0.000653</td>
      <td>0.003976</td>
      <td>0.011727</td>
      <td>0.030150</td>
      <td>0.016687</td>
      <td>0.004626</td>
      <td>0.008435</td>
      <td>0.006079</td>
      <td>0.036373</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.110575</td>
      <td>0.010110</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.028627</td>
      <td>0.125917</td>
      <td>0.111895</td>
      <td>0.004947</td>
      <td>0.003630</td>
      <td>0.046306</td>
      <td>...</td>
      <td>0.016125</td>
      <td>0.000653</td>
      <td>0.004017</td>
      <td>0.011795</td>
      <td>0.030150</td>
      <td>0.016703</td>
      <td>0.004616</td>
      <td>0.008353</td>
      <td>0.006077</td>
      <td>0.036378</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.112957</td>
      <td>0.010117</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.029633</td>
      <td>0.125917</td>
      <td>0.111872</td>
      <td>0.004952</td>
      <td>0.003630</td>
      <td>0.046364</td>
      <td>...</td>
      <td>0.016263</td>
      <td>0.000653</td>
      <td>0.004060</td>
      <td>0.011858</td>
      <td>0.030150</td>
      <td>0.016720</td>
      <td>0.004606</td>
      <td>0.008309</td>
      <td>0.006075</td>
      <td>0.036393</td>
    </tr>
    <tr>
      <th>5</th>
      <td>0.115460</td>
      <td>0.010125</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.030625</td>
      <td>0.125917</td>
      <td>0.111851</td>
      <td>0.004957</td>
      <td>0.003630</td>
      <td>0.046434</td>
      <td>...</td>
      <td>0.016427</td>
      <td>0.000653</td>
      <td>0.004105</td>
      <td>0.011927</td>
      <td>0.030150</td>
      <td>0.016741</td>
      <td>0.004598</td>
      <td>0.008297</td>
      <td>0.006072</td>
      <td>0.036419</td>
    </tr>
    <tr>
      <th>6</th>
      <td>0.118505</td>
      <td>0.010138</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.031655</td>
      <td>0.125917</td>
      <td>0.111830</td>
      <td>0.004962</td>
      <td>0.003630</td>
      <td>0.046530</td>
      <td>...</td>
      <td>0.016638</td>
      <td>0.000653</td>
      <td>0.004150</td>
      <td>0.012011</td>
      <td>0.030150</td>
      <td>0.016773</td>
      <td>0.004591</td>
      <td>0.008324</td>
      <td>0.006070</td>
      <td>0.036465</td>
    </tr>
    <tr>
      <th>7</th>
      <td>0.123112</td>
      <td>0.010161</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.032694</td>
      <td>0.125917</td>
      <td>0.111804</td>
      <td>0.004966</td>
      <td>0.003630</td>
      <td>0.046688</td>
      <td>...</td>
      <td>0.016966</td>
      <td>0.000653</td>
      <td>0.004197</td>
      <td>0.012141</td>
      <td>0.030150</td>
      <td>0.016836</td>
      <td>0.004588</td>
      <td>0.008426</td>
      <td>0.006067</td>
      <td>0.036559</td>
    </tr>
    <tr>
      <th>8</th>
      <td>0.133734</td>
      <td>0.011645</td>
      <td>0.000340</td>
      <td>0.000870</td>
      <td>0.033575</td>
      <td>0.129235</td>
      <td>0.108779</td>
      <td>0.004688</td>
      <td>0.003063</td>
      <td>0.052367</td>
      <td>...</td>
      <td>0.018224</td>
      <td>0.001059</td>
      <td>0.004159</td>
      <td>0.016334</td>
      <td>0.035428</td>
      <td>0.018678</td>
      <td>0.003592</td>
      <td>0.009150</td>
      <td>0.006692</td>
      <td>0.042688</td>
    </tr>
    <tr>
      <th>9</th>
      <td>0.123435</td>
      <td>0.010154</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.033030</td>
      <td>0.125917</td>
      <td>0.111790</td>
      <td>0.004976</td>
      <td>0.003630</td>
      <td>0.046701</td>
      <td>...</td>
      <td>0.017093</td>
      <td>0.000653</td>
      <td>0.004302</td>
      <td>0.012159</td>
      <td>0.030150</td>
      <td>0.016790</td>
      <td>0.004592</td>
      <td>0.008234</td>
      <td>0.006062</td>
      <td>0.036497</td>
    </tr>
    <tr>
      <th>10</th>
      <td>0.119160</td>
      <td>0.010124</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.032334</td>
      <td>0.125917</td>
      <td>0.111801</td>
      <td>0.004981</td>
      <td>0.003630</td>
      <td>0.046557</td>
      <td>...</td>
      <td>0.016899</td>
      <td>0.000653</td>
      <td>0.004362</td>
      <td>0.012048</td>
      <td>0.030150</td>
      <td>0.016679</td>
      <td>0.004598</td>
      <td>0.007934</td>
      <td>0.006060</td>
      <td>0.036339</td>
    </tr>
    <tr>
      <th>11</th>
      <td>0.116464</td>
      <td>0.010103</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.031667</td>
      <td>0.125917</td>
      <td>0.111808</td>
      <td>0.004986</td>
      <td>0.003630</td>
      <td>0.046478</td>
      <td>...</td>
      <td>0.016835</td>
      <td>0.000653</td>
      <td>0.004425</td>
      <td>0.011984</td>
      <td>0.030150</td>
      <td>0.016595</td>
      <td>0.004602</td>
      <td>0.007689</td>
      <td>0.006057</td>
      <td>0.036223</td>
    </tr>
    <tr>
      <th>12</th>
      <td>0.114334</td>
      <td>0.010085</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.031067</td>
      <td>0.125917</td>
      <td>0.111815</td>
      <td>0.004991</td>
      <td>0.003630</td>
      <td>0.046428</td>
      <td>...</td>
      <td>0.016847</td>
      <td>0.000653</td>
      <td>0.004486</td>
      <td>0.011937</td>
      <td>0.030150</td>
      <td>0.016514</td>
      <td>0.004603</td>
      <td>0.007452</td>
      <td>0.006054</td>
      <td>0.036115</td>
    </tr>
    <tr>
      <th>13</th>
      <td>0.112345</td>
      <td>0.010067</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.030503</td>
      <td>0.125917</td>
      <td>0.111825</td>
      <td>0.004997</td>
      <td>0.003630</td>
      <td>0.046399</td>
      <td>...</td>
      <td>0.016938</td>
      <td>0.000653</td>
      <td>0.004540</td>
      <td>0.011898</td>
      <td>0.030150</td>
      <td>0.016423</td>
      <td>0.004597</td>
      <td>0.007184</td>
      <td>0.006051</td>
      <td>0.035997</td>
    </tr>
    <tr>
      <th>14</th>
      <td>0.110036</td>
      <td>0.010043</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.029966</td>
      <td>0.125917</td>
      <td>0.111846</td>
      <td>0.005002</td>
      <td>0.003630</td>
      <td>0.046398</td>
      <td>...</td>
      <td>0.017176</td>
      <td>0.000653</td>
      <td>0.004575</td>
      <td>0.011857</td>
      <td>0.030150</td>
      <td>0.016293</td>
      <td>0.004580</td>
      <td>0.006819</td>
      <td>0.006048</td>
      <td>0.035833</td>
    </tr>
  </tbody>
</table>
<p>15 rows × 21 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">imputed_MT_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;MT_metabolite_imputed_data.csv&#39;</span><span class="p">)</span>
<span class="n">imputed_MT_data</span> <span class="o">=</span> <span class="n">imputed_MT_data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>


<span class="n">imputed_MT_data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer12</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MT_data</span><span class="p">)</span>

<span class="n">infer12</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer12</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer13</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MT_data</span><span class="p">)</span>

<span class="n">infer13</span><span class="o">.</span><span class="n">run_inference_large_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer13</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">imputed_MP_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;MP_metabolite_imputed_data.csv&#39;</span><span class="p">)</span>
<span class="n">imputed_MP_data</span> <span class="o">=</span> <span class="n">imputed_MP_data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>

<span class="n">imputed_MP_data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer14</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">)</span>
<span class="n">infer14</span><span class="o">.</span><span class="n">run_inference_large_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer14</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<hr class="docutils" />
<section id="Reducing-number-of-variables-(to-the-3-or-5-most-abundant)-to-reduce-the-computational-cost-and-simplify-the-analysis">
<h2>Reducing number of variables (to the 3 or 5 most abundant) to reduce the computational cost and simplify the analysis<a class="headerlink" href="#Reducing-number-of-variables-(to-the-3-or-5-most-abundant)-to-reduce-the-computational-cost-and-simplify-the-analysis" title="Link to this heading">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># We are going to develop a method that selects the top 4 or 5 abundances that change the most over time</span>


<span class="k">def</span> <span class="nf">select_top_variable_abundances</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">5</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Selects the top N abundances that change the most over time.</span>
<span class="sd">    Metric Used: Variance of abundances over time.</span>
<span class="sd">    Steps:</span>
<span class="sd">        Converts the data to a NumPy array if it is a DataFrame.</span>
<span class="sd">        Calculates the variance of each abundance across all time points.</span>
<span class="sd">        Sorts the variances in ascending order.</span>
<span class="sd">        Selects the indices of the top N abundances with the highest variance.</span>
<span class="sd">        Output: Indices of the top N most variable abundances.</span>
<span class="sd">        Benefit: Variance captures the overall variability of the abundances, providing a sense of how spread out the values are over time.</span>

<span class="sd">    Args:</span>
<span class="sd">    data (pd.DataFrame or np.ndarray): The abundance data.</span>
<span class="sd">    top_n (int): The number of top variable abundances to select.</span>

<span class="sd">    Returns:</span>
<span class="sd">    np.ndarray: The indices of the selected abundances.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">values</span>

    <span class="n">variances</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">top_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">variances</span><span class="p">)[</span><span class="o">-</span><span class="n">top_n</span><span class="p">:]</span>
    <span class="k">return</span> <span class="n">data</span><span class="p">[:,</span> <span class="n">top_indices</span><span class="p">]</span>


<span class="k">def</span> <span class="nf">top_abundance_changes</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">5</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Selects the top N abundances that change the most over time based on total changes.</span>
<span class="sd">    Metric Used: Sum of absolute changes in abundance between consecutive time points.</span>
<span class="sd">    Steps:</span>
<span class="sd">        Calculates the difference between consecutive time points to find changes.</span>
<span class="sd">        Takes the absolute value of these changes and sums them for each abundance.</span>
<span class="sd">        Sorts these summed changes in descending order.</span>
<span class="sd">        Selects the top N abundances with the highest summed changes.</span>
<span class="sd">        Output: A Series with the top N abundances that have the most change.</span>
<span class="sd">        Benefit: This method focuses on the magnitude of changes between consecutive time points, highlighting abundances that fluctuate the most over the entire period.</span>

<span class="sd">    Args:</span>
<span class="sd">    data (pd.DataFrame): The abundance data.</span>
<span class="sd">    top_n (int): The number of top variable abundances to select.</span>

<span class="sd">    Returns:</span>
<span class="sd">    pd.Series: The abundances with the most changes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Calculate the change in abundance for each genus</span>
    <span class="n">data_change</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">diff</span><span class="p">()</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">top_indices</span> <span class="o">=</span> <span class="n">data_change</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top_n</span><span class="p">)</span><span class="o">.</span><span class="n">index</span>
    <span class="k">return</span> <span class="n">data</span><span class="p">[</span><span class="n">top_indices</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now pick the top 5 most variable abundances for the MG (imputed_data) and the mp (imputed_MP_data) datasets</span>


<span class="n">top_variable_abundances_MG</span> <span class="o">=</span> <span class="n">top_abundance_changes</span><span class="p">(</span><span class="n">imputed_data</span><span class="p">)</span>

<span class="n">top_variable_abundances_MP</span> <span class="o">=</span> <span class="n">top_abundance_changes</span><span class="p">(</span><span class="n">imputed_MP_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Do the inference for these reduced datasets</span>

<span class="n">infer16</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">top_variable_abundances_MG</span><span class="p">,</span>
                    <span class="n">dataS</span><span class="o">=</span><span class="n">top_variable_abundances_MP</span><span class="p">)</span>
<span class="n">infer16</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer16</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">imputed_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;imputed_data.csv&#39;</span><span class="p">)</span>
<span class="c1"># create a pandas dataframe of the original imputed data_set with days as the index</span>
<span class="n">imputed_data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;days&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">imputed_MP_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;MP_metabolite_imputed_data.csv&#39;</span><span class="p">)</span>
<span class="n">imputed_MP_data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># now use the VAR model with these as priors</span>
<span class="n">inferWithoutPrior</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">)</span>
<span class="n">inferWithoutPrior</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 21:26<00:00 Sampling 2 chains, 1,307 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 1290 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.126  0.142  -0.016    0.268      0.100    0.085       3.0
Ah[0, 1]    0.027  0.021   0.006    0.048      0.015    0.013       2.0
Ah[0, 2]    0.127  0.104   0.023    0.231      0.073    0.062       3.0
Ah[0, 3]    0.081  0.002   0.079    0.083      0.001    0.001       2.0
Ah[0, 4]    0.176  0.089   0.086    0.265      0.063    0.053       2.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]  -0.539  0.217  -0.760   -0.319      0.153    0.130       3.0
Bh[20, 8]  -0.378  0.037  -0.416   -0.341      0.026    0.022       3.0
Bh[20, 9]   0.367  0.392  -0.026    0.761      0.277    0.235       3.0
Bh[20, 10] -0.135  0.528  -0.663    0.393      0.373    0.316       3.0
Bh[20, 11] -0.229  0.009  -0.238   -0.219      0.006    0.005       3.0

            ess_tail  r_hat
Ah[0, 0]        33.0   1.89
Ah[0, 1]        16.0   2.82
Ah[0, 2]        13.0   2.23
Ah[0, 3]        14.0   3.03
Ah[0, 4]        12.0   2.65
...              ...    ...
Bh[20, 7]       11.0   2.24
Bh[20, 8]       14.0   2.25
Bh[20, 9]       14.0   2.19
Bh[20, 10]      12.0   2.27
Bh[20, 11]      27.0   1.99

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_57_5.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_57_5.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithoutPrior</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.126  0.142  -0.016    0.268      0.100    0.085       3.0
Ah[0, 1]    0.027  0.021   0.006    0.048      0.015    0.013       2.0
Ah[0, 2]    0.127  0.104   0.023    0.231      0.073    0.062       3.0
Ah[0, 3]    0.081  0.002   0.079    0.083      0.001    0.001       2.0
Ah[0, 4]    0.176  0.089   0.086    0.265      0.063    0.053       2.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]  -0.539  0.217  -0.760   -0.319      0.153    0.130       3.0
Bh[20, 8]  -0.378  0.037  -0.416   -0.341      0.026    0.022       3.0
Bh[20, 9]   0.367  0.392  -0.026    0.761      0.277    0.235       3.0
Bh[20, 10] -0.135  0.528  -0.663    0.393      0.373    0.316       3.0
Bh[20, 11] -0.229  0.009  -0.238   -0.219      0.006    0.005       3.0

            ess_tail  r_hat
Ah[0, 0]        33.0   1.89
Ah[0, 1]        16.0   2.82
Ah[0, 2]        13.0   2.23
Ah[0, 3]        14.0   3.03
Ah[0, 4]        12.0   2.65
...              ...    ...
Bh[20, 7]       11.0   2.24
Bh[20, 8]       14.0   2.25
Bh[20, 9]       14.0   2.19
Bh[20, 10]      12.0   2.27
Bh[20, 11]      27.0   1.99

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_58_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_58_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_58_2.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_58_2.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statsmodels.tsa.api</span> <span class="kn">import</span> <span class="n">VAR</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># make VAR model with the imputed data set and statsmodels library</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">VAR</span><span class="p">(</span><span class="n">imputed_data</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">maxlags</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ml&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning: An unsupported index was provided and will be ignored when e.g. forecasting.
  self._init_dates(dates, freq)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">coefficients</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">params</span>
<span class="n">covariance_matrix</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">sigma_u</span>
<span class="n">intercepts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">results</span><span class="o">.</span><span class="n">params</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

<span class="c1"># remove the first row of the coefficients matrix</span>
<span class="n">coefficients</span> <span class="o">=</span> <span class="n">coefficients</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
<span class="c1"># transpose the coefficients matrix</span>
<span class="n">coefficients</span> <span class="o">=</span> <span class="n">coefficients</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">coefficients</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">,</span>
                           <span class="n">coefficients</span><span class="o">=</span><span class="n">coefficients</span><span class="p">,</span> <span class="n">intercepts</span><span class="o">=</span><span class="n">intercepts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 26:56<00:00 Sampling 2 chains, 0 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 1621 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.648  0.125   0.523    0.773      0.088    0.075       3.0
Ah[0, 1]    0.013  0.023  -0.011    0.036      0.016    0.014       2.0
Ah[0, 2]    0.356  0.091   0.265    0.446      0.064    0.054       2.0
Ah[0, 3]    0.072  0.020   0.051    0.092      0.014    0.012       3.0
Ah[0, 4]   -0.063  0.020  -0.083   -0.044      0.014    0.012       2.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.019  0.253  -0.234    0.273      0.179    0.151       3.0
Bh[20, 8]  -0.544  0.037  -0.581   -0.507      0.026    0.022       3.0
Bh[20, 9]   0.172  0.382  -0.210    0.557      0.270    0.228       2.0
Bh[20, 10]  0.272  0.233   0.039    0.505      0.164    0.139       3.0
Bh[20, 11] -0.112  0.333  -0.446    0.221      0.235    0.199       3.0

            ess_tail  r_hat
Ah[0, 0]        17.0   2.03
Ah[0, 1]        12.0   2.82
Ah[0, 2]        11.0   2.71
Ah[0, 3]        12.0   2.16
Ah[0, 4]        22.0   2.41
...              ...    ...
Bh[20, 7]       30.0   2.20
Bh[20, 8]       32.0   1.85
Bh[20, 9]       14.0   2.48
Bh[20, 10]      13.0   2.03
Bh[20, 11]      21.0   2.07

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_64_5.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_64_5.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.648  0.125   0.523    0.773      0.088    0.075       3.0
Ah[0, 1]    0.013  0.023  -0.011    0.036      0.016    0.014       2.0
Ah[0, 2]    0.356  0.091   0.265    0.446      0.064    0.054       2.0
Ah[0, 3]    0.072  0.020   0.051    0.092      0.014    0.012       3.0
Ah[0, 4]   -0.063  0.020  -0.083   -0.044      0.014    0.012       2.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.019  0.253  -0.234    0.273      0.179    0.151       3.0
Bh[20, 8]  -0.544  0.037  -0.581   -0.507      0.026    0.022       3.0
Bh[20, 9]   0.172  0.382  -0.210    0.557      0.270    0.228       2.0
Bh[20, 10]  0.272  0.233   0.039    0.505      0.164    0.139       3.0
Bh[20, 11] -0.112  0.333  -0.446    0.221      0.235    0.199       3.0

            ess_tail  r_hat
Ah[0, 0]        17.0   2.03
Ah[0, 1]        12.0   2.82
Ah[0, 2]        11.0   2.71
Ah[0, 3]        12.0   2.16
Ah[0, 4]        22.0   2.41
...              ...    ...
Bh[20, 7]       30.0   2.20
Bh[20, 8]       32.0   1.85
Bh[20, 9]       14.0   2.48
Bh[20, 10]      13.0   2.03
Bh[20, 11]      21.0   2.07

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_65_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_65_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_65_2.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_65_2.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior2</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">,</span>
                            <span class="n">coefficients</span><span class="o">=</span><span class="n">coefficients</span><span class="p">,</span> <span class="n">intercepts</span><span class="o">=</span><span class="n">intercepts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior2</span><span class="o">.</span><span class="n">run_inference_large_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [c2_A, tau_A, lam_A, Ah, c2_B, tau_B, lam_B, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='24000' class='' max='24000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [24000/24000 1:21:45<00:00 Sampling 4 chains, 0 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 4 chains for 2_000 tune and 4_000 draw iterations (8_000 + 16_000 draws total) took 4913 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Chain 0 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 1 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 2 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 3 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.318  0.361   0.074    0.943      0.180    0.138       4.0
Ah[0, 1]    0.020  0.027  -0.004    0.065      0.014    0.010       5.0
Ah[0, 2]    0.025  0.165  -0.256    0.176      0.083    0.063       4.0
Ah[0, 3]    0.116  0.067  -0.001    0.169      0.034    0.026       4.0
Ah[0, 4]    0.013  0.023  -0.023    0.039      0.012    0.009       5.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]  -0.101  0.490  -0.749    0.404      0.245    0.188       4.0
Bh[20, 8]   0.179  0.168  -0.033    0.349      0.084    0.064       4.0
Bh[20, 9]   0.000  0.510  -0.713    0.557      0.255    0.195       4.0
Bh[20, 10]  0.191  0.328  -0.373    0.443      0.164    0.126       4.0
Bh[20, 11] -0.230  0.217  -0.532    0.084      0.109    0.083       4.0

            ess_tail  r_hat
Ah[0, 0]        11.0   3.96
Ah[0, 1]        23.0   2.96
Ah[0, 2]        16.0   3.19
Ah[0, 3]        11.0   3.75
Ah[0, 4]        23.0   2.97
...              ...    ...
Bh[20, 7]       14.0   3.54
Bh[20, 8]       13.0   3.16
Bh[20, 9]       11.0   3.76
Bh[20, 10]      12.0   3.85
Bh[20, 11]      15.0   3.34

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_67_5.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_67_5.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior2</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.318  0.361   0.074    0.943      0.180    0.138       4.0
Ah[0, 1]    0.020  0.027  -0.004    0.065      0.014    0.010       5.0
Ah[0, 2]    0.025  0.165  -0.256    0.176      0.083    0.063       4.0
Ah[0, 3]    0.116  0.067  -0.001    0.169      0.034    0.026       4.0
Ah[0, 4]    0.013  0.023  -0.023    0.039      0.012    0.009       5.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]  -0.101  0.490  -0.749    0.404      0.245    0.188       4.0
Bh[20, 8]   0.179  0.168  -0.033    0.349      0.084    0.064       4.0
Bh[20, 9]   0.000  0.510  -0.713    0.557      0.255    0.195       4.0
Bh[20, 10]  0.191  0.328  -0.373    0.443      0.164    0.126       4.0
Bh[20, 11] -0.230  0.217  -0.532    0.084      0.109    0.083       4.0

            ess_tail  r_hat
Ah[0, 0]        11.0   3.96
Ah[0, 1]        23.0   2.96
Ah[0, 2]        16.0   3.19
Ah[0, 3]        11.0   3.75
Ah[0, 4]        23.0   2.97
...              ...    ...
Bh[20, 7]       14.0   3.54
Bh[20, 8]       13.0   3.16
Bh[20, 9]       11.0   3.76
Bh[20, 10]      12.0   3.85
Bh[20, 11]      15.0   3.34

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_68_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_68_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_68_2.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_68_2.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">normalVAR</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">normalVAR</span><span class="o">.</span><span class="n">run_inference</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
x0: Shape.0
data[1:, :]: (409, 12)
x0: (12,)
mu: Shape.0
data: (409, 12)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [x0, A, noise_chol]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 19:50<00:00 Sampling 2 chains, 0 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 1194 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
            mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
x0[0]      0.566  0.513  -0.323    1.583      0.016    0.012     972.0
x0[1]      1.655  0.941  -0.161    3.381      0.021    0.016    1924.0
x0[2]     -0.112  0.421  -0.922    0.662      0.016    0.011     715.0
x0[3]      1.560  0.916  -0.251    3.168      0.025    0.018    1375.0
x0[4]      2.400  0.772   0.929    3.823      0.023    0.016    1166.0
...          ...    ...     ...      ...        ...      ...       ...
A[11, 7]   0.014  0.007   0.001    0.027      0.000    0.000    1262.0
A[11, 8]   0.021  0.006   0.010    0.032      0.000    0.000    1107.0
A[11, 9]  -0.111  0.023  -0.156   -0.069      0.000    0.000    2661.0
A[11, 10]  0.018  0.021  -0.022    0.058      0.000    0.000    2109.0
A[11, 11]  1.013  0.010   0.994    1.030      0.000    0.000    1356.0

           ess_tail  r_hat
x0[0]        1894.0    1.0
x0[1]        1895.0    1.0
x0[2]        1382.0    1.0
x0[3]        1614.0    1.0
x0[4]        1791.0    1.0
...             ...    ...
A[11, 7]     2130.0    1.0
A[11, 8]     2063.0    1.0
A[11, 9]     2788.0    1.0
A[11, 10]    2686.0    1.0
A[11, 11]    2022.0    1.0

[156 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (156) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_70_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_70_6.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">posteriorVAR</span> <span class="o">=</span> <span class="n">az</span><span class="o">.</span><span class="n">from_netcdf</span><span class="p">(</span><span class="s1">&#39;model_posterior.nc&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">coefficients</span> <span class="o">=</span> <span class="n">posteriorVAR</span><span class="o">.</span><span class="n">posterior</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">coefficients_median</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">coefficients</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)))</span>
<span class="n">intercepts</span> <span class="o">=</span> <span class="n">posteriorVAR</span><span class="o">.</span><span class="n">posterior</span><span class="p">[</span><span class="s1">&#39;x0&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">intercepts_median</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">intercepts</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior2</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">,</span>
                            <span class="n">coefficients</span><span class="o">=</span><span class="n">coefficients_median</span><span class="p">,</span> <span class="n">intercepts</span><span class="o">=</span><span class="n">intercepts_median</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior2</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 20:19<00:00 Sampling 2 chains, 2,000 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 1224 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.554  0.266   0.287    0.821      0.188    0.159       3.0
Ah[0, 1]    0.003  0.012  -0.013    0.016      0.009    0.007       2.0
Ah[0, 2]   -0.100  0.066  -0.165   -0.023      0.046    0.039       2.0
Ah[0, 3]    0.051  0.023   0.028    0.074      0.016    0.014       3.0
Ah[0, 4]    0.069  0.056   0.014    0.126      0.040    0.033       3.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.184  0.212  -0.029    0.396      0.150    0.127       3.0
Bh[20, 8]  -0.002  0.056  -0.059    0.054      0.040    0.034       2.0
Bh[20, 9]  -0.558  0.360  -0.922   -0.198      0.254    0.215       2.0
Bh[20, 10]  0.258  0.444  -0.187    0.702      0.314    0.266       3.0
Bh[20, 11] -0.267  0.194  -0.461   -0.073      0.137    0.116       3.0

            ess_tail  r_hat
Ah[0, 0]        13.0   2.08
Ah[0, 1]        11.0   2.42
Ah[0, 2]        11.0   2.37
Ah[0, 3]        12.0   2.29
Ah[0, 4]        12.0   2.18
...              ...    ...
Bh[20, 7]       13.0   2.31
Bh[20, 8]       14.0   2.71
Bh[20, 9]       19.0   2.89
Bh[20, 10]      32.0   1.90
Bh[20, 11]      11.0   2.22

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-red-intense-fg ansi-bold">---------------------------------------------------------------------------</span>
<span class="ansi-red-intense-fg ansi-bold">KeyError</span>                                  Traceback (most recent call last)
File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\xarray\backends\file_manager.py:211</span>, in <span class="ansi-cyan-fg">CachingFileManager._acquire_with_cache_info</span><span class="ansi-blue-intense-fg ansi-bold">(self, needs_lock)</span>
<span class="ansi-green-fg">    210</span> <span class="ansi-bold" style="color: rgb(0,135,0)">try</span>:
<span class="ansi-green-intense-fg ansi-bold">--&gt; 211</span>     file <span style="color: rgb(98,98,98)">=</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_cache[<span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_key]
<span class="ansi-green-fg">    212</span> <span class="ansi-bold" style="color: rgb(0,135,0)">except</span> <span class="ansi-bold" style="color: rgb(215,95,95)">KeyError</span>:

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\xarray\backends\lru_cache.py:56</span>, in <span class="ansi-cyan-fg">LRUCache.__getitem__</span><span class="ansi-blue-intense-fg ansi-bold">(self, key)</span>
<span class="ansi-green-fg">     55</span> <span class="ansi-bold" style="color: rgb(0,135,0)">with</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_lock:
<span class="ansi-green-intense-fg ansi-bold">---&gt; 56</span>     value <span style="color: rgb(98,98,98)">=</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_cache[key]
<span class="ansi-green-fg">     57</span>     <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_cache<span style="color: rgb(98,98,98)">.</span>move_to_end(key)

<span class="ansi-red-intense-fg ansi-bold">KeyError</span>: [&lt;class &#39;h5netcdf.core.File&#39;&gt;, (&#39;c:\\Users\\User\\Documents\\GitHub\\MIMIC\\examples\\MultiModel\\Herold\\model_posterior.nc&#39;,), &#39;a&#39;, ((&#39;decode_vlen_strings&#39;, True), (&#39;invalid_netcdf&#39;, None)), &#39;9344176b-2960-4eca-b4ca-d211c56398d1&#39;]

During handling of the above exception, another exception occurred:

<span class="ansi-red-intense-fg ansi-bold">OSError</span>                                   Traceback (most recent call last)
Cell <span class="ansi-green-intense-fg ansi-bold">In[47], line 1</span>
<span class="ansi-green-intense-fg ansi-bold">----&gt; 1</span> inferWithPrior2<span style="color: rgb(98,98,98)">.</span>run_inference_xs()

File <span class="ansi-green-intense-fg ansi-bold">~\Documents\GitHub\MIMIC\mimic\model_infer\infer_VAR_bayes.py:348</span>, in <span class="ansi-cyan-fg">infer_VAR.run_inference_xs</span><span class="ansi-blue-intense-fg ansi-bold">(self, samples, tune, cores)</span>
<span class="ansi-green-fg">    346</span> az<span style="color: rgb(98,98,98)">.</span>plot_posterior(idata, var_names<span style="color: rgb(98,98,98)">=</span>[<span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">Ah</span><span style="color: rgb(175,0,0)">&#34;</span>, <span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">Bh</span><span style="color: rgb(175,0,0)">&#34;</span>])
<span class="ansi-green-fg">    347</span> plt<span style="color: rgb(98,98,98)">.</span>savefig(<span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">posterior_plot.pdf</span><span style="color: rgb(175,0,0)">&#34;</span>)
<span class="ansi-green-intense-fg ansi-bold">--&gt; 348</span> az<span style="color: rgb(98,98,98)">.</span>to_netcdf(idata, <span style="color: rgb(175,0,0)">&#39;</span><span style="color: rgb(175,0,0)">model_posterior.nc</span><span style="color: rgb(175,0,0)">&#39;</span>)
<span class="ansi-green-fg">    349</span> np<span style="color: rgb(98,98,98)">.</span>savez(<span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">data.npz</span><span style="color: rgb(175,0,0)">&#34;</span>, dataX<span style="color: rgb(98,98,98)">=</span>dataX, dataS<span style="color: rgb(98,98,98)">=</span>dataS)

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\data\io_netcdf.py:67</span>, in <span class="ansi-cyan-fg">to_netcdf</span><span class="ansi-blue-intense-fg ansi-bold">(data, filename, group, engine, coords, dims)</span>
<span class="ansi-green-fg">     42</span> <span style="color: rgb(175,0,0)">&#34;&#34;&#34;Save dataset as a netcdf file.</span>
<span class="ansi-green-fg">     43</span>
<span class="ansi-green-fg">     44</span> <span style="color: rgb(175,0,0)">WARNING: Only idempotent in case `data` is InferenceData</span>
<span class="ansi-green-intense-fg ansi-bold">   (...)</span>
<span class="ansi-green-fg">     64</span> <span style="color: rgb(175,0,0)">    filename saved to</span>
<span class="ansi-green-fg">     65</span> <span style="color: rgb(175,0,0)">&#34;&#34;&#34;</span>
<span class="ansi-green-fg">     66</span> inference_data <span style="color: rgb(98,98,98)">=</span> convert_to_inference_data(data, group<span style="color: rgb(98,98,98)">=</span>group, coords<span style="color: rgb(98,98,98)">=</span>coords, dims<span style="color: rgb(98,98,98)">=</span>dims)
<span class="ansi-green-intense-fg ansi-bold">---&gt; 67</span> file_name <span style="color: rgb(98,98,98)">=</span> inference_data<span style="color: rgb(98,98,98)">.</span>to_netcdf(filename, engine<span style="color: rgb(98,98,98)">=</span>engine)
<span class="ansi-green-fg">     68</span> <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> file_name

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\data\inference_data.py:497</span>, in <span class="ansi-cyan-fg">InferenceData.to_netcdf</span><span class="ansi-blue-intense-fg ansi-bold">(self, filename, compress, groups, engine, base_group, overwrite_existing)</span>
<span class="ansi-green-fg">    491</span> <span class="ansi-bold" style="color: rgb(0,135,0)">if</span> compress:
<span class="ansi-green-fg">    492</span>     kwargs[<span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">encoding</span><span style="color: rgb(175,0,0)">&#34;</span>] <span style="color: rgb(98,98,98)">=</span> {
<span class="ansi-green-fg">    493</span>         var_name: {<span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">zlib</span><span style="color: rgb(175,0,0)">&#34;</span>: <span class="ansi-bold" style="color: rgb(0,135,0)">True</span>}
<span class="ansi-green-fg">    494</span>         <span class="ansi-bold" style="color: rgb(0,135,0)">for</span> var_name, values <span class="ansi-bold" style="color: rgb(175,0,255)">in</span> data<span style="color: rgb(98,98,98)">.</span>variables<span style="color: rgb(98,98,98)">.</span>items()
<span class="ansi-green-fg">    495</span>         <span class="ansi-bold" style="color: rgb(0,135,0)">if</span> _compressible_dtype(values<span style="color: rgb(98,98,98)">.</span>dtype)
<span class="ansi-green-fg">    496</span>     }
<span class="ansi-green-intense-fg ansi-bold">--&gt; 497</span> data<span style="color: rgb(98,98,98)">.</span>to_netcdf(filename, mode<span style="color: rgb(98,98,98)">=</span>mode, group<span style="color: rgb(98,98,98)">=</span><span style="color: rgb(175,0,0)">f</span><span style="color: rgb(175,0,0)">&#34;</span><span class="ansi-bold" style="color: rgb(175,95,135)">{</span>base_group<span class="ansi-bold" style="color: rgb(175,95,135)">}</span><span style="color: rgb(175,0,0)">/</span><span class="ansi-bold" style="color: rgb(175,95,135)">{</span>group<span class="ansi-bold" style="color: rgb(175,95,135)">}</span><span style="color: rgb(175,0,0)">&#34;</span>, <span style="color: rgb(98,98,98)">*</span><span style="color: rgb(98,98,98)">*</span>kwargs)
<span class="ansi-green-fg">    498</span> data<span style="color: rgb(98,98,98)">.</span>close()
<span class="ansi-green-fg">    499</span> mode <span style="color: rgb(98,98,98)">=</span> <span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">a</span><span style="color: rgb(175,0,0)">&#34;</span>

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\xarray\core\dataset.py:1957</span>, in <span class="ansi-cyan-fg">Dataset.to_netcdf</span><span class="ansi-blue-intense-fg ansi-bold">(self, path, mode, format, group, engine, encoding, unlimited_dims, compute, invalid_netcdf)</span>
<span class="ansi-green-fg">   1954</span>     encoding <span style="color: rgb(98,98,98)">=</span> {}
<span class="ansi-green-fg">   1955</span> <span class="ansi-bold" style="color: rgb(0,135,0)">from</span> <span class="ansi-bold" style="color: rgb(0,0,255)">xarray</span><span class="ansi-bold" style="color: rgb(0,0,255)">.</span><span class="ansi-bold" style="color: rgb(0,0,255)">backends</span><span class="ansi-bold" style="color: rgb(0,0,255)">.</span><span class="ansi-bold" style="color: rgb(0,0,255)">api</span> <span class="ansi-bold" style="color: rgb(0,135,0)">import</span> to_netcdf
<span class="ansi-green-intense-fg ansi-bold">-&gt; 1957</span> <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> to_netcdf(  <span style="color: rgb(95,135,135)"># type: ignore  # mypy cannot resolve the overloads:(</span>
<span class="ansi-green-fg">   1958</span>     <span style="color: rgb(0,135,0)">self</span>,
<span class="ansi-green-fg">   1959</span>     path,
<span class="ansi-green-fg">   1960</span>     mode<span style="color: rgb(98,98,98)">=</span>mode,
<span class="ansi-green-fg">   1961</span>     <span style="color: rgb(0,135,0)">format</span><span style="color: rgb(98,98,98)">=</span><span style="color: rgb(0,135,0)">format</span>,
<span class="ansi-green-fg">   1962</span>     group<span style="color: rgb(98,98,98)">=</span>group,
<span class="ansi-green-fg">   1963</span>     engine<span style="color: rgb(98,98,98)">=</span>engine,
<span class="ansi-green-fg">   1964</span>     encoding<span style="color: rgb(98,98,98)">=</span>encoding,
<span class="ansi-green-fg">   1965</span>     unlimited_dims<span style="color: rgb(98,98,98)">=</span>unlimited_dims,
<span class="ansi-green-fg">   1966</span>     compute<span style="color: rgb(98,98,98)">=</span>compute,
<span class="ansi-green-fg">   1967</span>     multifile<span style="color: rgb(98,98,98)">=</span><span class="ansi-bold" style="color: rgb(0,135,0)">False</span>,
<span class="ansi-green-fg">   1968</span>     invalid_netcdf<span style="color: rgb(98,98,98)">=</span>invalid_netcdf,
<span class="ansi-green-fg">   1969</span> )

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\xarray\backends\api.py:1255</span>, in <span class="ansi-cyan-fg">to_netcdf</span><span class="ansi-blue-intense-fg ansi-bold">(dataset, path_or_file, mode, format, group, engine, encoding, unlimited_dims, compute, multifile, invalid_netcdf)</span>
<span class="ansi-green-fg">   1251</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">else</span>:
<span class="ansi-green-fg">   1252</span>         <span class="ansi-bold" style="color: rgb(0,135,0)">raise</span> <span class="ansi-bold" style="color: rgb(215,95,95)">ValueError</span>(
<span class="ansi-green-fg">   1253</span>             <span style="color: rgb(175,0,0)">f</span><span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">unrecognized option </span><span style="color: rgb(175,0,0)">&#39;</span><span style="color: rgb(175,0,0)">invalid_netcdf</span><span style="color: rgb(175,0,0)">&#39;</span><span style="color: rgb(175,0,0)"> for engine </span><span class="ansi-bold" style="color: rgb(175,95,135)">{</span>engine<span class="ansi-bold" style="color: rgb(175,95,135)">}</span><span style="color: rgb(175,0,0)">&#34;</span>
<span class="ansi-green-fg">   1254</span>         )
<span class="ansi-green-intense-fg ansi-bold">-&gt; 1255</span> store <span style="color: rgb(98,98,98)">=</span> store_open(target, mode, <span style="color: rgb(0,135,0)">format</span>, group, <span style="color: rgb(98,98,98)">*</span><span style="color: rgb(98,98,98)">*</span>kwargs)
<span class="ansi-green-fg">   1257</span> <span class="ansi-bold" style="color: rgb(0,135,0)">if</span> unlimited_dims <span class="ansi-bold" style="color: rgb(175,0,255)">is</span> <span class="ansi-bold" style="color: rgb(0,135,0)">None</span>:
<span class="ansi-green-fg">   1258</span>     unlimited_dims <span style="color: rgb(98,98,98)">=</span> dataset<span style="color: rgb(98,98,98)">.</span>encoding<span style="color: rgb(98,98,98)">.</span>get(<span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">unlimited_dims</span><span style="color: rgb(175,0,0)">&#34;</span>, <span class="ansi-bold" style="color: rgb(0,135,0)">None</span>)

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\xarray\backends\h5netcdf_.py:176</span>, in <span class="ansi-cyan-fg">H5NetCDFStore.open</span><span class="ansi-blue-intense-fg ansi-bold">(cls, filename, mode, format, group, lock, autoclose, invalid_netcdf, phony_dims, decode_vlen_strings)</span>
<span class="ansi-green-fg">    173</span>         lock <span style="color: rgb(98,98,98)">=</span> combine_locks([HDF5_LOCK, get_write_lock(filename)])
<span class="ansi-green-fg">    175</span> manager <span style="color: rgb(98,98,98)">=</span> CachingFileManager(h5netcdf<span style="color: rgb(98,98,98)">.</span>File, filename, mode<span style="color: rgb(98,98,98)">=</span>mode, kwargs<span style="color: rgb(98,98,98)">=</span>kwargs)
<span class="ansi-green-intense-fg ansi-bold">--&gt; 176</span> <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> <span style="color: rgb(0,135,0)">cls</span>(manager, group<span style="color: rgb(98,98,98)">=</span>group, mode<span style="color: rgb(98,98,98)">=</span>mode, lock<span style="color: rgb(98,98,98)">=</span>lock, autoclose<span style="color: rgb(98,98,98)">=</span>autoclose)

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\xarray\backends\h5netcdf_.py:127</span>, in <span class="ansi-cyan-fg">H5NetCDFStore.__init__</span><span class="ansi-blue-intense-fg ansi-bold">(self, manager, group, mode, lock, autoclose)</span>
<span class="ansi-green-fg">    124</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>format <span style="color: rgb(98,98,98)">=</span> <span class="ansi-bold" style="color: rgb(0,135,0)">None</span>
<span class="ansi-green-fg">    125</span> <span style="color: rgb(95,135,135)"># todo: utilizing find_root_and_group seems a bit clunky</span>
<span class="ansi-green-fg">    126</span> <span style="color: rgb(95,135,135)">#  making filename available on h5netcdf.Group seems better</span>
<span class="ansi-green-intense-fg ansi-bold">--&gt; 127</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_filename <span style="color: rgb(98,98,98)">=</span> find_root_and_group(<span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>ds)[<span style="color: rgb(98,98,98)">0</span>]<span style="color: rgb(98,98,98)">.</span>filename
<span class="ansi-green-fg">    128</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>is_remote <span style="color: rgb(98,98,98)">=</span> is_remote_uri(<span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_filename)
<span class="ansi-green-fg">    129</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>lock <span style="color: rgb(98,98,98)">=</span> ensure_lock(lock)

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\xarray\backends\h5netcdf_.py:187</span>, in <span class="ansi-cyan-fg">H5NetCDFStore.ds</span><span class="ansi-blue-intense-fg ansi-bold">(self)</span>
<span class="ansi-green-fg">    185</span> <span style="color: rgb(175,0,255)">@property</span>
<span class="ansi-green-fg">    186</span> <span class="ansi-bold" style="color: rgb(0,135,0)">def</span> <span style="color: rgb(0,0,255)">ds</span>(<span style="color: rgb(0,135,0)">self</span>):
<span class="ansi-green-intense-fg ansi-bold">--&gt; 187</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_acquire()

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\xarray\backends\h5netcdf_.py:179</span>, in <span class="ansi-cyan-fg">H5NetCDFStore._acquire</span><span class="ansi-blue-intense-fg ansi-bold">(self, needs_lock)</span>
<span class="ansi-green-fg">    178</span> <span class="ansi-bold" style="color: rgb(0,135,0)">def</span> <span style="color: rgb(0,0,255)">_acquire</span>(<span style="color: rgb(0,135,0)">self</span>, needs_lock<span style="color: rgb(98,98,98)">=</span><span class="ansi-bold" style="color: rgb(0,135,0)">True</span>):
<span class="ansi-green-intense-fg ansi-bold">--&gt; 179</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">with</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_manager<span style="color: rgb(98,98,98)">.</span>acquire_context(needs_lock) <span class="ansi-bold" style="color: rgb(0,135,0)">as</span> root:
<span class="ansi-green-fg">    180</span>         ds <span style="color: rgb(98,98,98)">=</span> _nc4_require_group(
<span class="ansi-green-fg">    181</span>             root, <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_group, <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_mode, create_group<span style="color: rgb(98,98,98)">=</span>_h5netcdf_create_group
<span class="ansi-green-fg">    182</span>         )
<span class="ansi-green-fg">    183</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> ds

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\contextlib.py:137</span>, in <span class="ansi-cyan-fg">_GeneratorContextManager.__enter__</span><span class="ansi-blue-intense-fg ansi-bold">(self)</span>
<span class="ansi-green-fg">    135</span> <span class="ansi-bold" style="color: rgb(0,135,0)">del</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>args, <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>kwds, <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>func
<span class="ansi-green-fg">    136</span> <span class="ansi-bold" style="color: rgb(0,135,0)">try</span>:
<span class="ansi-green-intense-fg ansi-bold">--&gt; 137</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> <span style="color: rgb(0,135,0)">next</span>(<span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>gen)
<span class="ansi-green-fg">    138</span> <span class="ansi-bold" style="color: rgb(0,135,0)">except</span> <span class="ansi-bold" style="color: rgb(215,95,95)">StopIteration</span>:
<span class="ansi-green-fg">    139</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">raise</span> <span class="ansi-bold" style="color: rgb(215,95,95)">RuntimeError</span>(<span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">generator didn</span><span style="color: rgb(175,0,0)">&#39;</span><span style="color: rgb(175,0,0)">t yield</span><span style="color: rgb(175,0,0)">&#34;</span>) <span class="ansi-bold" style="color: rgb(0,135,0)">from</span> <span class="ansi-bold" style="color: rgb(0,135,0)">None</span>

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\xarray\backends\file_manager.py:199</span>, in <span class="ansi-cyan-fg">CachingFileManager.acquire_context</span><span class="ansi-blue-intense-fg ansi-bold">(self, needs_lock)</span>
<span class="ansi-green-fg">    196</span> <span style="color: rgb(175,0,255)">@contextlib</span><span style="color: rgb(98,98,98)">.</span>contextmanager
<span class="ansi-green-fg">    197</span> <span class="ansi-bold" style="color: rgb(0,135,0)">def</span> <span style="color: rgb(0,0,255)">acquire_context</span>(<span style="color: rgb(0,135,0)">self</span>, needs_lock<span style="color: rgb(98,98,98)">=</span><span class="ansi-bold" style="color: rgb(0,135,0)">True</span>):
<span class="ansi-green-fg">    198</span> <span style="color: rgb(188,188,188)">    </span><span style="color: rgb(175,0,0)">&#34;&#34;&#34;Context manager for acquiring a file.&#34;&#34;&#34;</span>
<span class="ansi-green-intense-fg ansi-bold">--&gt; 199</span>     file, cached <span style="color: rgb(98,98,98)">=</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_acquire_with_cache_info(needs_lock)
<span class="ansi-green-fg">    200</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">try</span>:
<span class="ansi-green-fg">    201</span>         <span class="ansi-bold" style="color: rgb(0,135,0)">yield</span> file

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\xarray\backends\file_manager.py:217</span>, in <span class="ansi-cyan-fg">CachingFileManager._acquire_with_cache_info</span><span class="ansi-blue-intense-fg ansi-bold">(self, needs_lock)</span>
<span class="ansi-green-fg">    215</span>     kwargs <span style="color: rgb(98,98,98)">=</span> kwargs<span style="color: rgb(98,98,98)">.</span>copy()
<span class="ansi-green-fg">    216</span>     kwargs[<span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">mode</span><span style="color: rgb(175,0,0)">&#34;</span>] <span style="color: rgb(98,98,98)">=</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_mode
<span class="ansi-green-intense-fg ansi-bold">--&gt; 217</span> file <span style="color: rgb(98,98,98)">=</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_opener(<span style="color: rgb(98,98,98)">*</span><span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_args, <span style="color: rgb(98,98,98)">*</span><span style="color: rgb(98,98,98)">*</span>kwargs)
<span class="ansi-green-fg">    218</span> <span class="ansi-bold" style="color: rgb(0,135,0)">if</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_mode <span style="color: rgb(98,98,98)">==</span> <span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">w</span><span style="color: rgb(175,0,0)">&#34;</span>:
<span class="ansi-green-fg">    219</span>     <span style="color: rgb(95,135,135)"># ensure file doesn&#39;t get overridden when opened again</span>
<span class="ansi-green-fg">    220</span>     <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_mode <span style="color: rgb(98,98,98)">=</span> <span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">a</span><span style="color: rgb(175,0,0)">&#34;</span>

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\h5netcdf\core.py:1051</span>, in <span class="ansi-cyan-fg">File.__init__</span><span class="ansi-blue-intense-fg ansi-bold">(self, path, mode, invalid_netcdf, phony_dims, **kwargs)</span>
<span class="ansi-green-fg">   1049</span>         <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_preexisting_file <span style="color: rgb(98,98,98)">=</span> os<span style="color: rgb(98,98,98)">.</span>path<span style="color: rgb(98,98,98)">.</span>exists(path) <span class="ansi-bold" style="color: rgb(175,0,255)">and</span> mode <span style="color: rgb(98,98,98)">!=</span> <span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">w</span><span style="color: rgb(175,0,0)">&#34;</span>
<span class="ansi-green-fg">   1050</span>         <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_h5py <span style="color: rgb(98,98,98)">=</span> h5py
<span class="ansi-green-intense-fg ansi-bold">-&gt; 1051</span>         <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_h5file <span style="color: rgb(98,98,98)">=</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_h5py<span style="color: rgb(98,98,98)">.</span>File(
<span class="ansi-green-fg">   1052</span>             path, mode, track_order<span style="color: rgb(98,98,98)">=</span>track_order, <span style="color: rgb(98,98,98)">*</span><span style="color: rgb(98,98,98)">*</span>kwargs
<span class="ansi-green-fg">   1053</span>         )
<span class="ansi-green-fg">   1054</span> <span class="ansi-bold" style="color: rgb(0,135,0)">else</span>:  <span style="color: rgb(95,135,135)"># file-like object</span>
<span class="ansi-green-fg">   1055</span>     <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_preexisting_file <span style="color: rgb(98,98,98)">=</span> mode <span class="ansi-bold" style="color: rgb(175,0,255)">in</span> {<span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">r</span><span style="color: rgb(175,0,0)">&#34;</span>, <span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">r+</span><span style="color: rgb(175,0,0)">&#34;</span>, <span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">a</span><span style="color: rgb(175,0,0)">&#34;</span>}

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\h5py\_hl\files.py:562</span>, in <span class="ansi-cyan-fg">File.__init__</span><span class="ansi-blue-intense-fg ansi-bold">(self, name, mode, driver, libver, userblock_size, swmr, rdcc_nslots, rdcc_nbytes, rdcc_w0, track_order, fs_strategy, fs_persist, fs_threshold, fs_page_size, page_buf_size, min_meta_keep, min_raw_keep, locking, alignment_threshold, alignment_interval, meta_block_size, **kwds)</span>
<span class="ansi-green-fg">    553</span>     fapl <span style="color: rgb(98,98,98)">=</span> make_fapl(driver, libver, rdcc_nslots, rdcc_nbytes, rdcc_w0,
<span class="ansi-green-fg">    554</span>                      locking, page_buf_size, min_meta_keep, min_raw_keep,
<span class="ansi-green-fg">    555</span>                      alignment_threshold<span style="color: rgb(98,98,98)">=</span>alignment_threshold,
<span class="ansi-green-fg">    556</span>                      alignment_interval<span style="color: rgb(98,98,98)">=</span>alignment_interval,
<span class="ansi-green-fg">    557</span>                      meta_block_size<span style="color: rgb(98,98,98)">=</span>meta_block_size,
<span class="ansi-green-fg">    558</span>                      <span style="color: rgb(98,98,98)">*</span><span style="color: rgb(98,98,98)">*</span>kwds)
<span class="ansi-green-fg">    559</span>     fcpl <span style="color: rgb(98,98,98)">=</span> make_fcpl(track_order<span style="color: rgb(98,98,98)">=</span>track_order, fs_strategy<span style="color: rgb(98,98,98)">=</span>fs_strategy,
<span class="ansi-green-fg">    560</span>                      fs_persist<span style="color: rgb(98,98,98)">=</span>fs_persist, fs_threshold<span style="color: rgb(98,98,98)">=</span>fs_threshold,
<span class="ansi-green-fg">    561</span>                      fs_page_size<span style="color: rgb(98,98,98)">=</span>fs_page_size)
<span class="ansi-green-intense-fg ansi-bold">--&gt; 562</span>     fid <span style="color: rgb(98,98,98)">=</span> make_fid(name, mode, userblock_size, fapl, fcpl, swmr<span style="color: rgb(98,98,98)">=</span>swmr)
<span class="ansi-green-fg">    564</span> <span class="ansi-bold" style="color: rgb(0,135,0)">if</span> <span style="color: rgb(0,135,0)">isinstance</span>(libver, <span style="color: rgb(0,135,0)">tuple</span>):
<span class="ansi-green-fg">    565</span>     <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>_libver <span style="color: rgb(98,98,98)">=</span> libver

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\h5py\_hl\files.py:241</span>, in <span class="ansi-cyan-fg">make_fid</span><span class="ansi-blue-intense-fg ansi-bold">(name, mode, userblock_size, fapl, fcpl, swmr)</span>
<span class="ansi-green-fg">    239</span>     fid <span style="color: rgb(98,98,98)">=</span> h5f<span style="color: rgb(98,98,98)">.</span>create(name, h5f<span style="color: rgb(98,98,98)">.</span>ACC_EXCL, fapl<span style="color: rgb(98,98,98)">=</span>fapl, fcpl<span style="color: rgb(98,98,98)">=</span>fcpl)
<span class="ansi-green-fg">    240</span> <span class="ansi-bold" style="color: rgb(0,135,0)">elif</span> mode <span style="color: rgb(98,98,98)">==</span> <span style="color: rgb(175,0,0)">&#39;</span><span style="color: rgb(175,0,0)">w</span><span style="color: rgb(175,0,0)">&#39;</span>:
<span class="ansi-green-intense-fg ansi-bold">--&gt; 241</span>     fid <span style="color: rgb(98,98,98)">=</span> h5f<span style="color: rgb(98,98,98)">.</span>create(name, h5f<span style="color: rgb(98,98,98)">.</span>ACC_TRUNC, fapl<span style="color: rgb(98,98,98)">=</span>fapl, fcpl<span style="color: rgb(98,98,98)">=</span>fcpl)
<span class="ansi-green-fg">    242</span> <span class="ansi-bold" style="color: rgb(0,135,0)">elif</span> mode <span style="color: rgb(98,98,98)">==</span> <span style="color: rgb(175,0,0)">&#39;</span><span style="color: rgb(175,0,0)">a</span><span style="color: rgb(175,0,0)">&#39;</span>:
<span class="ansi-green-fg">    243</span>     <span style="color: rgb(95,135,135)"># Open in append mode (read/write).</span>
<span class="ansi-green-fg">    244</span>     <span style="color: rgb(95,135,135)"># If that fails, create a new file only if it won&#39;t clobber an</span>
<span class="ansi-green-fg">    245</span>     <span style="color: rgb(95,135,135)"># existing one (ACC_EXCL)</span>
<span class="ansi-green-fg">    246</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">try</span>:

File <span class="ansi-green-intense-fg ansi-bold">h5py\_objects.pyx:54</span>, in <span class="ansi-cyan-fg">h5py._objects.with_phil.wrapper</span><span class="ansi-blue-intense-fg ansi-bold">()</span>

File <span class="ansi-green-intense-fg ansi-bold">h5py\_objects.pyx:55</span>, in <span class="ansi-cyan-fg">h5py._objects.with_phil.wrapper</span><span class="ansi-blue-intense-fg ansi-bold">()</span>

File <span class="ansi-green-intense-fg ansi-bold">h5py\h5f.pyx:122</span>, in <span class="ansi-cyan-fg">h5py.h5f.create</span><span class="ansi-blue-intense-fg ansi-bold">()</span>

<span class="ansi-red-intense-fg ansi-bold">OSError</span>: Unable to synchronously create file (unable to truncate a file which is already open)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_74_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_74_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior2</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-red-intense-fg ansi-bold">---------------------------------------------------------------------------</span>
<span class="ansi-red-intense-fg ansi-bold">KeyError</span>                                  Traceback (most recent call last)
Cell <span class="ansi-green-intense-fg ansi-bold">In[48], line 1</span>
<span class="ansi-green-intense-fg ansi-bold">----&gt; 1</span> inferWithPrior2<span style="color: rgb(98,98,98)">.</span>posterior_analysis()

File <span class="ansi-green-intense-fg ansi-bold">~\Documents\GitHub\MIMIC\mimic\model_infer\infer_VAR_bayes.py:438</span>, in <span class="ansi-cyan-fg">infer_VAR.posterior_analysis</span><span class="ansi-blue-intense-fg ansi-bold">(self, A, B)</span>
<span class="ansi-green-fg">    436</span> <span class="ansi-bold" style="color: rgb(0,135,0)">with</span> np<span style="color: rgb(98,98,98)">.</span>load(<span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">data.npz</span><span style="color: rgb(175,0,0)">&#34;</span>, allow_pickle<span style="color: rgb(98,98,98)">=</span><span class="ansi-bold" style="color: rgb(0,135,0)">True</span>) <span class="ansi-bold" style="color: rgb(0,135,0)">as</span> xsdata:
<span class="ansi-green-fg">    437</span>     dataX <span style="color: rgb(98,98,98)">=</span> xsdata[<span style="color: rgb(175,0,0)">&#39;</span><span style="color: rgb(175,0,0)">dataX</span><span style="color: rgb(175,0,0)">&#39;</span>]
<span class="ansi-green-intense-fg ansi-bold">--&gt; 438</span>     dataS <span style="color: rgb(98,98,98)">=</span> xsdata[<span style="color: rgb(175,0,0)">&#39;</span><span style="color: rgb(175,0,0)">dataS</span><span style="color: rgb(175,0,0)">&#39;</span>]
<span class="ansi-green-fg">    440</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>make_plot_stacked(dataX, dataS)
<span class="ansi-green-fg">    442</span> idata <span style="color: rgb(98,98,98)">=</span> az<span style="color: rgb(98,98,98)">.</span>from_netcdf(<span style="color: rgb(175,0,0)">&#39;</span><span style="color: rgb(175,0,0)">model_posterior.nc</span><span style="color: rgb(175,0,0)">&#39;</span>)

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\numpy\lib\npyio.py:263</span>, in <span class="ansi-cyan-fg">NpzFile.__getitem__</span><span class="ansi-blue-intense-fg ansi-bold">(self, key)</span>
<span class="ansi-green-fg">    261</span>         <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> <span style="color: rgb(0,135,0)">self</span><span style="color: rgb(98,98,98)">.</span>zip<span style="color: rgb(98,98,98)">.</span>read(key)
<span class="ansi-green-fg">    262</span> <span class="ansi-bold" style="color: rgb(0,135,0)">else</span>:
<span class="ansi-green-intense-fg ansi-bold">--&gt; 263</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">raise</span> <span class="ansi-bold" style="color: rgb(215,95,95)">KeyError</span>(<span style="color: rgb(175,0,0)">f</span><span style="color: rgb(175,0,0)">&#34;</span><span class="ansi-bold" style="color: rgb(175,95,135)">{</span>key<span class="ansi-bold" style="color: rgb(175,95,135)">}</span><span style="color: rgb(175,0,0)"> is not a file in the archive</span><span style="color: rgb(175,0,0)">&#34;</span>)

<span class="ansi-red-intense-fg ansi-bold">KeyError</span>: &#39;dataS is not a file in the archive&#39;
</pre></div></div>
</div>
<p>So, these inferences show that there is no easy correlation to infer between the species relative abundances and the metabolite concentrations. This could be due to us trying to infer the causal inference between the relative abundance, and not the change in relative abundance of the species. In the next step, we will try to infer the causal relationships between the change in relative abundance of the species and the metabolite concentrations.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Let&#39;s calculate the change in relative abundance for each genus over time from the imputed_data dataset</span>

<span class="c1"># Calculate the change in relative abundance for each genus</span>
<span class="n">data_change</span> <span class="o">=</span> <span class="n">imputed_data</span><span class="o">.</span><span class="n">diff</span><span class="p">()</span>

<span class="c1"># Make the first row as 0.0 replacing NaN</span>
<span class="n">data_change</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
<br/><br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChange</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data_change</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">)</span>
<span class="n">inferChange</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 25:07<00:00 Sampling 2 chains, 0 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 1512 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.450  0.430  -0.014    0.905      0.304    0.257       3.0
Ah[0, 1]    0.005  0.005  -0.001    0.014      0.003    0.002       4.0
Ah[0, 2]   -0.113  0.035  -0.175   -0.051      0.019    0.015       4.0
Ah[0, 3]    0.015  0.008   0.003    0.030      0.004    0.003       4.0
Ah[0, 4]   -0.005  0.033  -0.067    0.041      0.022    0.019       2.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.072  0.642  -1.179    1.161      0.281    0.211       5.0
Bh[20, 8]  -0.061  0.134  -0.324    0.193      0.044    0.032      10.0
Bh[20, 9]  -0.115  0.850  -1.499    1.713      0.144    0.102      37.0
Bh[20, 10] -0.329  0.523  -1.303    0.686      0.243    0.184       4.0
Bh[20, 11]  0.148  0.412  -0.560    0.820      0.197    0.150       5.0

            ess_tail  r_hat
Ah[0, 0]        29.0   2.33
Ah[0, 1]        49.0   1.70
Ah[0, 2]        14.0   1.48
Ah[0, 3]        16.0   1.42
Ah[0, 4]        11.0   2.40
...              ...    ...
Bh[20, 7]       57.0   1.29
Bh[20, 8]       14.0   1.15
Bh[20, 9]       76.0   1.09
Bh[20, 10]      11.0   1.42
Bh[20, 11]      28.0   1.37

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_78_5.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_78_5.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChange</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.450  0.430  -0.014    0.905      0.304    0.257       3.0
Ah[0, 1]    0.005  0.005  -0.001    0.014      0.003    0.002       4.0
Ah[0, 2]   -0.113  0.035  -0.175   -0.051      0.019    0.015       4.0
Ah[0, 3]    0.015  0.008   0.003    0.030      0.004    0.003       4.0
Ah[0, 4]   -0.005  0.033  -0.067    0.041      0.022    0.019       2.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.072  0.642  -1.179    1.161      0.281    0.211       5.0
Bh[20, 8]  -0.061  0.134  -0.324    0.193      0.044    0.032      10.0
Bh[20, 9]  -0.115  0.850  -1.499    1.713      0.144    0.102      37.0
Bh[20, 10] -0.329  0.523  -1.303    0.686      0.243    0.184       4.0
Bh[20, 11]  0.148  0.412  -0.560    0.820      0.197    0.150       5.0

            ess_tail  r_hat
Ah[0, 0]        29.0   2.33
Ah[0, 1]        49.0   1.70
Ah[0, 2]        14.0   1.48
Ah[0, 3]        16.0   1.42
Ah[0, 4]        11.0   2.40
...              ...    ...
Bh[20, 7]       57.0   1.29
Bh[20, 8]       14.0   1.15
Bh[20, 9]       76.0   1.09
Bh[20, 10]      11.0   1.42
Bh[20, 11]      28.0   1.37

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_79_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_79_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_79_2.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_79_2.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statsmodels.tsa.api</span> <span class="kn">import</span> <span class="n">VAR</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># make VAR model with the imputed data set and statsmodels library</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">VAR</span><span class="p">(</span><span class="n">data_change</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">maxlags</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ml&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning: An unsupported index was provided and will be ignored when e.g. forecasting.
  self._init_dates(dates, freq)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">coefficients</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">params</span>
<span class="n">covariance_matrix</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">sigma_u</span>
<span class="n">intercepts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">results</span><span class="o">.</span><span class="n">params</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

<span class="c1"># remove the first row of the coefficients matrix</span>
<span class="n">coefficients</span> <span class="o">=</span> <span class="n">coefficients</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
<span class="c1"># transpose the coefficients matrix</span>
<span class="n">coefficients</span> <span class="o">=</span> <span class="n">coefficients</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data_change</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">,</span>
                           <span class="n">coefficients</span><span class="o">=</span><span class="n">coefficients</span><span class="p">,</span> <span class="n">intercepts</span><span class="o">=</span><span class="n">intercepts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">(</span><span class="n">tune</span><span class="o">=</span><span class="mi">5000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='14000' class='' max='14000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [14000/14000 54:33<00:00 Sampling 2 chains, 0 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 5_000 tune and 2_000 draw iterations (10_000 + 4_000 draws total) took 3278 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.901  0.021   0.863    0.940      0.002    0.001     191.0
Ah[0, 1]    0.010  0.005   0.001    0.020      0.001    0.001      47.0
Ah[0, 2]   -0.087  0.023  -0.133   -0.046      0.002    0.002      96.0
Ah[0, 3]    0.023  0.009   0.008    0.041      0.002    0.001      26.0
Ah[0, 4]    0.019  0.009   0.002    0.036      0.001    0.001     156.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]  -0.053  0.318  -0.564    0.505      0.178    0.139       3.0
Bh[20, 8]  -0.082  0.222  -0.432    0.329      0.098    0.074       6.0
Bh[20, 9]  -0.030  0.955  -1.862    1.626      0.120    0.085      64.0
Bh[20, 10]  0.158  0.786  -1.210    1.839      0.139    0.099      33.0
Bh[20, 11] -0.276  0.682  -1.746    0.773      0.157    0.201      21.0

            ess_tail  r_hat
Ah[0, 0]       382.0   1.02
Ah[0, 1]       154.0   1.05
Ah[0, 2]       278.0   1.05
Ah[0, 3]        60.0   1.07
Ah[0, 4]       441.0   1.01
...              ...    ...
Bh[20, 7]       19.0   1.58
Bh[20, 8]       30.0   1.29
Bh[20, 9]      244.0   1.01
Bh[20, 10]      52.0   1.07
Bh[20, 11]      31.0   1.23

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_84_5.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_84_5.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.901  0.021   0.863    0.940      0.002    0.001     191.0
Ah[0, 1]    0.010  0.005   0.001    0.020      0.001    0.001      47.0
Ah[0, 2]   -0.087  0.023  -0.133   -0.046      0.002    0.002      96.0
Ah[0, 3]    0.023  0.009   0.008    0.041      0.002    0.001      26.0
Ah[0, 4]    0.019  0.009   0.002    0.036      0.001    0.001     156.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]  -0.053  0.318  -0.564    0.505      0.178    0.139       3.0
Bh[20, 8]  -0.082  0.222  -0.432    0.329      0.098    0.074       6.0
Bh[20, 9]  -0.030  0.955  -1.862    1.626      0.120    0.085      64.0
Bh[20, 10]  0.158  0.786  -1.210    1.839      0.139    0.099      33.0
Bh[20, 11] -0.276  0.682  -1.746    0.773      0.157    0.201      21.0

            ess_tail  r_hat
Ah[0, 0]       382.0   1.02
Ah[0, 1]       154.0   1.05
Ah[0, 2]       278.0   1.05
Ah[0, 3]        60.0   1.07
Ah[0, 4]       441.0   1.01
...              ...    ...
Bh[20, 7]       19.0   1.58
Bh[20, 8]       30.0   1.29
Bh[20, 9]      244.0   1.01
Bh[20, 10]      52.0   1.07
Bh[20, 11]      31.0   1.23

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_85_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_85_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_85_2.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_85_2.png" />
</div>
</div>
<p>We can also increase the increase the burn-in and number of samples to get more accurate results.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior2</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data_change</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">,</span>
                           <span class="n">coefficients</span><span class="o">=</span><span class="n">coefficients</span><span class="p">,</span> <span class="n">intercepts</span><span class="o">=</span><span class="n">intercepts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior2</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">(</span><span class="n">tune</span><span class="o">=</span><span class="mi">8000</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="mi">4000</span><span class="p">,</span> <span class="n">cores</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='48000' class='' max='48000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [48000/48000 8:44:26<00:00 Sampling 4 chains, 0 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 4 chains for 8_000 tune and 4_000 draw iterations (32_000 + 16_000 draws total) took 31473 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Chain 0 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 1 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 2 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 3 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.900  0.022   0.859    0.941      0.001    0.001     361.0
Ah[0, 1]    0.010  0.005   0.001    0.018      0.000    0.000     153.0
Ah[0, 2]   -0.086  0.022  -0.126   -0.044      0.001    0.001     416.0
Ah[0, 3]    0.023  0.007   0.010    0.037      0.001    0.001      56.0
Ah[0, 4]    0.019  0.010   0.001    0.040      0.001    0.000     289.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.071  0.320  -0.467    0.729      0.097    0.070      11.0
Bh[20, 8]   0.065  0.252  -0.367    0.574      0.046    0.033      28.0
Bh[20, 9]  -0.062  0.976  -1.856    1.787      0.056    0.039     308.0
Bh[20, 10]  0.087  0.503  -0.831    1.133      0.050    0.035     101.0
Bh[20, 11]  0.071  0.419  -0.653    0.943      0.061    0.044      46.0

            ess_tail  r_hat
Ah[0, 0]       910.0   1.01
Ah[0, 1]       329.0   1.02
Ah[0, 2]       966.0   1.01
Ah[0, 3]       256.0   1.07
Ah[0, 4]       619.0   1.01
...              ...    ...
Bh[20, 7]       85.0   1.29
Bh[20, 8]       94.0   1.14
Bh[20, 9]      573.0   1.02
Bh[20, 10]     176.0   1.02
Bh[20, 11]     114.0   1.09

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_88_5.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_88_5.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior2</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.900  0.022   0.859    0.941      0.001    0.001     361.0
Ah[0, 1]    0.010  0.005   0.001    0.018      0.000    0.000     153.0
Ah[0, 2]   -0.086  0.022  -0.126   -0.044      0.001    0.001     416.0
Ah[0, 3]    0.023  0.007   0.010    0.037      0.001    0.001      56.0
Ah[0, 4]    0.019  0.010   0.001    0.040      0.001    0.000     289.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.071  0.320  -0.467    0.729      0.097    0.070      11.0
Bh[20, 8]   0.065  0.252  -0.367    0.574      0.046    0.033      28.0
Bh[20, 9]  -0.062  0.976  -1.856    1.787      0.056    0.039     308.0
Bh[20, 10]  0.087  0.503  -0.831    1.133      0.050    0.035     101.0
Bh[20, 11]  0.071  0.419  -0.653    0.943      0.061    0.044      46.0

            ess_tail  r_hat
Ah[0, 0]       910.0   1.01
Ah[0, 1]       329.0   1.02
Ah[0, 2]       966.0   1.01
Ah[0, 3]       256.0   1.07
Ah[0, 4]       619.0   1.01
...              ...    ...
Bh[20, 7]       85.0   1.29
Bh[20, 8]       94.0   1.14
Bh[20, 9]      573.0   1.02
Bh[20, 10]     176.0   1.02
Bh[20, 11]     114.0   1.09

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_89_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_89_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_89_2.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_89_2.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior3</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data_change</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">,</span>
                           <span class="n">coefficients</span><span class="o">=</span><span class="n">coefficients</span><span class="p">,</span> <span class="n">intercepts</span><span class="o">=</span><span class="n">intercepts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior3</span><span class="o">.</span><span class="n">run_inference_large_xs</span><span class="p">(</span><span class="n">tune</span><span class="o">=</span><span class="mi">5000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [c2_A, tau_A, lam_A, Ah, c2_B, tau_B, lam_B, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='12861' class='' max='36000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  35.73% [12861/36000 1:20:10<2:24:15 Sampling 4 chains, 0 divergences]
</div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-red-intense-fg ansi-bold">---------------------------------------------------------------------------</span>
<span class="ansi-red-intense-fg ansi-bold">ValueError</span>                                Traceback (most recent call last)
Cell <span class="ansi-green-intense-fg ansi-bold">In[15], line 1</span>
<span class="ansi-green-intense-fg ansi-bold">----&gt; 1</span> inferChageWithPrior3<span style="color: rgb(98,98,98)">.</span>run_inference_large_xs(tune<span style="color: rgb(98,98,98)">=</span><span style="color: rgb(98,98,98)">5000</span>)

File <span class="ansi-green-intense-fg ansi-bold">~\Documents\GitHub\MIMIC\mimic\model_infer\infer_VAR_bayes.py:418</span>, in <span class="ansi-cyan-fg">infer_VAR.run_inference_large_xs</span><span class="ansi-blue-intense-fg ansi-bold">(self, samples, tune, cores)</span>
<span class="ansi-green-fg">    414</span>     likelihood <span style="color: rgb(98,98,98)">=</span> pm<span style="color: rgb(98,98,98)">.</span>Normal(<span style="color: rgb(175,0,0)">&#39;</span><span style="color: rgb(175,0,0)">likelihood</span><span style="color: rgb(175,0,0)">&#39;</span>, mu<span style="color: rgb(98,98,98)">=</span>mu<span style="color: rgb(98,98,98)">.</span>T,
<span class="ansi-green-fg">    415</span>                            sigma<span style="color: rgb(98,98,98)">=</span>sigma, observed<span style="color: rgb(98,98,98)">=</span>data[<span style="color: rgb(98,98,98)">2</span>:, :])
<span class="ansi-green-fg">    417</span> <span class="ansi-bold" style="color: rgb(0,135,0)">with</span> var_model:
<span class="ansi-green-intense-fg ansi-bold">--&gt; 418</span>     trace <span style="color: rgb(98,98,98)">=</span> pm<span style="color: rgb(98,98,98)">.</span>sample(samples, tune<span style="color: rgb(98,98,98)">=</span>tune, cores<span style="color: rgb(98,98,98)">=</span>cores)
<span class="ansi-green-fg">    420</span> <span style="color: rgb(0,135,0)">print</span>(az<span style="color: rgb(98,98,98)">.</span>summary(trace, var_names<span style="color: rgb(98,98,98)">=</span>[<span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">Ah</span><span style="color: rgb(175,0,0)">&#34;</span>, <span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">Bh</span><span style="color: rgb(175,0,0)">&#34;</span>]))
<span class="ansi-green-fg">    421</span> az<span style="color: rgb(98,98,98)">.</span>plot_posterior(trace, var_names<span style="color: rgb(98,98,98)">=</span>[<span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">Ah</span><span style="color: rgb(175,0,0)">&#34;</span>, <span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">Bh</span><span style="color: rgb(175,0,0)">&#34;</span>])

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\pymc\sampling\mcmc.py:827</span>, in <span class="ansi-cyan-fg">sample</span><span class="ansi-blue-intense-fg ansi-bold">(draws, tune, chains, cores, random_seed, progressbar, step, nuts_sampler, initvals, init, jitter_max_retries, n_init, trace, discard_tuned_samples, compute_convergence_checks, keep_warning_stat, return_inferencedata, idata_kwargs, nuts_sampler_kwargs, callback, mp_ctx, model, **kwargs)</span>
<span class="ansi-green-fg">    823</span> t_sampling <span style="color: rgb(98,98,98)">=</span> time<span style="color: rgb(98,98,98)">.</span>time() <span style="color: rgb(98,98,98)">-</span> t_start
<span class="ansi-green-fg">    825</span> <span style="color: rgb(95,135,135)"># Packaging, validating and returning the result was extracted</span>
<span class="ansi-green-fg">    826</span> <span style="color: rgb(95,135,135)"># into a function to make it easier to test and refactor.</span>
<span class="ansi-green-intense-fg ansi-bold">--&gt; 827</span> <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> _sample_return(
<span class="ansi-green-fg">    828</span>     run<span style="color: rgb(98,98,98)">=</span>run,
<span class="ansi-green-fg">    829</span>     traces<span style="color: rgb(98,98,98)">=</span>traces,
<span class="ansi-green-fg">    830</span>     tune<span style="color: rgb(98,98,98)">=</span>tune,
<span class="ansi-green-fg">    831</span>     t_sampling<span style="color: rgb(98,98,98)">=</span>t_sampling,
<span class="ansi-green-fg">    832</span>     discard_tuned_samples<span style="color: rgb(98,98,98)">=</span>discard_tuned_samples,
<span class="ansi-green-fg">    833</span>     compute_convergence_checks<span style="color: rgb(98,98,98)">=</span>compute_convergence_checks,
<span class="ansi-green-fg">    834</span>     return_inferencedata<span style="color: rgb(98,98,98)">=</span>return_inferencedata,
<span class="ansi-green-fg">    835</span>     keep_warning_stat<span style="color: rgb(98,98,98)">=</span>keep_warning_stat,
<span class="ansi-green-fg">    836</span>     idata_kwargs<span style="color: rgb(98,98,98)">=</span>idata_kwargs <span class="ansi-bold" style="color: rgb(175,0,255)">or</span> {},
<span class="ansi-green-fg">    837</span>     model<span style="color: rgb(98,98,98)">=</span>model,
<span class="ansi-green-fg">    838</span> )

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\pymc\sampling\mcmc.py:858</span>, in <span class="ansi-cyan-fg">_sample_return</span><span class="ansi-blue-intense-fg ansi-bold">(run, traces, tune, t_sampling, discard_tuned_samples, compute_convergence_checks, return_inferencedata, keep_warning_stat, idata_kwargs, model)</span>
<span class="ansi-green-fg">    856</span> <span style="color: rgb(95,135,135)"># Pick and slice chains to keep the maximum number of samples</span>
<span class="ansi-green-fg">    857</span> <span class="ansi-bold" style="color: rgb(0,135,0)">if</span> discard_tuned_samples:
<span class="ansi-green-intense-fg ansi-bold">--&gt; 858</span>     traces, length <span style="color: rgb(98,98,98)">=</span> _choose_chains(traces, tune)
<span class="ansi-green-fg">    859</span> <span class="ansi-bold" style="color: rgb(0,135,0)">else</span>:
<span class="ansi-green-fg">    860</span>     traces, length <span style="color: rgb(98,98,98)">=</span> _choose_chains(traces, <span style="color: rgb(98,98,98)">0</span>)

File <span class="ansi-green-intense-fg ansi-bold">c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\pymc\backends\base.py:601</span>, in <span class="ansi-cyan-fg">_choose_chains</span><span class="ansi-blue-intense-fg ansi-bold">(traces, tune)</span>
<span class="ansi-green-fg">    599</span> lengths <span style="color: rgb(98,98,98)">=</span> [<span style="color: rgb(0,135,0)">max</span>(<span style="color: rgb(98,98,98)">0</span>, <span style="color: rgb(0,135,0)">len</span>(trace) <span style="color: rgb(98,98,98)">-</span> tune) <span class="ansi-bold" style="color: rgb(0,135,0)">for</span> trace <span class="ansi-bold" style="color: rgb(175,0,255)">in</span> traces]
<span class="ansi-green-fg">    600</span> <span class="ansi-bold" style="color: rgb(0,135,0)">if</span> <span class="ansi-bold" style="color: rgb(175,0,255)">not</span> <span style="color: rgb(0,135,0)">sum</span>(lengths):
<span class="ansi-green-intense-fg ansi-bold">--&gt; 601</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">raise</span> <span class="ansi-bold" style="color: rgb(215,95,95)">ValueError</span>(<span style="color: rgb(175,0,0)">&#34;</span><span style="color: rgb(175,0,0)">Not enough samples to build a trace.</span><span style="color: rgb(175,0,0)">&#34;</span>)
<span class="ansi-green-fg">    603</span> idxs <span style="color: rgb(98,98,98)">=</span> np<span style="color: rgb(98,98,98)">.</span>argsort(lengths)
<span class="ansi-green-fg">    604</span> l_sort <span style="color: rgb(98,98,98)">=</span> np<span style="color: rgb(98,98,98)">.</span>array(lengths)[idxs]

<span class="ansi-red-intense-fg ansi-bold">ValueError</span>: Not enough samples to build a trace.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior3</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean   sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.912  0.0   0.912    0.912        0.0      0.0     298.0
Ah[0, 1]    0.006  0.0   0.006    0.006        0.0      0.0     169.0
Ah[0, 2]   -0.084  0.0  -0.084   -0.084        0.0      0.0     512.0
Ah[0, 3]    0.014  0.0   0.014    0.014        0.0      0.0     211.0
Ah[0, 4]    0.016  0.0   0.016    0.016        0.0      0.0     157.0
...           ...  ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.000  0.0  -0.000    0.000        0.0      0.0     128.0
Bh[20, 8]   0.000  0.0  -0.000    0.000        0.0      0.0     201.0
Bh[20, 9]   0.000  0.0  -0.000    0.000        0.0      0.0     363.0
Bh[20, 10] -0.000  0.0  -0.000    0.000        0.0      0.0     322.0
Bh[20, 11]  0.000  0.0  -0.000    0.000        0.0      0.0     520.0

            ess_tail  r_hat
Ah[0, 0]       185.0   1.15
Ah[0, 1]        71.0   1.13
Ah[0, 2]       409.0   1.10
Ah[0, 3]       150.0   1.09
Ah[0, 4]        55.0   1.08
...              ...    ...
Bh[20, 7]       33.0   1.11
Bh[20, 8]      141.0   1.09
Bh[20, 9]      219.0   1.09
Bh[20, 10]     241.0   1.12
Bh[20, 11]     244.0   1.05

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_92_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_92_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_92_2.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-Metabolites_92_2.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now we are going to check for stationarity in the data using adf and kpss tests from stasmodels library</span>
<span class="c1"># We will use the imputed data set</span>

<span class="kn">import</span> <span class="nn">statsmodels.api</span> <span class="k">as</span> <span class="nn">sm</span>
<span class="kn">from</span> <span class="nn">statsmodels.tsa.stattools</span> <span class="kn">import</span> <span class="n">adfuller</span><span class="p">,</span> <span class="n">kpss</span>


<span class="k">def</span> <span class="nf">adf_test</span><span class="p">(</span><span class="n">timeseries</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Results of Dickey-Fuller Test:&quot;</span><span class="p">)</span>
    <span class="n">dftest</span> <span class="o">=</span> <span class="n">adfuller</span><span class="p">(</span><span class="n">timeseries</span><span class="p">,</span> <span class="n">autolag</span><span class="o">=</span><span class="s2">&quot;AIC&quot;</span><span class="p">)</span>
    <span class="n">dfoutput</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
        <span class="n">dftest</span><span class="p">[:</span><span class="mi">4</span><span class="p">],</span>
        <span class="n">index</span><span class="o">=</span><span class="p">[</span>
            <span class="s2">&quot;Test Statistic&quot;</span><span class="p">,</span>
            <span class="s2">&quot;p-value&quot;</span><span class="p">,</span>
            <span class="s2">&quot;#Lags Used&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Number of Observations Used&quot;</span><span class="p">,</span>
        <span class="p">],</span>
    <span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">dftest</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">dfoutput</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Critical Value (</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">dfoutput</span><span class="p">)</span>
    <span class="c1"># if p-value is less than 0.05, we reject the null hypothesis and say that the data is stationary</span>
    <span class="c1"># print if the data is stationary or not</span>
    <span class="k">if</span> <span class="n">dftest</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mf">0.05</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ADF Test: Data is stationary&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ADF Test: Data is not stationary&quot;</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">kpss_test</span><span class="p">(</span><span class="n">timeseries</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Results of KPSS Test:&quot;</span><span class="p">)</span>
    <span class="n">kpsstest</span> <span class="o">=</span> <span class="n">kpss</span><span class="p">(</span><span class="n">timeseries</span><span class="p">,</span> <span class="n">regression</span><span class="o">=</span><span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="n">nlags</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">)</span>
    <span class="n">kpss_output</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
        <span class="n">kpsstest</span><span class="p">[:</span><span class="mi">3</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Test Statistic&quot;</span><span class="p">,</span> <span class="s2">&quot;p-value&quot;</span><span class="p">,</span> <span class="s2">&quot;Lags Used&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">kpsstest</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">kpss_output</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Critical Value (</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">kpss_output</span><span class="p">)</span>
    <span class="c1"># if p-value is greater than 0.05, we reject the null hypothesis and say that the data is stationary</span>
    <span class="c1"># print if the data is stationary or not</span>
    <span class="k">if</span> <span class="n">kpsstest</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mf">0.05</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;KPSS test: Data is stationary&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;KPSS test: Data is not stationary&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check for stationarity in the data imputed_dataset (without the &#39;days&#39; column)</span>
<span class="k">for</span> <span class="n">genus</span> <span class="ow">in</span> <span class="n">data_change</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">genus</span><span class="p">)</span>
    <span class="n">adf_test</span><span class="p">(</span><span class="n">data_change</span><span class="p">[</span><span class="n">genus</span><span class="p">])</span>
    <span class="n">kpss_test</span><span class="p">(</span><span class="n">data_change</span><span class="p">[</span><span class="n">genus</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Acidimicrobium
Results of Dickey-Fuller Test:
Test Statistic                  -4.877743
p-value                          0.000039
#Lags Used                      16.000000
Number of Observations Used    393.000000
Critical Value (1%)             -3.447099
Critical Value (5%)             -2.868923
Critical Value (10%)            -2.570703
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.032404
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Acinetobacter
Results of Dickey-Fuller Test:
Test Statistic                  -4.598946
p-value                          0.000130
#Lags Used                      17.000000
Number of Observations Used    392.000000
Critical Value (1%)             -3.447142
Critical Value (5%)             -2.868941
Critical Value (10%)            -2.570713
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.173686
p-value                   0.100000
Lags Used                10.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Albidiferax
Results of Dickey-Fuller Test:
Test Statistic                  -3.829655
p-value                          0.002618
#Lags Used                      11.000000
Number of Observations Used    398.000000
Critical Value (1%)             -3.446888
Critical Value (5%)             -2.868829
Critical Value (10%)            -2.570653
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.043204
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Candidatus Microthrix
Results of Dickey-Fuller Test:
Test Statistic                -7.937074e+00
p-value                        3.391025e-12
#Lags Used                     6.000000e+00
Number of Observations Used    4.030000e+02
Critical Value (1%)           -3.446681e+00
Critical Value (5%)           -2.868739e+00
Critical Value (10%)          -2.570605e+00
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.094575
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Chitinophaga pinensis
Results of Dickey-Fuller Test:
Test Statistic                  -3.816479
p-value                          0.002743
#Lags Used                      16.000000
Number of Observations Used    393.000000
Critical Value (1%)             -3.447099
Critical Value (5%)             -2.868923
Critical Value (10%)            -2.570703
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.059832
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Dechloromonas
Results of Dickey-Fuller Test:
Test Statistic                  -3.320016
p-value                          0.014007
#Lags Used                      15.000000
Number of Observations Used    394.000000
Critical Value (1%)             -3.447057
Critical Value (5%)             -2.868904
Critical Value (10%)            -2.570693
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.106131
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Haliscomenobacter
Results of Dickey-Fuller Test:
Test Statistic                  -3.790187
p-value                          0.003009
#Lags Used                      17.000000
Number of Observations Used    392.000000
Critical Value (1%)             -3.447142
Critical Value (5%)             -2.868941
Critical Value (10%)            -2.570713
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.099982
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Intrasporangium
Results of Dickey-Fuller Test:
Test Statistic                  -4.249279
p-value                          0.000543
#Lags Used                      18.000000
Number of Observations Used    391.000000
Critical Value (1%)             -3.447186
Critical Value (5%)             -2.868960
Critical Value (10%)            -2.570723
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.197429
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Leptospira
Results of Dickey-Fuller Test:
Test Statistic                  -3.825629
p-value                          0.002656
#Lags Used                      17.000000
Number of Observations Used    392.000000
Critical Value (1%)             -3.447142
Critical Value (5%)             -2.868941
Critical Value (10%)            -2.570713
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.039661
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Other
Results of Dickey-Fuller Test:
Test Statistic                -7.502466e+00
p-value                        4.218178e-11
#Lags Used                     6.000000e+00
Number of Observations Used    4.030000e+02
Critical Value (1%)           -3.446681e+00
Critical Value (5%)           -2.868739e+00
Critical Value (10%)          -2.570605e+00
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.075135
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Xanthomonas
Results of Dickey-Fuller Test:
Test Statistic                  -3.684859
p-value                          0.004331
#Lags Used                      18.000000
Number of Observations Used    391.000000
Critical Value (1%)             -3.447186
Critical Value (5%)             -2.868960
Critical Value (10%)            -2.570723
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.1455
p-value                   0.1000
Lags Used                11.0000
Critical Value (10%)      0.3470
Critical Value (5%)       0.4630
Critical Value (2.5%)     0.5740
Critical Value (1%)       0.7390
dtype: float64
KPSS test: Data is stationary


mean abundance &lt; 2%
Results of Dickey-Fuller Test:
Test Statistic                  -4.761013
p-value                          0.000065
#Lags Used                       7.000000
Number of Observations Used    402.000000
Critical Value (1%)             -3.446722
Critical Value (5%)             -2.868757
Critical Value (10%)            -2.570614
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.189282
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\User\AppData\Local\Temp\ipykernel_42636\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_42636\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_42636\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_42636\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_42636\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_42636\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_42636\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_42636\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_42636\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_42636\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_42636\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_42636\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import the matrix A from the model</span>

<span class="n">posterior_samples</span> <span class="o">=</span> <span class="n">az</span><span class="o">.</span><span class="n">from_netcdf</span><span class="p">(</span><span class="s1">&#39;model_posterior.nc&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">posterior</span><span class="p">[</span><span class="s1">&#39;Ah&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute the median of the posterior samples</span>
<span class="n">median_Ah</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">posterior_samples</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to check stability of a single matrix</span>
<span class="k">def</span> <span class="nf">is_stable</span><span class="p">(</span><span class="n">matrix</span><span class="p">):</span>
    <span class="n">eigenvalues</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">eigvals</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
    <span class="c1">#return all(np.sqrt(np.real(eigenvalue)**2 + np.imag(eigenvalue)**2) &lt; 1 for eigenvalue in eigenvalues)</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">eigenvalues</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">)</span>

<span class="c1"># Check stability of the median matrix</span>
<span class="n">stability_median_Ah</span> <span class="o">=</span> <span class="n">is_stable</span><span class="p">(</span><span class="n">median_Ah</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Median Ah matrix stability:&quot;</span><span class="p">,</span> <span class="n">stability_median_Ah</span><span class="p">)</span>
<span class="k">if</span> <span class="n">stability_median_Ah</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The median Ah matrix is stable.&quot;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The median Ah matrix is not stable.&quot;</span><span class="p">)</span>
<br/><br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Median Ah matrix stability: True
The median Ah matrix is stable.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to check stability of a single matrix</span>
<span class="k">def</span> <span class="nf">is_stable</span><span class="p">(</span><span class="n">matrix</span><span class="p">):</span>
    <span class="n">eigenvalues</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">eigvals</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">eigenvalues</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">)</span>

<span class="c1"># Check stability for each sampled matrix</span>
<span class="n">num_chains</span><span class="p">,</span> <span class="n">num_samples</span><span class="p">,</span> <span class="n">nX</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">posterior_samples</span><span class="o">.</span><span class="n">shape</span>
<span class="n">stability_results</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">num_chains</span><span class="p">,</span> <span class="n">num_samples</span><span class="p">))</span>

<span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_chains</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_samples</span><span class="p">):</span>
        <span class="n">stability_results</span><span class="p">[</span><span class="n">chain</span><span class="p">,</span> <span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">is_stable</span><span class="p">(</span><span class="n">posterior_samples</span><span class="p">[</span><span class="n">chain</span><span class="p">,</span> <span class="n">sample</span><span class="p">])</span>

<span class="c1"># Percentage of stable samples</span>
<span class="n">stable_percentage</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">stability_results</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Percentage of stable samples: </span><span class="si">{</span><span class="n">stable_percentage</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>

<span class="k">if</span> <span class="n">stable_percentage</span> <span class="o">&lt;</span> <span class="mi">100</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: Not all samples are stable.&quot;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;All samples are stable.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Percentage of stable samples: 83.94%
Warning: Not all samples are stable.
</pre></div></div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">MIMIC</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../readme.html">Modelling and Inference of MICrobiomes Project (MIMIC)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">mimic</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../../examples.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="examples-Herold.html">Using the Data Imputator on a Real Public Dataset, Herold et al. (2020)</a></li>
<li class="toctree-l2"><a class="reference internal" href="examples-Herold.html#Separate-the-data-into-MG-and-MT-datasets">Separate the data into MG and MT datasets</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Using the Data Imputator on a Real Public Dataset, Herold et al. (2020)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Separate-the-data-into-MG-and-MT-datasets">Separate the data into MG and MT datasets</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Reducing-number-of-variables-(to-the-3-or-5-most-abundant)-to-reduce-the-computational-cost-and-simplify-the-analysis">Reducing number of variables (to the 3 or 5 most abundant) to reduce the computational cost and simplify the analysis</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../gMLV/examples-sim-gMLV.html">Simulate some time course data from the gMLV</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gMLV/examples-ridge-lasso-gMLV.html">Run ridge regression and lasso for the gMLV model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../MVAR/examples-infer-MVAR.html">MVAR Simulation for Microbiota and Metabolites</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../MVAR/examples-sim-MVAR.html">MVAR Simulation for Microbiota and Metabolites</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../VAR/examples-sim-VAR.html">Vector AutoRegression (VAR) Simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../VAR/examples-bayes-VAR.html">Bayesian Inference for Vector AutoRegression (VAR) Models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gLV/examples-Stein.html">Repeat Stein et al. 2013 analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gLV/examples-bayes-gLV.html">Used Bayesian inference to infer the parameters of a (linearised) gLV model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gLV/examples-lasso-gLV.html">Comparison of ridge vs lasso regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gLV/examples-sim-gLV.html">Simulate some time course data from the gLV</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gLV/examples-ridge-gLV.html">Simulate some time course data and perform ridge regression as in Stein et al. 2013</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../GP/examples-impute-GP_Stein.html">Using the Data Imputator on a Real Public Dataset - Stein et al. (2013)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../GP/examples-impute-GP.html">Imputing Missing Values using Gaussian Process</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../contributing.html">Contributing to MIMIC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../history.html">History</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../../examples.html">Examples</a><ul>
      <li>Previous: <a href="examples_impute_data.html" title="previous chapter">&lt;no title&gt;</a></li>
      <li>Next: <a href="../../gMLV/examples-sim-gMLV.html" title="next chapter">Simulate some time course data from the gMLV</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2024, Pedro Fontanarrosa.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.3.7</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
      |
      <a href="../../../_sources/notebooks/MultiModel/Herold/examples-Herold-Metabolites.ipynb.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>