<!DOCTYPE html>

<html lang="en" data-content_root="../../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Sparse VAR on a Real Public Dataset, Herold et al. (2020) &#8212; MIMIC 0.1.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="../../../_static/alabaster.css?v=12dfc556" />
    <link rel="stylesheet" type="text/css" href="../../../_static/nbsphinx-code-cells.css?v=2aa19091" />
    <script src="../../../_static/documentation_options.js?v=01f34227"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Data imputation on a Real Public Dataset, Herold et al. (2020)" href="examples_impute_data.html" />
    <link rel="prev" title="VAR inference on a Real Public Dataset, Herold et al. (2020)" href="examples-Herold-VAR.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="Sparse-VAR-on-a-Real-Public-Dataset,-Herold-et-al.-(2020)">
<h1>Sparse VAR on a Real Public Dataset, Herold et al. (2020)<a class="headerlink" href="#Sparse-VAR-on-a-Real-Public-Dataset,-Herold-et-al.-(2020)" title="Link to this heading">¶</a></h1>
<p>In this example, we are goint to import the metabolite data and do a sparse VAR inference method to infer the causal relationships between the variables (relative abundances and metabolite concentrations) in the time series dataset available at <a class="reference external" href="https://doi.org/10.1038/s41467-020-19006-2">Herold et al., 2020</a>.</p>
<p>First, we will import the neccessary libraries and load the dataset.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>


<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>


<span class="kn">from</span> <span class="nn">mimic.data_imputation.impute_GP</span> <span class="kn">import</span> <span class="n">GPImputer</span>
<span class="kn">from</span> <span class="nn">mimic.model_infer.infer_VAR_bayes</span> <span class="kn">import</span> <span class="o">*</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING:tensorflow:From c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\tf_keras\src\losses.py:2976: The name tf.losses.sparse_softmax_cross_entropy is deprecated. Please use tf.compat.v1.losses.sparse_softmax_cross_entropy instead.

WARNING:tensorflow:From c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\tensorflow_probability\python\internal\backend\numpy\_utils.py:48: The name tf.logging.TaskLevelStatusMessage is deprecated. Please use tf.compat.v1.logging.TaskLevelStatusMessage instead.

WARNING:tensorflow:From c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\tensorflow_probability\python\internal\backend\numpy\_utils.py:48: The name tf.control_flow_v2_enabled is deprecated. Please use tf.compat.v1.control_flow_v2_enabled instead.

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
WARNING (pytensor.tensor.blas): Using NumPy C-API based implementation for BLAS functions.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Source Data/fig3ab.tsv&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>genus</th>
      <th>date</th>
      <th>val</th>
      <th>type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Acidimicrobium</td>
      <td>2011-03-21</td>
      <td>8.322484</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Acidimicrobium</td>
      <td>2011-03-29</td>
      <td>7.664534</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Acidimicrobium</td>
      <td>2011-04-05</td>
      <td>7.871188</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Acidimicrobium</td>
      <td>2011-04-14</td>
      <td>6.695900</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Acidimicrobium</td>
      <td>2011-04-21</td>
      <td>6.695177</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Acidimicrobium</td>
      <td>2011-04-29</td>
      <td>6.304652</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>6</th>
      <td>Acidimicrobium</td>
      <td>2011-05-06</td>
      <td>7.462918</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Acidimicrobium</td>
      <td>2011-05-13</td>
      <td>7.742366</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Acidimicrobium</td>
      <td>2011-05-20</td>
      <td>6.627358</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>9</th>
      <td>Acidimicrobium</td>
      <td>2011-05-27</td>
      <td>7.425880</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>10</th>
      <td>Acidimicrobium</td>
      <td>2011-06-03</td>
      <td>7.890684</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>11</th>
      <td>Acidimicrobium</td>
      <td>2011-06-09</td>
      <td>6.306718</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>12</th>
      <td>Acidimicrobium</td>
      <td>2011-06-17</td>
      <td>6.607237</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>13</th>
      <td>Acidimicrobium</td>
      <td>2011-06-24</td>
      <td>8.888402</td>
      <td>MG</td>
    </tr>
    <tr>
      <th>14</th>
      <td>Acidimicrobium</td>
      <td>2011-07-01</td>
      <td>9.086751</td>
      <td>MG</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
<section id="Separate-the-data-into-MG-and-MT-datasets">
<h1>Separate the data into MG and MT datasets<a class="headerlink" href="#Separate-the-data-into-MG-and-MT-datasets" title="Link to this heading">¶</a></h1>
<p>The data is separated into two datasets, MG and MT. The MG dataset contains the meta-genomics data, while the MT dataset contains the meta-transcriptomics. We will be using the MG dataset in this example.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># separate the data into MG and MT datasets</span>

<span class="n">dataMG</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;MG&#39;</span><span class="p">]</span>

<span class="n">dataMG</span><span class="o">.</span><span class="n">info</span><span class="p">()</span>

<span class="n">dataMT</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;MT&#39;</span><span class="p">]</span>

<span class="n">dataMT</span><span class="o">.</span><span class="n">info</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;class &#39;pandas.core.frame.DataFrame&#39;&gt;
Index: 612 entries, 0 to 611
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   genus   561 non-null    object
 1   date    612 non-null    object
 2   val     612 non-null    float64
 3   type    612 non-null    object
dtypes: float64(1), object(3)
memory usage: 23.9+ KB
&lt;class &#39;pandas.core.frame.DataFrame&#39;&gt;
Index: 561 entries, 612 to 1172
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   genus   510 non-null    object
 1   date    561 non-null    object
 2   val     561 non-null    float64
 3   type    561 non-null    object
dtypes: float64(1), object(3)
memory usage: 21.9+ KB
</pre></div></div>
</div>
<p>We will proceed with some data wrangling to prepare the data for the imputation process.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Data cleaning</span>
<span class="c1"># Rename Nan in the genus column to &#39;Other&#39;</span>

<span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="s1">&#39;Other&#39;</span><span class="p">)</span>

<span class="c1"># Drop the &#39;type&#39; column</span>
<span class="n">dataMG</span> <span class="o">=</span> <span class="n">dataMG</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">])</span>

<span class="c1"># Change relative Date columns to days</span>
<span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_datetime</span><span class="p">(</span><span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">days</span>

<span class="c1"># Change the &#39;genus&#39; column to a categorical column</span>
<span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>


<span class="n">dataMG</span><span class="o">.</span><span class="n">info</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;class &#39;pandas.core.frame.DataFrame&#39;&gt;
Index: 612 entries, 0 to 611
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   genus   612 non-null    category
 1   date    612 non-null    int64
 2   val     612 non-null    float64
dtypes: category(1), float64(1), int64(1)
memory usage: 15.3 KB
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\User\AppData\Local\Temp\ipykernel_41148\1446871442.py:4: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataMG[&#39;genus&#39;] = dataMG[&#39;genus&#39;].replace(np.nan, &#39;Other&#39;)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the data, relative abundance (value) vs. days (date) for each genus</span>


<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataMG</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Microbial Growth Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1057: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  grouped_data = data.groupby(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_9_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_9_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Pivot the data so each &#39;genus&#39; is a column</span>
<span class="n">dataMG_pivot</span> <span class="o">=</span> <span class="n">dataMG</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Create a stacked area chart</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">stackplot</span><span class="p">(</span><span class="n">dataMG_pivot</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">dataMG_pivot</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">dataMG_pivot</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Microbial Growth Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_10_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_10_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Do the same data cleaning for the MT data set</span>

<span class="c1"># Rename Nan in the genus column to &#39;Other&#39;</span>

<span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="s1">&#39;Other&#39;</span><span class="p">)</span>

<span class="c1"># Drop the &#39;type&#39; column</span>
<span class="n">dataMT</span> <span class="o">=</span> <span class="n">dataMT</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">])</span>

<span class="c1"># Change relative Date columns to days</span>
<span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_datetime</span><span class="p">(</span><span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">days</span>

<span class="c1"># Change the &#39;genus&#39; column to a categorical column</span>
<span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\User\AppData\Local\Temp\ipykernel_41148\2794291335.py:5: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataMT[&#39;genus&#39;] = dataMT[&#39;genus&#39;].replace(np.nan, &#39;Other&#39;)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the data, relative abundance (value) vs. days (date) for each genus</span>


<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataMT</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Microbial Meta Transcriptomics Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1057: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  grouped_data = data.groupby(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Text(0.5, 0, &#39;Days&#39;)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_12_2.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_12_2.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Pivot the data so each &#39;genus&#39; is a column</span>
<span class="n">dataMT_pivot</span> <span class="o">=</span> <span class="n">dataMT</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Pivot the MG data so each &#39;genus&#39; is a column</span>
<span class="n">dataMG_pivot</span> <span class="o">=</span> <span class="n">dataMG</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metabolites</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Source Data/fig4.tsv&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>This dataset contains several columns, each representing different attributes related to microbial activity and metabolite concentrations. Here is a detailed explanation of each column:</p>
<ul class="simple">
<li><p><strong>date</strong>: The date when the sample was collected.</p></li>
<li><p><strong>L1</strong>: The functional category or pathway associated with the data point.</p></li>
<li><p><strong>mp_tot</strong>: Total metaproteomic count for that date and pathway.</p></li>
<li><p><strong>mt_tot</strong>: Total metatranscriptomic count for that date and pathway.</p></li>
<li><p><strong>L1sum_MP</strong>: Sum of metaproteomic counts for that specific functional category (L1).</p></li>
<li><p><strong>L1sum_MT</strong>: Sum of metatranscriptomic counts for that specific functional category (L1).</p></li>
<li><p><strong>L1rel_MP</strong>: Relative abundance of metaproteomic counts within the functional category.</p></li>
<li><p><strong>L1rel_MT</strong>: Relative abundance of metatranscriptomic counts within the functional category.</p></li>
<li><p><strong>MP_std_L1</strong>: Standard deviation of metaproteomic counts for that functional category.</p></li>
<li><p><strong>MT_std_L1</strong>: Standard deviation of metatranscriptomic counts for that functional category.</p></li>
<li><p><strong>cor</strong>: Correlation value between metaproteomic and metatranscriptomic data for that pathway.</p></li>
<li><p><strong>newL1</strong>: A reformatted version of the functional category name, often including a correlation coefficient (r value) for easier readability in the figure.</p></li>
</ul>
<p>We will be using the <code class="docutils literal notranslate"><span class="pre">L1rel_MP</span></code> and <code class="docutils literal notranslate"><span class="pre">L1rel_MT</span></code> columns in this example to use it with the VAR inference method to infer the causal relationships between the variables in the time series with the meta transcriptomics (MT) and meta genomics (MG) data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Letr&#39;s create 2 dataframes, one for the L1rel_MP and the other for the L1rel_MT data, using the date, L1 and L2tel values</span>

<span class="n">dataL1_MP</span> <span class="o">=</span> <span class="n">metabolites</span><span class="p">[[</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="s1">&#39;L1rel_MP&#39;</span><span class="p">]]</span>

<span class="n">dataL1_MT</span> <span class="o">=</span> <span class="n">metabolites</span><span class="p">[[</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="s1">&#39;L1rel_MT&#39;</span><span class="p">]]</span>
<span class="n">dataL1_MP</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>date</th>
      <th>L1</th>
      <th>L1rel_MP</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2011-03-21</td>
      <td>01_Fermentation</td>
      <td>0.096899</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2011-03-21</td>
      <td>02_Homoacetogenesis</td>
      <td>0.009890</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2011-03-21</td>
      <td>03_Superpathway of thiosulfate metabolism (Des...</td>
      <td>0.000579</td>
    </tr>
    <tr>
      <th>3</th>
      <td>2011-03-21</td>
      <td>04_Utililization of sugar, conversion of pento...</td>
      <td>0.000535</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2011-03-21</td>
      <td>05_Fatty acid oxidation</td>
      <td>0.025483</td>
    </tr>
    <tr>
      <th>5</th>
      <td>2011-03-21</td>
      <td>06_Amino acid utilization biosynthesis metabolism</td>
      <td>0.122784</td>
    </tr>
    <tr>
      <th>6</th>
      <td>2011-03-21</td>
      <td>07_Nucleic acid metabolism</td>
      <td>0.119531</td>
    </tr>
    <tr>
      <th>7</th>
      <td>2011-03-21</td>
      <td>08_Hydrocarbon degradation</td>
      <td>0.005524</td>
    </tr>
    <tr>
      <th>8</th>
      <td>2011-03-21</td>
      <td>09_Carbohydrate Active enzyme - CAZy</td>
      <td>0.003921</td>
    </tr>
    <tr>
      <th>9</th>
      <td>2011-03-21</td>
      <td>10_TCA cycle</td>
      <td>0.045398</td>
    </tr>
    <tr>
      <th>10</th>
      <td>2011-03-21</td>
      <td>11_Nitrogen cycle</td>
      <td>0.026063</td>
    </tr>
    <tr>
      <th>11</th>
      <td>2011-03-21</td>
      <td>12_Transporters</td>
      <td>0.015593</td>
    </tr>
    <tr>
      <th>12</th>
      <td>2011-03-21</td>
      <td>13_Hydrogen metabolism</td>
      <td>0.000846</td>
    </tr>
    <tr>
      <th>13</th>
      <td>2011-03-21</td>
      <td>14_Methanogenesis</td>
      <td>0.003520</td>
    </tr>
    <tr>
      <th>14</th>
      <td>2011-03-21</td>
      <td>15_Methylotrophy</td>
      <td>0.009267</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now let&#39;s visualize the data</span>

<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;L1rel_MP&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataL1_MP</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Metaproteomic Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_17_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_17_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Same for the metatranscriptomic data</span>

<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;L1rel_MT&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataL1_MT</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Metatranscriptomic Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_18_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_18_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># And do the same as we did for the MG and MT data sets</span>

<span class="c1"># Rename Nan in the L1 column to &#39;Other&#39;</span>

<span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="s1">&#39;Other&#39;</span><span class="p">)</span>

<span class="c1"># Change relative Date columns to days</span>
<span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_datetime</span><span class="p">(</span><span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">days</span>

<span class="c1"># Change the &#39;L1&#39; column to a categorical column</span>
<span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\User\AppData\Local\Temp\ipykernel_41148\300627487.py:5: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MP[&#39;L1&#39;] = dataL1_MP[&#39;L1&#39;].replace(np.nan, &#39;Other&#39;)
C:\Users\User\AppData\Local\Temp\ipykernel_41148\300627487.py:8: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MP[&#39;date&#39;] = pd.to_datetime(dataL1_MP[&#39;date&#39;])
C:\Users\User\AppData\Local\Temp\ipykernel_41148\300627487.py:9: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MP[&#39;date&#39;] = (dataL1_MP[&#39;date&#39;] - dataL1_MP[&#39;date&#39;].min()).dt.days
C:\Users\User\AppData\Local\Temp\ipykernel_41148\300627487.py:12: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MP[&#39;L1&#39;] = dataL1_MP[&#39;L1&#39;].astype(&#39;category&#39;)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># do the same for the L1rel_MT data set</span>

<span class="c1"># Rename Nan in the L1 column to &#39;Other&#39;</span>

<span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="s1">&#39;Other&#39;</span><span class="p">)</span>

<span class="c1"># Change relative Date columns to days</span>
<span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_datetime</span><span class="p">(</span><span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">days</span>

<span class="c1"># Change the &#39;L1&#39; column to a categorical column</span>
<span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>

<span class="c1"># Plot the data, relative abundance (value) vs. days (date) for each L1</span>


<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;L1rel_MP&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataL1_MP</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Metaproteomic Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\User\AppData\Local\Temp\ipykernel_41148\4211173979.py:5: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MT[&#39;L1&#39;] = dataL1_MT[&#39;L1&#39;].replace(np.nan, &#39;Other&#39;)
C:\Users\User\AppData\Local\Temp\ipykernel_41148\4211173979.py:8: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MT[&#39;date&#39;] = pd.to_datetime(dataL1_MT[&#39;date&#39;])
C:\Users\User\AppData\Local\Temp\ipykernel_41148\4211173979.py:9: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MT[&#39;date&#39;] = (dataL1_MT[&#39;date&#39;] - dataL1_MT[&#39;date&#39;].min()).dt.days
C:\Users\User\AppData\Local\Temp\ipykernel_41148\4211173979.py:12: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MT[&#39;L1&#39;] = dataL1_MT[&#39;L1&#39;].astype(&#39;category&#39;)
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1057: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  grouped_data = data.groupby(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_20_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_20_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># And now pivot the data so each &#39;L1&#39; is a column for each data set</span>

<span class="n">dataL1_MP_pivot</span> <span class="o">=</span> <span class="n">dataL1_MP</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;L1rel_MP&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="n">dataL1_MT_pivot</span> <span class="o">=</span> <span class="n">dataL1_MT</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;L1rel_MT&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now let&#39;s try to infer the var model for both the L1rel_MP or L1rel_MT data sets with the MG or MT data sets</span>
<span class="c1"># So we will have 4 combinations to try</span>
<span class="c1"># 1. L1rel_MP with MG</span>
<span class="c1"># 2. L1rel_MP with MT</span>
<span class="c1"># 3. L1rel_MT with MG</span>
<span class="c1"># 4. L1rel_MT with MT</span>

<span class="n">infer12</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dataMG_pivot</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">dataL1_MP_pivot</span><span class="p">)</span>

<span class="n">infer12</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 06:18<00:00 Sampling 2 chains, 14 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 384 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.058  0.009   0.047    0.068      0.006    0.005       2.0
Ah[0, 1]   -0.006  0.007  -0.013    0.001      0.005    0.004       2.0
Ah[0, 2]    0.085  0.037   0.047    0.124      0.026    0.022       3.0
Ah[0, 3]    0.093  0.015   0.078    0.108      0.010    0.009       3.0
Ah[0, 4]    0.072  0.037   0.035    0.110      0.026    0.022       2.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.250  0.462  -0.221    0.740      0.326    0.276       3.0
Bh[20, 8]   0.024  0.238  -0.220    0.334      0.168    0.142       3.0
Bh[20, 9]  -0.510  0.167  -0.695   -0.330      0.118    0.100       3.0
Bh[20, 10] -0.174  0.277  -0.471    0.125      0.195    0.165       3.0
Bh[20, 11]  0.347  0.479  -0.136    0.830      0.338    0.286       2.0

            ess_tail  r_hat
Ah[0, 0]        23.0   2.48
Ah[0, 1]        11.0   2.61
Ah[0, 2]        11.0   2.26
Ah[0, 3]        43.0   2.16
Ah[0, 4]        13.0   2.64
...              ...    ...
Bh[20, 7]       11.0   2.09
Bh[20, 8]       20.0   2.26
Bh[20, 9]       15.0   2.24
Bh[20, 10]      14.0   1.99
Bh[20, 11]      11.0   3.02

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_xs_v2.nc
Data file: data_xs_v2.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_22_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_22_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer12</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_23_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_23_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_23_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_23_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer13</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dataMT_pivot</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">dataL1_MP_pivot</span><span class="p">)</span>

<span class="n">infer13</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 06:10<00:00 Sampling 2 chains, 274 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 375 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.087  0.022   0.065    0.111      0.016    0.013       3.0
Ah[0, 1]    0.029  0.008   0.021    0.037      0.006    0.005       3.0
Ah[0, 2]   -0.214  0.283  -0.497    0.071      0.200    0.169       3.0
Ah[0, 3]    0.085  0.071   0.014    0.156      0.050    0.042       2.0
Ah[0, 4]    0.265  0.121   0.143    0.386      0.085    0.072       3.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 6]   0.395  0.060   0.323    0.457      0.042    0.035       3.0
Bh[20, 7]   0.175  0.191  -0.019    0.378      0.135    0.114       3.0
Bh[20, 8]   0.055  0.071  -0.019    0.140      0.050    0.042       2.0
Bh[20, 9]   0.371  0.092   0.275    0.475      0.065    0.055       3.0
Bh[20, 10] -0.067  0.480  -0.560    0.413      0.339    0.287       3.0

            ess_tail  r_hat
Ah[0, 0]        15.0   2.24
Ah[0, 1]        11.0   2.14
Ah[0, 2]        15.0   2.07
Ah[0, 3]        13.0   2.67
Ah[0, 4]        11.0   1.96
...              ...    ...
Bh[20, 6]       11.0   2.01
Bh[20, 7]       25.0   2.01
Bh[20, 8]       22.0   2.91
Bh[20, 9]       29.0   1.99
Bh[20, 10]      24.0   2.26

[352 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (352) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_xs_v3.nc
Data file: data_xs_v3.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_24_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_24_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer13</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_25_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_25_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_25_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_25_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer14</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dataMG_pivot</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">dataL1_MT_pivot</span><span class="p">)</span>

<span class="n">infer14</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 06:15<00:00 Sampling 2 chains, 254 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 379 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.377  0.171   0.205    0.548      0.120    0.102       3.0
Ah[0, 1]   -0.008  0.008  -0.017    0.000      0.006    0.005       3.0
Ah[0, 2]    0.296  0.063   0.233    0.360      0.044    0.037       2.0
Ah[0, 3]    0.013  0.026  -0.013    0.041      0.019    0.016       3.0
Ah[0, 4]    0.054  0.015   0.039    0.071      0.011    0.009       2.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]  -0.140  0.338  -0.491    0.199      0.238    0.202       2.0
Bh[20, 8]  -0.158  0.151  -0.309    0.000      0.106    0.090       2.0
Bh[20, 9]   0.487  0.040   0.401    0.559      0.020    0.016       4.0
Bh[20, 10] -0.233  0.173  -0.462   -0.067      0.121    0.102       3.0
Bh[20, 11] -0.114  0.207  -0.322    0.105      0.146    0.123       2.0

            ess_tail  r_hat
Ah[0, 0]        30.0   2.17
Ah[0, 1]        21.0   2.15
Ah[0, 2]        12.0   2.84
Ah[0, 3]        22.0   2.23
Ah[0, 4]        11.0   3.00
...              ...    ...
Bh[20, 7]       11.0   2.37
Bh[20, 8]       14.0   2.57
Bh[20, 9]       11.0   1.92
Bh[20, 10]      35.0   2.27
Bh[20, 11]      12.0   2.67

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_xs_v4.nc
Data file: data_xs_v4.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_26_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_26_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer14</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_27_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_27_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_27_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_27_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer15</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dataMT_pivot</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">dataL1_MT_pivot</span><span class="p">)</span>

<span class="n">infer15</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 06:00<00:00 Sampling 2 chains, 106 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 365 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.133  0.146  -0.013    0.280      0.103    0.087       3.0
Ah[0, 1]    0.013  0.014  -0.001    0.027      0.010    0.008       3.0
Ah[0, 2]   -0.102  0.481  -0.584    0.379      0.339    0.287       2.0
Ah[0, 3]    0.089  0.065   0.024    0.155      0.046    0.039       2.0
Ah[0, 4]    0.192  0.101   0.090    0.292      0.071    0.060       3.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 6]   0.274  0.209   0.061    0.503      0.147    0.125       2.0
Bh[20, 7]  -0.202  0.145  -0.357   -0.041      0.102    0.087       3.0
Bh[20, 8]  -0.231  0.311  -0.544    0.081      0.219    0.186       3.0
Bh[20, 9]   0.606  0.165   0.435    0.781      0.116    0.098       3.0
Bh[20, 10] -0.447  0.572  -1.025    0.128      0.404    0.342       3.0

            ess_tail  r_hat
Ah[0, 0]        18.0   2.25
Ah[0, 1]        12.0   2.23
Ah[0, 2]        11.0   2.88
Ah[0, 3]        11.0   2.73
Ah[0, 4]        15.0   2.13
...              ...    ...
Bh[20, 6]       11.0   2.64
Bh[20, 7]       11.0   2.26
Bh[20, 8]       11.0   2.00
Bh[20, 9]       13.0   2.17
Bh[20, 10]      38.0   1.84

[352 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (352) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_xs_v5.nc
Data file: data_xs_v5.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_28_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_28_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer15</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_29_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_29_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_29_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_29_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dataL1_MP_pivot</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>L1</th>
      <th>01_Fermentation</th>
      <th>02_Homoacetogenesis</th>
      <th>03_Superpathway of thiosulfate metabolism (Desulfovibrio sulfodismutans)</th>
      <th>04_Utililization of sugar, conversion of pentose to EMP pathway intermediates</th>
      <th>05_Fatty acid oxidation</th>
      <th>06_Amino acid utilization biosynthesis metabolism</th>
      <th>07_Nucleic acid metabolism</th>
      <th>08_Hydrocarbon degradation</th>
      <th>09_Carbohydrate Active enzyme - CAZy</th>
      <th>10_TCA cycle</th>
      <th>...</th>
      <th>12_Transporters</th>
      <th>13_Hydrogen metabolism</th>
      <th>14_Methanogenesis</th>
      <th>15_Methylotrophy</th>
      <th>16_Embden Meyerhof-Parnos (EMP)</th>
      <th>17_Gluconeogenesis</th>
      <th>18_Sulfur compounds metabolism</th>
      <th>19_Saccharide and derivated synthesis</th>
      <th>20_Hydrolysis of polymers</th>
      <th>21_Cellular response to stress</th>
    </tr>
    <tr>
      <th>date</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.096899</td>
      <td>0.009890</td>
      <td>0.000579</td>
      <td>0.000535</td>
      <td>0.025483</td>
      <td>0.122784</td>
      <td>0.119531</td>
      <td>0.005524</td>
      <td>0.003921</td>
      <td>0.045398</td>
      <td>...</td>
      <td>0.015593</td>
      <td>0.000846</td>
      <td>0.003520</td>
      <td>0.009267</td>
      <td>0.030473</td>
      <td>0.016261</td>
      <td>0.005346</td>
      <td>0.009267</td>
      <td>0.006193</td>
      <td>0.037334</td>
    </tr>
    <tr>
      <th>8</th>
      <td>0.133734</td>
      <td>0.011645</td>
      <td>0.000340</td>
      <td>0.000870</td>
      <td>0.033575</td>
      <td>0.129235</td>
      <td>0.108779</td>
      <td>0.004688</td>
      <td>0.003063</td>
      <td>0.052367</td>
      <td>...</td>
      <td>0.018224</td>
      <td>0.001059</td>
      <td>0.004159</td>
      <td>0.016334</td>
      <td>0.035428</td>
      <td>0.018678</td>
      <td>0.003592</td>
      <td>0.009150</td>
      <td>0.006692</td>
      <td>0.042688</td>
    </tr>
    <tr>
      <th>15</th>
      <td>0.105901</td>
      <td>0.009051</td>
      <td>0.000754</td>
      <td>0.000266</td>
      <td>0.029503</td>
      <td>0.130568</td>
      <td>0.117214</td>
      <td>0.004570</td>
      <td>0.003061</td>
      <td>0.048048</td>
      <td>...</td>
      <td>0.018500</td>
      <td>0.001109</td>
      <td>0.005368</td>
      <td>0.011180</td>
      <td>0.024933</td>
      <td>0.013886</td>
      <td>0.007143</td>
      <td>0.005501</td>
      <td>0.005102</td>
      <td>0.029148</td>
    </tr>
    <tr>
      <th>24</th>
      <td>0.111451</td>
      <td>0.010700</td>
      <td>0.000262</td>
      <td>0.000349</td>
      <td>0.029435</td>
      <td>0.121233</td>
      <td>0.108612</td>
      <td>0.005634</td>
      <td>0.004061</td>
      <td>0.045725</td>
      <td>...</td>
      <td>0.015591</td>
      <td>0.000699</td>
      <td>0.003319</td>
      <td>0.012184</td>
      <td>0.029697</td>
      <td>0.014717</td>
      <td>0.002751</td>
      <td>0.007730</td>
      <td>0.008036</td>
      <td>0.037296</td>
    </tr>
    <tr>
      <th>31</th>
      <td>0.101236</td>
      <td>0.008762</td>
      <td>0.000313</td>
      <td>0.000235</td>
      <td>0.033602</td>
      <td>0.130300</td>
      <td>0.103701</td>
      <td>0.004772</td>
      <td>0.005359</td>
      <td>0.043029</td>
      <td>...</td>
      <td>0.014865</td>
      <td>0.000352</td>
      <td>0.003286</td>
      <td>0.012752</td>
      <td>0.034736</td>
      <td>0.018346</td>
      <td>0.003286</td>
      <td>0.007393</td>
      <td>0.004381</td>
      <td>0.033876</td>
    </tr>
    <tr>
      <th>39</th>
      <td>0.112423</td>
      <td>0.010556</td>
      <td>0.000052</td>
      <td>0.000416</td>
      <td>0.029692</td>
      <td>0.116999</td>
      <td>0.110603</td>
      <td>0.005616</td>
      <td>0.003016</td>
      <td>0.047163</td>
      <td>...</td>
      <td>0.015444</td>
      <td>0.000468</td>
      <td>0.003380</td>
      <td>0.011232</td>
      <td>0.032708</td>
      <td>0.016068</td>
      <td>0.001040</td>
      <td>0.008320</td>
      <td>0.006344</td>
      <td>0.036244</td>
    </tr>
    <tr>
      <th>46</th>
      <td>0.097988</td>
      <td>0.007198</td>
      <td>0.000315</td>
      <td>0.000368</td>
      <td>0.028950</td>
      <td>0.120265</td>
      <td>0.116377</td>
      <td>0.003468</td>
      <td>0.003625</td>
      <td>0.041559</td>
      <td>...</td>
      <td>0.011769</td>
      <td>0.000630</td>
      <td>0.004518</td>
      <td>0.009457</td>
      <td>0.022908</td>
      <td>0.013660</td>
      <td>0.001524</td>
      <td>0.008459</td>
      <td>0.003888</td>
      <td>0.037146</td>
    </tr>
    <tr>
      <th>53</th>
      <td>0.083523</td>
      <td>0.009196</td>
      <td>0.000000</td>
      <td>0.000928</td>
      <td>0.027335</td>
      <td>0.157597</td>
      <td>0.097022</td>
      <td>0.007171</td>
      <td>0.002700</td>
      <td>0.038556</td>
      <td>...</td>
      <td>0.015186</td>
      <td>0.000422</td>
      <td>0.006918</td>
      <td>0.013752</td>
      <td>0.031722</td>
      <td>0.016030</td>
      <td>0.001687</td>
      <td>0.004218</td>
      <td>0.007002</td>
      <td>0.028516</td>
    </tr>
    <tr>
      <th>60</th>
      <td>0.100174</td>
      <td>0.009553</td>
      <td>0.001742</td>
      <td>0.000106</td>
      <td>0.022220</td>
      <td>0.131472</td>
      <td>0.139547</td>
      <td>0.006439</td>
      <td>0.002006</td>
      <td>0.043543</td>
      <td>...</td>
      <td>0.016995</td>
      <td>0.000633</td>
      <td>0.003431</td>
      <td>0.008445</td>
      <td>0.027128</td>
      <td>0.014039</td>
      <td>0.014778</td>
      <td>0.006756</td>
      <td>0.005753</td>
      <td>0.034253</td>
    </tr>
    <tr>
      <th>67</th>
      <td>0.113514</td>
      <td>0.010484</td>
      <td>0.000000</td>
      <td>0.001208</td>
      <td>0.031672</td>
      <td>0.112416</td>
      <td>0.111648</td>
      <td>0.004666</td>
      <td>0.003239</td>
      <td>0.050390</td>
      <td>...</td>
      <td>0.011692</td>
      <td>0.000659</td>
      <td>0.006312</td>
      <td>0.012954</td>
      <td>0.029202</td>
      <td>0.015095</td>
      <td>0.003458</td>
      <td>0.006532</td>
      <td>0.006642</td>
      <td>0.038424</td>
    </tr>
    <tr>
      <th>74</th>
      <td>0.113738</td>
      <td>0.012006</td>
      <td>0.000263</td>
      <td>0.001106</td>
      <td>0.033174</td>
      <td>0.129061</td>
      <td>0.101469</td>
      <td>0.003844</td>
      <td>0.004739</td>
      <td>0.043863</td>
      <td>...</td>
      <td>0.016797</td>
      <td>0.000737</td>
      <td>0.006108</td>
      <td>0.016745</td>
      <td>0.039018</td>
      <td>0.019114</td>
      <td>0.002949</td>
      <td>0.007583</td>
      <td>0.007741</td>
      <td>0.036386</td>
    </tr>
    <tr>
      <th>80</th>
      <td>0.117201</td>
      <td>0.010857</td>
      <td>0.000372</td>
      <td>0.001004</td>
      <td>0.041310</td>
      <td>0.113259</td>
      <td>0.112776</td>
      <td>0.003830</td>
      <td>0.003793</td>
      <td>0.050978</td>
      <td>...</td>
      <td>0.011043</td>
      <td>0.000669</td>
      <td>0.003979</td>
      <td>0.010634</td>
      <td>0.031903</td>
      <td>0.018108</td>
      <td>0.006284</td>
      <td>0.008143</td>
      <td>0.006284</td>
      <td>0.038299</td>
    </tr>
    <tr>
      <th>88</th>
      <td>0.107418</td>
      <td>0.010560</td>
      <td>0.000173</td>
      <td>0.000476</td>
      <td>0.032762</td>
      <td>0.131308</td>
      <td>0.115684</td>
      <td>0.003765</td>
      <td>0.003549</td>
      <td>0.058643</td>
      <td>...</td>
      <td>0.010863</td>
      <td>0.000736</td>
      <td>0.005193</td>
      <td>0.011729</td>
      <td>0.037869</td>
      <td>0.021380</td>
      <td>0.001904</td>
      <td>0.008786</td>
      <td>0.004804</td>
      <td>0.038821</td>
    </tr>
    <tr>
      <th>95</th>
      <td>0.137137</td>
      <td>0.012536</td>
      <td>0.000040</td>
      <td>0.001477</td>
      <td>0.029903</td>
      <td>0.138494</td>
      <td>0.113582</td>
      <td>0.004272</td>
      <td>0.002435</td>
      <td>0.044075</td>
      <td>...</td>
      <td>0.016408</td>
      <td>0.000958</td>
      <td>0.004352</td>
      <td>0.012256</td>
      <td>0.026230</td>
      <td>0.018005</td>
      <td>0.002595</td>
      <td>0.006627</td>
      <td>0.003912</td>
      <td>0.037248</td>
    </tr>
    <tr>
      <th>102</th>
      <td>0.135713</td>
      <td>0.009189</td>
      <td>0.000169</td>
      <td>0.001778</td>
      <td>0.037305</td>
      <td>0.105522</td>
      <td>0.077363</td>
      <td>0.002752</td>
      <td>0.004319</td>
      <td>0.058731</td>
      <td>...</td>
      <td>0.014482</td>
      <td>0.000635</td>
      <td>0.003345</td>
      <td>0.008088</td>
      <td>0.025152</td>
      <td>0.019732</td>
      <td>0.000932</td>
      <td>0.006309</td>
      <td>0.005843</td>
      <td>0.041878</td>
    </tr>
  </tbody>
</table>
<p>15 rows × 21 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">dataL1_MP_pivot</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Index([  0,   8,  15,  24,  31,  39,  46,  53,  60,  67,  74,  80,  88,  95,
       102, 109, 137, 143, 151, 161, 168, 175, 182, 191, 198, 205, 226, 231,
       240, 247, 253, 275, 282, 288, 296, 304, 310, 317, 324, 330, 339, 345,
       353, 359, 367, 373, 380, 386, 393, 401, 409],
      dtype=&#39;int64&#39;, name=&#39;date&#39;)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Step 1: Determine the minimum and maximum index</span>
<span class="n">min_index</span> <span class="o">=</span> <span class="n">dataL1_MP_pivot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<span class="n">max_index</span> <span class="o">=</span> <span class="n">dataL1_MP_pivot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>

<span class="c1"># Step 2: Create a new index range from min_index to max_index</span>
<span class="n">new_index</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">min_index</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">max_index</span><span class="p">)</span> <span class="o">+</span>
                  <span class="mi">1</span><span class="p">)</span>  <span class="c1"># Ensure it&#39;s an integer range</span>

<span class="c1"># Step 3: Re-index the DataFrame</span>
<span class="n">dataL1_MP_complete</span> <span class="o">=</span> <span class="n">dataL1_MP_pivot</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">new_index</span><span class="p">)</span>

<span class="n">dataL1_MP_complete</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>L1</th>
      <th>01_Fermentation</th>
      <th>02_Homoacetogenesis</th>
      <th>03_Superpathway of thiosulfate metabolism (Desulfovibrio sulfodismutans)</th>
      <th>04_Utililization of sugar, conversion of pentose to EMP pathway intermediates</th>
      <th>05_Fatty acid oxidation</th>
      <th>06_Amino acid utilization biosynthesis metabolism</th>
      <th>07_Nucleic acid metabolism</th>
      <th>08_Hydrocarbon degradation</th>
      <th>09_Carbohydrate Active enzyme - CAZy</th>
      <th>10_TCA cycle</th>
      <th>...</th>
      <th>12_Transporters</th>
      <th>13_Hydrogen metabolism</th>
      <th>14_Methanogenesis</th>
      <th>15_Methylotrophy</th>
      <th>16_Embden Meyerhof-Parnos (EMP)</th>
      <th>17_Gluconeogenesis</th>
      <th>18_Sulfur compounds metabolism</th>
      <th>19_Saccharide and derivated synthesis</th>
      <th>20_Hydrolysis of polymers</th>
      <th>21_Cellular response to stress</th>
    </tr>
    <tr>
      <th>date</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.096899</td>
      <td>0.009890</td>
      <td>0.000579</td>
      <td>0.000535</td>
      <td>0.025483</td>
      <td>0.122784</td>
      <td>0.119531</td>
      <td>0.005524</td>
      <td>0.003921</td>
      <td>0.045398</td>
      <td>...</td>
      <td>0.015593</td>
      <td>0.000846</td>
      <td>0.003520</td>
      <td>0.009267</td>
      <td>0.030473</td>
      <td>0.016261</td>
      <td>0.005346</td>
      <td>0.009267</td>
      <td>0.006193</td>
      <td>0.037334</td>
    </tr>
    <tr>
      <th>1</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>5</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>6</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>7</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>8</th>
      <td>0.133734</td>
      <td>0.011645</td>
      <td>0.000340</td>
      <td>0.000870</td>
      <td>0.033575</td>
      <td>0.129235</td>
      <td>0.108779</td>
      <td>0.004688</td>
      <td>0.003063</td>
      <td>0.052367</td>
      <td>...</td>
      <td>0.018224</td>
      <td>0.001059</td>
      <td>0.004159</td>
      <td>0.016334</td>
      <td>0.035428</td>
      <td>0.018678</td>
      <td>0.003592</td>
      <td>0.009150</td>
      <td>0.006692</td>
      <td>0.042688</td>
    </tr>
    <tr>
      <th>9</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>10</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>11</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>12</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>13</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>14</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>15 rows × 21 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reset the index</span>
<span class="n">complete_dataMP_reset</span> <span class="o">=</span> <span class="n">dataL1_MP_complete</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

<span class="c1"># Make index as float</span>
<span class="n">complete_dataMP_reset</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">complete_dataMP_reset</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">complete_dataMP_reset</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>L1</th>
      <th>date</th>
      <th>01_Fermentation</th>
      <th>02_Homoacetogenesis</th>
      <th>03_Superpathway of thiosulfate metabolism (Desulfovibrio sulfodismutans)</th>
      <th>04_Utililization of sugar, conversion of pentose to EMP pathway intermediates</th>
      <th>05_Fatty acid oxidation</th>
      <th>06_Amino acid utilization biosynthesis metabolism</th>
      <th>07_Nucleic acid metabolism</th>
      <th>08_Hydrocarbon degradation</th>
      <th>09_Carbohydrate Active enzyme - CAZy</th>
      <th>...</th>
      <th>12_Transporters</th>
      <th>13_Hydrogen metabolism</th>
      <th>14_Methanogenesis</th>
      <th>15_Methylotrophy</th>
      <th>16_Embden Meyerhof-Parnos (EMP)</th>
      <th>17_Gluconeogenesis</th>
      <th>18_Sulfur compounds metabolism</th>
      <th>19_Saccharide and derivated synthesis</th>
      <th>20_Hydrolysis of polymers</th>
      <th>21_Cellular response to stress</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.0</td>
      <td>0.096899</td>
      <td>0.009890</td>
      <td>0.000579</td>
      <td>0.000535</td>
      <td>0.025483</td>
      <td>0.122784</td>
      <td>0.119531</td>
      <td>0.005524</td>
      <td>0.003921</td>
      <td>...</td>
      <td>0.015593</td>
      <td>0.000846</td>
      <td>0.003520</td>
      <td>0.009267</td>
      <td>0.030473</td>
      <td>0.016261</td>
      <td>0.005346</td>
      <td>0.009267</td>
      <td>0.006193</td>
      <td>0.037334</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>5</th>
      <td>5.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>6</th>
      <td>6.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>7</th>
      <td>7.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>8</th>
      <td>8.0</td>
      <td>0.133734</td>
      <td>0.011645</td>
      <td>0.000340</td>
      <td>0.000870</td>
      <td>0.033575</td>
      <td>0.129235</td>
      <td>0.108779</td>
      <td>0.004688</td>
      <td>0.003063</td>
      <td>...</td>
      <td>0.018224</td>
      <td>0.001059</td>
      <td>0.004159</td>
      <td>0.016334</td>
      <td>0.035428</td>
      <td>0.018678</td>
      <td>0.003592</td>
      <td>0.009150</td>
      <td>0.006692</td>
      <td>0.042688</td>
    </tr>
    <tr>
      <th>9</th>
      <td>9.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>10</th>
      <td>10.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>11</th>
      <td>11.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>12</th>
      <td>12.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>13</th>
      <td>13.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>14</th>
      <td>14.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>15 rows × 22 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># do the imputation for each column (metabolite) in the dataset</span>
<span class="c1"># Copy the original dataset</span>
<span class="n">new_dataset_complete</span> <span class="o">=</span> <span class="n">complete_dataMP_reset</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="k">for</span> <span class="n">metabolite</span> <span class="ow">in</span> <span class="n">dataL1_MP_complete</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="c1"># Impute the missing values for the current column</span>
    <span class="n">imputed_data</span> <span class="o">=</span> <span class="n">GPImputer</span><span class="p">()</span><span class="o">.</span><span class="n">impute_missing_values</span><span class="p">(</span>
        <span class="n">complete_dataMP_reset</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">metabolite</span><span class="p">],</span> <span class="n">metabolite</span>
    <span class="p">)</span>

    <span class="c1"># Update the column in the new dataset with the imputed data</span>
    <span class="n">new_dataset_complete</span><span class="p">[</span><span class="n">metabolite</span><span class="p">]</span> <span class="o">=</span> <span class="n">imputed_data</span><span class="p">[</span><span class="n">metabolite</span><span class="p">]</span>

<span class="c1"># Save the dataset (new_dataset_complete) to a csv file</span>
<span class="n">new_dataset_complete</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;MP_metabolite_imputed_data.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 316.24954273578993
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 543.8966103725907
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_3.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_3.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Cosine&#39;&gt;
BIC: 607.9660604736986
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_5.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_5.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 608.9563540235508
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_7.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_7.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 401.7797626907321
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_9.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_9.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 325.3865873305258
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_11.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_11.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 296.9544274862874
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_13.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_13.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 564.5813046134983
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_15.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_15.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 585.5469855571225
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_17.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_17.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 388.40705634294005
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_19.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_19.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 458.8175786976297
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_21.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_21.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 480.62151656395025
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_23.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_23.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Cosine&#39;&gt;
BIC: 614.8411989518773
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_25.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_25.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 531.0659831191797
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_27.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_27.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 494.17518442169035
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_29.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_29.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 408.56692970134554
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_31.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_31.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 493.5793389680505
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_33.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_33.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 449.3139425867986
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_35.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_35.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 475.65821053719867
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_37.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_37.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 531.9685673673611
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_39.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_39.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 410.2962130321013
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_41.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_35_41.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Let&#39;s do the same for the MT dataset</span>

<span class="c1"># Step 1: Determine the minimum and maximum index</span>
<span class="n">min_index</span> <span class="o">=</span> <span class="n">dataL1_MT_pivot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<span class="n">max_index</span> <span class="o">=</span> <span class="n">dataL1_MT_pivot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>

<span class="c1"># Step 2: Create a new index range from min_index to max_index</span>
<span class="n">new_index</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">min_index</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">max_index</span><span class="p">)</span> <span class="o">+</span>
                  <span class="mi">1</span><span class="p">)</span>  <span class="c1"># Ensure it&#39;s an integer range</span>

<span class="c1"># Step 3: Re-index the DataFrame</span>
<span class="n">dataL1_MT_complete</span> <span class="o">=</span> <span class="n">dataL1_MT_pivot</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">new_index</span><span class="p">)</span>

<span class="n">dataL1_MT_complete</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>L1</th>
      <th>01_Fermentation</th>
      <th>02_Homoacetogenesis</th>
      <th>03_Superpathway of thiosulfate metabolism (Desulfovibrio sulfodismutans)</th>
      <th>04_Utililization of sugar, conversion of pentose to EMP pathway intermediates</th>
      <th>05_Fatty acid oxidation</th>
      <th>06_Amino acid utilization biosynthesis metabolism</th>
      <th>07_Nucleic acid metabolism</th>
      <th>08_Hydrocarbon degradation</th>
      <th>09_Carbohydrate Active enzyme - CAZy</th>
      <th>10_TCA cycle</th>
      <th>...</th>
      <th>12_Transporters</th>
      <th>13_Hydrogen metabolism</th>
      <th>14_Methanogenesis</th>
      <th>15_Methylotrophy</th>
      <th>16_Embden Meyerhof-Parnos (EMP)</th>
      <th>17_Gluconeogenesis</th>
      <th>18_Sulfur compounds metabolism</th>
      <th>19_Saccharide and derivated synthesis</th>
      <th>20_Hydrolysis of polymers</th>
      <th>21_Cellular response to stress</th>
    </tr>
    <tr>
      <th>date</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.032555</td>
      <td>0.004332</td>
      <td>0.000655</td>
      <td>0.000551</td>
      <td>0.007245</td>
      <td>0.050208</td>
      <td>0.018190</td>
      <td>0.001434</td>
      <td>0.007091</td>
      <td>0.007711</td>
      <td>...</td>
      <td>0.013046</td>
      <td>0.000764</td>
      <td>0.001359</td>
      <td>0.003455</td>
      <td>0.009246</td>
      <td>0.004036</td>
      <td>0.002204</td>
      <td>0.015901</td>
      <td>0.007996</td>
      <td>0.044404</td>
    </tr>
    <tr>
      <th>1</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>5</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>6</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>7</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>8</th>
      <td>0.032122</td>
      <td>0.004362</td>
      <td>0.000836</td>
      <td>0.000600</td>
      <td>0.007049</td>
      <td>0.045215</td>
      <td>0.019025</td>
      <td>0.001449</td>
      <td>0.008429</td>
      <td>0.007428</td>
      <td>...</td>
      <td>0.011506</td>
      <td>0.000694</td>
      <td>0.001097</td>
      <td>0.003125</td>
      <td>0.009361</td>
      <td>0.003913</td>
      <td>0.002097</td>
      <td>0.015956</td>
      <td>0.008593</td>
      <td>0.045070</td>
    </tr>
    <tr>
      <th>9</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>10</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>11</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>12</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>13</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>14</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>15 rows × 21 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reset the index</span>
<span class="n">complete_dataMT_reset</span> <span class="o">=</span> <span class="n">dataL1_MT_complete</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

<span class="c1"># Make index as float</span>
<span class="n">complete_dataMT_reset</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">complete_dataMT_reset</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># do the imputation for each column (metabolite) in the dataset</span>
<span class="c1"># Copy the original dataset</span>
<span class="n">new_dataset_complete</span> <span class="o">=</span> <span class="n">complete_dataMT_reset</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="k">for</span> <span class="n">metabolite</span> <span class="ow">in</span> <span class="n">dataL1_MT_complete</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="c1"># Impute the missing values for the current column</span>
    <span class="n">imputed_data</span> <span class="o">=</span> <span class="n">GPImputer</span><span class="p">()</span><span class="o">.</span><span class="n">impute_missing_values</span><span class="p">(</span>
        <span class="n">complete_dataMT_reset</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">metabolite</span><span class="p">],</span> <span class="n">metabolite</span>
    <span class="p">)</span>

    <span class="c1"># Update the column in the new dataset with the imputed data</span>
    <span class="n">new_dataset_complete</span><span class="p">[</span><span class="n">metabolite</span><span class="p">]</span> <span class="o">=</span> <span class="n">imputed_data</span><span class="p">[</span><span class="n">metabolite</span><span class="p">]</span>

<span class="c1"># Save the dataset (new_dataset_complete) to a csv file</span>
<span class="n">new_dataset_complete</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;MT_metabolite_imputed_data.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern12&#39;&gt;
BIC: 478.859730026591
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 614.2657044827979
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_3.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_3.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 608.8043346603398
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_5.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_5.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Cosine&#39;&gt;
BIC: 617.8351979733203
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_7.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_7.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 573.9485590748709
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_9.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_9.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 435.39759108490625
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_11.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_11.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 514.0421771707646
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_13.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_13.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 618.2696916720081
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_15.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_15.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 522.3910029832333
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_17.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_17.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 581.3090177713289
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_19.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_19.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 527.9448745709293
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_21.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_21.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 528.552383605427
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_23.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_23.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 607.2700168991987
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_25.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_25.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 612.1185334241881
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_27.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_27.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 612.844341963594
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_29.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_29.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 591.550679192629
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_31.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_31.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 614.472046581256
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_33.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_33.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 615.2226785686273
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_35.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_35.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 503.610308665269
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_37.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_37.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 586.0012210101237
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_39.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_39.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.RationalQuadratic&#39;&gt;
BIC: 473.07506092202203
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_41.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_38_41.png" />
</div>
</div>
<p>So now, let’s try doing the inference with the VAR method and the imputed data to see if it can infer the causal relationships between the variables in the time series.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reload the imputed data into a new DataFrame</span>

<span class="n">imputed_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;imputed_data.csv&#39;</span><span class="p">)</span>

<span class="c1"># remove the &#39;days&#39; column</span>
<span class="n">imputed_data</span> <span class="o">=</span> <span class="n">imputed_data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;days&#39;</span><span class="p">])</span>
<span class="n">imputed_data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Acidimicrobium</th>
      <th>Acinetobacter</th>
      <th>Albidiferax</th>
      <th>Candidatus Microthrix</th>
      <th>Chitinophaga pinensis</th>
      <th>Dechloromonas</th>
      <th>Haliscomenobacter</th>
      <th>Intrasporangium</th>
      <th>Leptospira</th>
      <th>Other</th>
      <th>Xanthomonas</th>
      <th>mean abundance &lt; 2%</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>8.322484</td>
      <td>0.791870</td>
      <td>6.383491</td>
      <td>26.180838</td>
      <td>7.055585</td>
      <td>4.998287</td>
      <td>15.142620</td>
      <td>11.445072</td>
      <td>4.742982</td>
      <td>2.357318</td>
      <td>0.853658</td>
      <td>11.725795</td>
    </tr>
    <tr>
      <th>1</th>
      <td>8.265177</td>
      <td>0.763703</td>
      <td>6.362575</td>
      <td>25.593890</td>
      <td>7.081694</td>
      <td>5.072711</td>
      <td>15.397948</td>
      <td>11.546582</td>
      <td>4.791269</td>
      <td>2.382206</td>
      <td>0.864697</td>
      <td>11.836536</td>
    </tr>
    <tr>
      <th>2</th>
      <td>8.180701</td>
      <td>0.781881</td>
      <td>6.330330</td>
      <td>25.226861</td>
      <td>7.127157</td>
      <td>5.147886</td>
      <td>15.663537</td>
      <td>11.636934</td>
      <td>4.833619</td>
      <td>2.399779</td>
      <td>0.879132</td>
      <td>11.947293</td>
    </tr>
    <tr>
      <th>3</th>
      <td>8.078737</td>
      <td>0.801890</td>
      <td>6.290243</td>
      <td>24.849371</td>
      <td>7.185821</td>
      <td>5.219273</td>
      <td>15.926969</td>
      <td>11.710476</td>
      <td>4.868031</td>
      <td>2.418784</td>
      <td>0.894873</td>
      <td>12.058067</td>
    </tr>
    <tr>
      <th>4</th>
      <td>7.969807</td>
      <td>0.823758</td>
      <td>6.246734</td>
      <td>24.505012</td>
      <td>7.249498</td>
      <td>5.280679</td>
      <td>16.172482</td>
      <td>11.759176</td>
      <td>4.892380</td>
      <td>2.438263</td>
      <td>0.909303</td>
      <td>12.168858</td>
    </tr>
    <tr>
      <th>5</th>
      <td>7.863614</td>
      <td>0.847488</td>
      <td>6.204391</td>
      <td>24.245813</td>
      <td>7.309111</td>
      <td>5.324508</td>
      <td>16.382314</td>
      <td>11.772278</td>
      <td>4.904550</td>
      <td>2.457029</td>
      <td>0.919595</td>
      <td>12.279667</td>
    </tr>
    <tr>
      <th>6</th>
      <td>7.770084</td>
      <td>0.873055</td>
      <td>6.167656</td>
      <td>24.127242</td>
      <td>7.355842</td>
      <td>5.342460</td>
      <td>16.538718</td>
      <td>11.737040</td>
      <td>4.902619</td>
      <td>2.473734</td>
      <td>0.923138</td>
      <td>12.390494</td>
    </tr>
    <tr>
      <th>7</th>
      <td>7.700314</td>
      <td>0.900405</td>
      <td>6.140528</td>
      <td>24.199845</td>
      <td>7.382293</td>
      <td>5.326927</td>
      <td>16.626579</td>
      <td>11.641157</td>
      <td>4.885061</td>
      <td>2.487002</td>
      <td>0.918012</td>
      <td>12.501338</td>
    </tr>
    <tr>
      <th>8</th>
      <td>7.664534</td>
      <td>0.911291</td>
      <td>6.125822</td>
      <td>23.641355</td>
      <td>7.383681</td>
      <td>5.273044</td>
      <td>16.636265</td>
      <td>11.476739</td>
      <td>4.850959</td>
      <td>2.520630</td>
      <td>0.903478</td>
      <td>12.612201</td>
    </tr>
    <tr>
      <th>9</th>
      <td>7.666801</td>
      <td>0.960073</td>
      <td>6.124129</td>
      <td>25.024940</td>
      <td>7.358852</td>
      <td>5.180566</td>
      <td>16.565955</td>
      <td>11.244474</td>
      <td>4.800196</td>
      <td>2.499159</td>
      <td>0.880339</td>
      <td>12.321359</td>
    </tr>
    <tr>
      <th>10</th>
      <td>7.701597</td>
      <td>0.992116</td>
      <td>6.133189</td>
      <td>25.753165</td>
      <td>7.310707</td>
      <td>5.055128</td>
      <td>16.422573</td>
      <td>10.955857</td>
      <td>4.733611</td>
      <td>2.497398</td>
      <td>0.851104</td>
      <td>12.030537</td>
    </tr>
    <tr>
      <th>11</th>
      <td>7.756540</td>
      <td>1.025393</td>
      <td>6.148177</td>
      <td>26.621835</td>
      <td>7.245802</td>
      <td>4.907436</td>
      <td>16.220878</td>
      <td>10.632238</td>
      <td>4.653122</td>
      <td>2.490776</td>
      <td>0.819790</td>
      <td>11.739731</td>
    </tr>
    <tr>
      <th>12</th>
      <td>7.816900</td>
      <td>1.059680</td>
      <td>6.162504</td>
      <td>27.550701</td>
      <td>7.173270</td>
      <td>4.751605</td>
      <td>15.981051</td>
      <td>10.301403</td>
      <td>4.561808</td>
      <td>2.479674</td>
      <td>0.791448</td>
      <td>11.448942</td>
    </tr>
    <tr>
      <th>13</th>
      <td>7.867041</td>
      <td>1.094721</td>
      <td>6.168573</td>
      <td>28.452806</td>
      <td>7.103389</td>
      <td>4.602845</td>
      <td>15.725567</td>
      <td>9.993155</td>
      <td>4.463977</td>
      <td>2.464200</td>
      <td>0.771427</td>
      <td>11.158170</td>
    </tr>
    <tr>
      <th>14</th>
      <td>7.890203</td>
      <td>1.130229</td>
      <td>6.158709</td>
      <td>29.247651</td>
      <td>7.046066</td>
      <td>4.475146</td>
      <td>15.476159</td>
      <td>9.734647</td>
      <td>4.365213</td>
      <td>2.444171</td>
      <td>0.764505</td>
      <td>10.867413</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">imputed_MP_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;MP_metabolite_imputed_data.csv&#39;</span><span class="p">)</span>
<span class="n">imputed_MP_data</span> <span class="o">=</span> <span class="n">imputed_MP_data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">imputed_MP_data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>01_Fermentation</th>
      <th>02_Homoacetogenesis</th>
      <th>03_Superpathway of thiosulfate metabolism (Desulfovibrio sulfodismutans)</th>
      <th>04_Utililization of sugar, conversion of pentose to EMP pathway intermediates</th>
      <th>05_Fatty acid oxidation</th>
      <th>06_Amino acid utilization biosynthesis metabolism</th>
      <th>07_Nucleic acid metabolism</th>
      <th>08_Hydrocarbon degradation</th>
      <th>09_Carbohydrate Active enzyme - CAZy</th>
      <th>10_TCA cycle</th>
      <th>...</th>
      <th>12_Transporters</th>
      <th>13_Hydrogen metabolism</th>
      <th>14_Methanogenesis</th>
      <th>15_Methylotrophy</th>
      <th>16_Embden Meyerhof-Parnos (EMP)</th>
      <th>17_Gluconeogenesis</th>
      <th>18_Sulfur compounds metabolism</th>
      <th>19_Saccharide and derivated synthesis</th>
      <th>20_Hydrolysis of polymers</th>
      <th>21_Cellular response to stress</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.096899</td>
      <td>0.009890</td>
      <td>0.000579</td>
      <td>0.000535</td>
      <td>0.025483</td>
      <td>0.122784</td>
      <td>0.119531</td>
      <td>0.005524</td>
      <td>0.003921</td>
      <td>0.045398</td>
      <td>...</td>
      <td>0.015593</td>
      <td>0.000846</td>
      <td>0.003520</td>
      <td>0.009267</td>
      <td>0.030473</td>
      <td>0.016261</td>
      <td>0.005346</td>
      <td>0.009267</td>
      <td>0.006193</td>
      <td>0.037334</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.104379</td>
      <td>0.010097</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.026447</td>
      <td>0.125917</td>
      <td>0.111970</td>
      <td>0.004937</td>
      <td>0.003630</td>
      <td>0.046199</td>
      <td>...</td>
      <td>0.015868</td>
      <td>0.000653</td>
      <td>0.003944</td>
      <td>0.011636</td>
      <td>0.030150</td>
      <td>0.016665</td>
      <td>0.004634</td>
      <td>0.008585</td>
      <td>0.006081</td>
      <td>0.036379</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.107954</td>
      <td>0.010104</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.027555</td>
      <td>0.125917</td>
      <td>0.111924</td>
      <td>0.004942</td>
      <td>0.003630</td>
      <td>0.046254</td>
      <td>...</td>
      <td>0.015999</td>
      <td>0.000653</td>
      <td>0.003976</td>
      <td>0.011727</td>
      <td>0.030150</td>
      <td>0.016687</td>
      <td>0.004626</td>
      <td>0.008435</td>
      <td>0.006079</td>
      <td>0.036373</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.110575</td>
      <td>0.010110</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.028627</td>
      <td>0.125917</td>
      <td>0.111895</td>
      <td>0.004947</td>
      <td>0.003630</td>
      <td>0.046306</td>
      <td>...</td>
      <td>0.016125</td>
      <td>0.000653</td>
      <td>0.004017</td>
      <td>0.011795</td>
      <td>0.030150</td>
      <td>0.016703</td>
      <td>0.004616</td>
      <td>0.008353</td>
      <td>0.006077</td>
      <td>0.036378</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.112957</td>
      <td>0.010117</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.029633</td>
      <td>0.125917</td>
      <td>0.111872</td>
      <td>0.004952</td>
      <td>0.003630</td>
      <td>0.046364</td>
      <td>...</td>
      <td>0.016263</td>
      <td>0.000653</td>
      <td>0.004060</td>
      <td>0.011858</td>
      <td>0.030150</td>
      <td>0.016720</td>
      <td>0.004606</td>
      <td>0.008309</td>
      <td>0.006075</td>
      <td>0.036393</td>
    </tr>
    <tr>
      <th>5</th>
      <td>0.115460</td>
      <td>0.010125</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.030625</td>
      <td>0.125917</td>
      <td>0.111851</td>
      <td>0.004957</td>
      <td>0.003630</td>
      <td>0.046434</td>
      <td>...</td>
      <td>0.016427</td>
      <td>0.000653</td>
      <td>0.004105</td>
      <td>0.011927</td>
      <td>0.030150</td>
      <td>0.016741</td>
      <td>0.004598</td>
      <td>0.008297</td>
      <td>0.006072</td>
      <td>0.036419</td>
    </tr>
    <tr>
      <th>6</th>
      <td>0.118505</td>
      <td>0.010138</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.031655</td>
      <td>0.125917</td>
      <td>0.111830</td>
      <td>0.004962</td>
      <td>0.003630</td>
      <td>0.046530</td>
      <td>...</td>
      <td>0.016638</td>
      <td>0.000653</td>
      <td>0.004150</td>
      <td>0.012011</td>
      <td>0.030150</td>
      <td>0.016773</td>
      <td>0.004591</td>
      <td>0.008324</td>
      <td>0.006070</td>
      <td>0.036465</td>
    </tr>
    <tr>
      <th>7</th>
      <td>0.123112</td>
      <td>0.010161</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.032694</td>
      <td>0.125917</td>
      <td>0.111804</td>
      <td>0.004966</td>
      <td>0.003630</td>
      <td>0.046688</td>
      <td>...</td>
      <td>0.016966</td>
      <td>0.000653</td>
      <td>0.004197</td>
      <td>0.012141</td>
      <td>0.030150</td>
      <td>0.016836</td>
      <td>0.004588</td>
      <td>0.008426</td>
      <td>0.006067</td>
      <td>0.036559</td>
    </tr>
    <tr>
      <th>8</th>
      <td>0.133734</td>
      <td>0.011645</td>
      <td>0.000340</td>
      <td>0.000870</td>
      <td>0.033575</td>
      <td>0.129235</td>
      <td>0.108779</td>
      <td>0.004688</td>
      <td>0.003063</td>
      <td>0.052367</td>
      <td>...</td>
      <td>0.018224</td>
      <td>0.001059</td>
      <td>0.004159</td>
      <td>0.016334</td>
      <td>0.035428</td>
      <td>0.018678</td>
      <td>0.003592</td>
      <td>0.009150</td>
      <td>0.006692</td>
      <td>0.042688</td>
    </tr>
    <tr>
      <th>9</th>
      <td>0.123435</td>
      <td>0.010154</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.033030</td>
      <td>0.125917</td>
      <td>0.111790</td>
      <td>0.004976</td>
      <td>0.003630</td>
      <td>0.046701</td>
      <td>...</td>
      <td>0.017093</td>
      <td>0.000653</td>
      <td>0.004302</td>
      <td>0.012159</td>
      <td>0.030150</td>
      <td>0.016790</td>
      <td>0.004592</td>
      <td>0.008234</td>
      <td>0.006062</td>
      <td>0.036497</td>
    </tr>
    <tr>
      <th>10</th>
      <td>0.119160</td>
      <td>0.010124</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.032334</td>
      <td>0.125917</td>
      <td>0.111801</td>
      <td>0.004981</td>
      <td>0.003630</td>
      <td>0.046557</td>
      <td>...</td>
      <td>0.016899</td>
      <td>0.000653</td>
      <td>0.004362</td>
      <td>0.012048</td>
      <td>0.030150</td>
      <td>0.016679</td>
      <td>0.004598</td>
      <td>0.007934</td>
      <td>0.006060</td>
      <td>0.036339</td>
    </tr>
    <tr>
      <th>11</th>
      <td>0.116464</td>
      <td>0.010103</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.031667</td>
      <td>0.125917</td>
      <td>0.111808</td>
      <td>0.004986</td>
      <td>0.003630</td>
      <td>0.046478</td>
      <td>...</td>
      <td>0.016835</td>
      <td>0.000653</td>
      <td>0.004425</td>
      <td>0.011984</td>
      <td>0.030150</td>
      <td>0.016595</td>
      <td>0.004602</td>
      <td>0.007689</td>
      <td>0.006057</td>
      <td>0.036223</td>
    </tr>
    <tr>
      <th>12</th>
      <td>0.114334</td>
      <td>0.010085</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.031067</td>
      <td>0.125917</td>
      <td>0.111815</td>
      <td>0.004991</td>
      <td>0.003630</td>
      <td>0.046428</td>
      <td>...</td>
      <td>0.016847</td>
      <td>0.000653</td>
      <td>0.004486</td>
      <td>0.011937</td>
      <td>0.030150</td>
      <td>0.016514</td>
      <td>0.004603</td>
      <td>0.007452</td>
      <td>0.006054</td>
      <td>0.036115</td>
    </tr>
    <tr>
      <th>13</th>
      <td>0.112345</td>
      <td>0.010067</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.030503</td>
      <td>0.125917</td>
      <td>0.111825</td>
      <td>0.004997</td>
      <td>0.003630</td>
      <td>0.046399</td>
      <td>...</td>
      <td>0.016938</td>
      <td>0.000653</td>
      <td>0.004540</td>
      <td>0.011898</td>
      <td>0.030150</td>
      <td>0.016423</td>
      <td>0.004597</td>
      <td>0.007184</td>
      <td>0.006051</td>
      <td>0.035997</td>
    </tr>
    <tr>
      <th>14</th>
      <td>0.110036</td>
      <td>0.010043</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.029966</td>
      <td>0.125917</td>
      <td>0.111846</td>
      <td>0.005002</td>
      <td>0.003630</td>
      <td>0.046398</td>
      <td>...</td>
      <td>0.017176</td>
      <td>0.000653</td>
      <td>0.004575</td>
      <td>0.011857</td>
      <td>0.030150</td>
      <td>0.016293</td>
      <td>0.004580</td>
      <td>0.006819</td>
      <td>0.006048</td>
      <td>0.035833</td>
    </tr>
  </tbody>
</table>
<p>15 rows × 21 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">imputed_MT_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;MT_metabolite_imputed_data.csv&#39;</span><span class="p">)</span>
<span class="n">imputed_MT_data</span> <span class="o">=</span> <span class="n">imputed_MT_data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>


<span class="n">imputed_MT_data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>01_Fermentation</th>
      <th>02_Homoacetogenesis</th>
      <th>03_Superpathway of thiosulfate metabolism (Desulfovibrio sulfodismutans)</th>
      <th>04_Utililization of sugar, conversion of pentose to EMP pathway intermediates</th>
      <th>05_Fatty acid oxidation</th>
      <th>06_Amino acid utilization biosynthesis metabolism</th>
      <th>07_Nucleic acid metabolism</th>
      <th>08_Hydrocarbon degradation</th>
      <th>09_Carbohydrate Active enzyme - CAZy</th>
      <th>10_TCA cycle</th>
      <th>...</th>
      <th>12_Transporters</th>
      <th>13_Hydrogen metabolism</th>
      <th>14_Methanogenesis</th>
      <th>15_Methylotrophy</th>
      <th>16_Embden Meyerhof-Parnos (EMP)</th>
      <th>17_Gluconeogenesis</th>
      <th>18_Sulfur compounds metabolism</th>
      <th>19_Saccharide and derivated synthesis</th>
      <th>20_Hydrolysis of polymers</th>
      <th>21_Cellular response to stress</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.032555</td>
      <td>0.004332</td>
      <td>6.551075e-04</td>
      <td>0.000551</td>
      <td>0.007245</td>
      <td>0.050208</td>
      <td>0.018190</td>
      <td>0.001434</td>
      <td>0.007091</td>
      <td>0.007711</td>
      <td>...</td>
      <td>0.013046</td>
      <td>7.636352e-04</td>
      <td>0.001359</td>
      <td>0.003455</td>
      <td>0.009246</td>
      <td>0.004036</td>
      <td>0.002204</td>
      <td>0.015901</td>
      <td>0.007996</td>
      <td>0.044404</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.032376</td>
      <td>0.004261</td>
      <td>1.417759e-20</td>
      <td>0.000649</td>
      <td>0.007283</td>
      <td>0.047614</td>
      <td>0.018241</td>
      <td>0.001163</td>
      <td>0.007493</td>
      <td>0.007491</td>
      <td>...</td>
      <td>0.012228</td>
      <td>1.703635e-19</td>
      <td>0.001268</td>
      <td>0.002820</td>
      <td>0.008107</td>
      <td>0.003502</td>
      <td>0.002264</td>
      <td>0.016145</td>
      <td>0.007861</td>
      <td>0.044433</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.032329</td>
      <td>0.004261</td>
      <td>1.417756e-20</td>
      <td>0.000649</td>
      <td>0.007282</td>
      <td>0.047304</td>
      <td>0.018290</td>
      <td>0.001163</td>
      <td>0.007542</td>
      <td>0.007490</td>
      <td>...</td>
      <td>0.012153</td>
      <td>1.703597e-19</td>
      <td>0.001268</td>
      <td>0.002820</td>
      <td>0.008107</td>
      <td>0.003502</td>
      <td>0.002264</td>
      <td>0.016171</td>
      <td>0.007860</td>
      <td>0.044494</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.032282</td>
      <td>0.004261</td>
      <td>1.417798e-20</td>
      <td>0.000649</td>
      <td>0.007282</td>
      <td>0.047108</td>
      <td>0.018347</td>
      <td>0.001163</td>
      <td>0.007596</td>
      <td>0.007489</td>
      <td>...</td>
      <td>0.012072</td>
      <td>1.703636e-19</td>
      <td>0.001268</td>
      <td>0.002820</td>
      <td>0.008107</td>
      <td>0.003502</td>
      <td>0.002264</td>
      <td>0.016193</td>
      <td>0.007859</td>
      <td>0.044560</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.032235</td>
      <td>0.004261</td>
      <td>1.417867e-20</td>
      <td>0.000649</td>
      <td>0.007282</td>
      <td>0.046958</td>
      <td>0.018413</td>
      <td>0.001163</td>
      <td>0.007650</td>
      <td>0.007488</td>
      <td>...</td>
      <td>0.011985</td>
      <td>1.703717e-19</td>
      <td>0.001268</td>
      <td>0.002820</td>
      <td>0.008107</td>
      <td>0.003502</td>
      <td>0.002264</td>
      <td>0.016214</td>
      <td>0.007858</td>
      <td>0.044633</td>
    </tr>
    <tr>
      <th>5</th>
      <td>0.032187</td>
      <td>0.004261</td>
      <td>1.417960e-20</td>
      <td>0.000649</td>
      <td>0.007282</td>
      <td>0.046824</td>
      <td>0.018489</td>
      <td>0.001163</td>
      <td>0.007695</td>
      <td>0.007487</td>
      <td>...</td>
      <td>0.011894</td>
      <td>1.703833e-19</td>
      <td>0.001268</td>
      <td>0.002820</td>
      <td>0.008107</td>
      <td>0.003502</td>
      <td>0.002264</td>
      <td>0.016235</td>
      <td>0.007857</td>
      <td>0.044712</td>
    </tr>
    <tr>
      <th>6</th>
      <td>0.032140</td>
      <td>0.004261</td>
      <td>1.418084e-20</td>
      <td>0.000649</td>
      <td>0.007282</td>
      <td>0.046689</td>
      <td>0.018577</td>
      <td>0.001163</td>
      <td>0.007726</td>
      <td>0.007485</td>
      <td>...</td>
      <td>0.011798</td>
      <td>1.703990e-19</td>
      <td>0.001268</td>
      <td>0.002820</td>
      <td>0.008107</td>
      <td>0.003502</td>
      <td>0.002264</td>
      <td>0.016256</td>
      <td>0.007856</td>
      <td>0.044790</td>
    </tr>
    <tr>
      <th>7</th>
      <td>0.032093</td>
      <td>0.004261</td>
      <td>1.418264e-20</td>
      <td>0.000649</td>
      <td>0.007282</td>
      <td>0.046514</td>
      <td>0.018671</td>
      <td>0.001163</td>
      <td>0.007735</td>
      <td>0.007484</td>
      <td>...</td>
      <td>0.011699</td>
      <td>1.704220e-19</td>
      <td>0.001268</td>
      <td>0.002820</td>
      <td>0.008107</td>
      <td>0.003502</td>
      <td>0.002264</td>
      <td>0.016273</td>
      <td>0.007854</td>
      <td>0.044850</td>
    </tr>
    <tr>
      <th>8</th>
      <td>0.032122</td>
      <td>0.004362</td>
      <td>8.355870e-04</td>
      <td>0.000600</td>
      <td>0.007049</td>
      <td>0.045215</td>
      <td>0.019025</td>
      <td>0.001449</td>
      <td>0.008429</td>
      <td>0.007428</td>
      <td>...</td>
      <td>0.011506</td>
      <td>6.943123e-04</td>
      <td>0.001097</td>
      <td>0.003125</td>
      <td>0.009361</td>
      <td>0.003913</td>
      <td>0.002097</td>
      <td>0.015956</td>
      <td>0.008593</td>
      <td>0.045070</td>
    </tr>
    <tr>
      <th>9</th>
      <td>0.031966</td>
      <td>0.004261</td>
      <td>1.418395e-20</td>
      <td>0.000649</td>
      <td>0.007283</td>
      <td>0.046505</td>
      <td>0.018789</td>
      <td>0.001163</td>
      <td>0.007679</td>
      <td>0.007480</td>
      <td>...</td>
      <td>0.011494</td>
      <td>1.704403e-19</td>
      <td>0.001268</td>
      <td>0.002820</td>
      <td>0.008107</td>
      <td>0.003502</td>
      <td>0.002264</td>
      <td>0.016354</td>
      <td>0.007850</td>
      <td>0.044836</td>
    </tr>
    <tr>
      <th>10</th>
      <td>0.031887</td>
      <td>0.004261</td>
      <td>1.418347e-20</td>
      <td>0.000649</td>
      <td>0.007284</td>
      <td>0.046678</td>
      <td>0.018818</td>
      <td>0.001163</td>
      <td>0.007616</td>
      <td>0.007478</td>
      <td>...</td>
      <td>0.011390</td>
      <td>1.704356e-19</td>
      <td>0.001268</td>
      <td>0.002820</td>
      <td>0.008107</td>
      <td>0.003502</td>
      <td>0.002264</td>
      <td>0.016421</td>
      <td>0.007848</td>
      <td>0.044762</td>
    </tr>
    <tr>
      <th>11</th>
      <td>0.031808</td>
      <td>0.004261</td>
      <td>1.418359e-20</td>
      <td>0.000649</td>
      <td>0.007284</td>
      <td>0.046822</td>
      <td>0.018866</td>
      <td>0.001163</td>
      <td>0.007537</td>
      <td>0.007476</td>
      <td>...</td>
      <td>0.011286</td>
      <td>1.704385e-19</td>
      <td>0.001268</td>
      <td>0.002820</td>
      <td>0.008107</td>
      <td>0.003502</td>
      <td>0.002264</td>
      <td>0.016494</td>
      <td>0.007846</td>
      <td>0.044671</td>
    </tr>
    <tr>
      <th>12</th>
      <td>0.031729</td>
      <td>0.004261</td>
      <td>1.418405e-20</td>
      <td>0.000649</td>
      <td>0.007286</td>
      <td>0.046989</td>
      <td>0.018941</td>
      <td>0.001163</td>
      <td>0.007454</td>
      <td>0.007474</td>
      <td>...</td>
      <td>0.011183</td>
      <td>1.704460e-19</td>
      <td>0.001268</td>
      <td>0.002820</td>
      <td>0.008107</td>
      <td>0.003502</td>
      <td>0.002264</td>
      <td>0.016584</td>
      <td>0.007844</td>
      <td>0.044574</td>
    </tr>
    <tr>
      <th>13</th>
      <td>0.031650</td>
      <td>0.004261</td>
      <td>1.418484e-20</td>
      <td>0.000649</td>
      <td>0.007287</td>
      <td>0.047226</td>
      <td>0.019043</td>
      <td>0.001163</td>
      <td>0.007376</td>
      <td>0.007472</td>
      <td>...</td>
      <td>0.011082</td>
      <td>1.704580e-19</td>
      <td>0.001268</td>
      <td>0.002820</td>
      <td>0.008107</td>
      <td>0.003502</td>
      <td>0.002264</td>
      <td>0.016709</td>
      <td>0.007841</td>
      <td>0.044473</td>
    </tr>
    <tr>
      <th>14</th>
      <td>0.031571</td>
      <td>0.004261</td>
      <td>1.418615e-20</td>
      <td>0.000649</td>
      <td>0.007288</td>
      <td>0.047657</td>
      <td>0.019140</td>
      <td>0.001163</td>
      <td>0.007313</td>
      <td>0.007470</td>
      <td>...</td>
      <td>0.010982</td>
      <td>1.704772e-19</td>
      <td>0.001268</td>
      <td>0.002820</td>
      <td>0.008107</td>
      <td>0.003502</td>
      <td>0.002264</td>
      <td>0.016922</td>
      <td>0.007838</td>
      <td>0.044353</td>
    </tr>
  </tbody>
</table>
<p>15 rows × 21 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer12</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MT_data</span><span class="p">)</span>

<span class="n">infer12</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 23:50<00:00 Sampling 2 chains, 2,000 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 1434 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.218  0.008   0.210    0.226      0.006    0.005       3.0
Ah[0, 1]   -0.004  0.000  -0.004   -0.003      0.000    0.000       2.0
Ah[0, 2]    0.078  0.040   0.038    0.118      0.028    0.024       3.0
Ah[0, 3]    0.058  0.006   0.051    0.064      0.004    0.004       2.0
Ah[0, 4]    0.071  0.001   0.070    0.072      0.001    0.001       3.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.149  0.366  -0.253    0.515      0.258    0.219       2.0
Bh[20, 8]   0.171  0.233  -0.062    0.405      0.164    0.139       3.0
Bh[20, 9]  -0.819  0.094  -0.954   -0.729      0.066    0.055       2.0
Bh[20, 10]  0.100  0.289  -0.191    0.389      0.204    0.172       2.0
Bh[20, 11]  0.017  0.119  -0.102    0.137      0.084    0.071       2.0

            ess_tail  r_hat
Ah[0, 0]        20.0   1.95
Ah[0, 1]        18.0   2.35
Ah[0, 2]        28.0   2.22
Ah[0, 3]        11.0   2.63
Ah[0, 4]        18.0   1.95
...              ...    ...
Bh[20, 7]       11.0   2.30
Bh[20, 8]       11.0   2.23
Bh[20, 9]       11.0   2.66
Bh[20, 10]      11.0   2.68
Bh[20, 11]      13.0   2.64

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_xs_v6.nc
Data file: data_xs_v6.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_43_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_43_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer12</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_44_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_44_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_44_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_44_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer13</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MT_data</span><span class="p">)</span>

<span class="n">infer13</span><span class="o">.</span><span class="n">run_inference_large_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [c2_A, tau_A, lam_A, Ah, c2_B, tau_B, lam_B, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='24000' class='' max='24000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [24000/24000 1:19:46<00:00 Sampling 4 chains, 4,001 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 4 chains for 2_000 tune and 4_000 draw iterations (8_000 + 16_000 draws total) took 4794 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
There were 4001 divergences after tuning. Increase `target_accept` or reparameterize.
Chain 1 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 2 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 3 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.328  0.217   0.096    0.671      0.108    0.083       4.0
Ah[0, 1]    0.016  0.030  -0.015    0.062      0.015    0.011       4.0
Ah[0, 2]   -0.098  0.141  -0.279    0.067      0.071    0.054       4.0
Ah[0, 3]    0.119  0.051   0.046    0.188      0.025    0.019       4.0
Ah[0, 4]    0.008  0.053  -0.082    0.058      0.027    0.020       4.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]  -0.207  0.196  -0.348    0.129      0.098    0.075       4.0
Bh[20, 8]   0.202  0.128   0.010    0.350      0.064    0.049       4.0
Bh[20, 9]  -0.063  0.362  -0.409    0.501      0.181    0.138       4.0
Bh[20, 10] -0.280  0.236  -0.562    0.041      0.118    0.090       4.0
Bh[20, 11] -0.136  0.309  -0.615    0.232      0.154    0.118       5.0

            ess_tail  r_hat
Ah[0, 0]        12.0   3.60
Ah[0, 1]        13.0   3.61
Ah[0, 2]        19.0   3.18
Ah[0, 3]        11.0   3.81
Ah[0, 4]        13.0   3.44
...              ...    ...
Bh[20, 7]       12.0   4.12
Bh[20, 8]       15.0   3.24
Bh[20, 9]       13.0   3.55
Bh[20, 10]      11.0   3.47
Bh[20, 11]      15.0   3.51

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_large_xs.nc
Data file: data_large_xs.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_45_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_45_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer13</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_46_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_46_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_46_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_46_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">imputed_MP_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;MP_metabolite_imputed_data.csv&#39;</span><span class="p">)</span>
<span class="n">imputed_MP_data</span> <span class="o">=</span> <span class="n">imputed_MP_data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>

<span class="n">imputed_MP_data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>01_Fermentation</th>
      <th>02_Homoacetogenesis</th>
      <th>03_Superpathway of thiosulfate metabolism (Desulfovibrio sulfodismutans)</th>
      <th>04_Utililization of sugar, conversion of pentose to EMP pathway intermediates</th>
      <th>05_Fatty acid oxidation</th>
      <th>06_Amino acid utilization biosynthesis metabolism</th>
      <th>07_Nucleic acid metabolism</th>
      <th>08_Hydrocarbon degradation</th>
      <th>09_Carbohydrate Active enzyme - CAZy</th>
      <th>10_TCA cycle</th>
      <th>...</th>
      <th>12_Transporters</th>
      <th>13_Hydrogen metabolism</th>
      <th>14_Methanogenesis</th>
      <th>15_Methylotrophy</th>
      <th>16_Embden Meyerhof-Parnos (EMP)</th>
      <th>17_Gluconeogenesis</th>
      <th>18_Sulfur compounds metabolism</th>
      <th>19_Saccharide and derivated synthesis</th>
      <th>20_Hydrolysis of polymers</th>
      <th>21_Cellular response to stress</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.096899</td>
      <td>0.009890</td>
      <td>0.000579</td>
      <td>0.000535</td>
      <td>0.025483</td>
      <td>0.122784</td>
      <td>0.119531</td>
      <td>0.005524</td>
      <td>0.003921</td>
      <td>0.045398</td>
      <td>...</td>
      <td>0.015593</td>
      <td>0.000846</td>
      <td>0.003520</td>
      <td>0.009267</td>
      <td>0.030473</td>
      <td>0.016261</td>
      <td>0.005346</td>
      <td>0.009267</td>
      <td>0.006193</td>
      <td>0.037334</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.104379</td>
      <td>0.010097</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.026447</td>
      <td>0.125917</td>
      <td>0.111970</td>
      <td>0.004937</td>
      <td>0.003630</td>
      <td>0.046199</td>
      <td>...</td>
      <td>0.015868</td>
      <td>0.000653</td>
      <td>0.003944</td>
      <td>0.011636</td>
      <td>0.030150</td>
      <td>0.016665</td>
      <td>0.004634</td>
      <td>0.008585</td>
      <td>0.006081</td>
      <td>0.036379</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.107954</td>
      <td>0.010104</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.027555</td>
      <td>0.125917</td>
      <td>0.111924</td>
      <td>0.004942</td>
      <td>0.003630</td>
      <td>0.046254</td>
      <td>...</td>
      <td>0.015999</td>
      <td>0.000653</td>
      <td>0.003976</td>
      <td>0.011727</td>
      <td>0.030150</td>
      <td>0.016687</td>
      <td>0.004626</td>
      <td>0.008435</td>
      <td>0.006079</td>
      <td>0.036373</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.110575</td>
      <td>0.010110</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.028627</td>
      <td>0.125917</td>
      <td>0.111895</td>
      <td>0.004947</td>
      <td>0.003630</td>
      <td>0.046306</td>
      <td>...</td>
      <td>0.016125</td>
      <td>0.000653</td>
      <td>0.004017</td>
      <td>0.011795</td>
      <td>0.030150</td>
      <td>0.016703</td>
      <td>0.004616</td>
      <td>0.008353</td>
      <td>0.006077</td>
      <td>0.036378</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.112957</td>
      <td>0.010117</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.029633</td>
      <td>0.125917</td>
      <td>0.111872</td>
      <td>0.004952</td>
      <td>0.003630</td>
      <td>0.046364</td>
      <td>...</td>
      <td>0.016263</td>
      <td>0.000653</td>
      <td>0.004060</td>
      <td>0.011858</td>
      <td>0.030150</td>
      <td>0.016720</td>
      <td>0.004606</td>
      <td>0.008309</td>
      <td>0.006075</td>
      <td>0.036393</td>
    </tr>
    <tr>
      <th>5</th>
      <td>0.115460</td>
      <td>0.010125</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.030625</td>
      <td>0.125917</td>
      <td>0.111851</td>
      <td>0.004957</td>
      <td>0.003630</td>
      <td>0.046434</td>
      <td>...</td>
      <td>0.016427</td>
      <td>0.000653</td>
      <td>0.004105</td>
      <td>0.011927</td>
      <td>0.030150</td>
      <td>0.016741</td>
      <td>0.004598</td>
      <td>0.008297</td>
      <td>0.006072</td>
      <td>0.036419</td>
    </tr>
    <tr>
      <th>6</th>
      <td>0.118505</td>
      <td>0.010138</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.031655</td>
      <td>0.125917</td>
      <td>0.111830</td>
      <td>0.004962</td>
      <td>0.003630</td>
      <td>0.046530</td>
      <td>...</td>
      <td>0.016638</td>
      <td>0.000653</td>
      <td>0.004150</td>
      <td>0.012011</td>
      <td>0.030150</td>
      <td>0.016773</td>
      <td>0.004591</td>
      <td>0.008324</td>
      <td>0.006070</td>
      <td>0.036465</td>
    </tr>
    <tr>
      <th>7</th>
      <td>0.123112</td>
      <td>0.010161</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.032694</td>
      <td>0.125917</td>
      <td>0.111804</td>
      <td>0.004966</td>
      <td>0.003630</td>
      <td>0.046688</td>
      <td>...</td>
      <td>0.016966</td>
      <td>0.000653</td>
      <td>0.004197</td>
      <td>0.012141</td>
      <td>0.030150</td>
      <td>0.016836</td>
      <td>0.004588</td>
      <td>0.008426</td>
      <td>0.006067</td>
      <td>0.036559</td>
    </tr>
    <tr>
      <th>8</th>
      <td>0.133734</td>
      <td>0.011645</td>
      <td>0.000340</td>
      <td>0.000870</td>
      <td>0.033575</td>
      <td>0.129235</td>
      <td>0.108779</td>
      <td>0.004688</td>
      <td>0.003063</td>
      <td>0.052367</td>
      <td>...</td>
      <td>0.018224</td>
      <td>0.001059</td>
      <td>0.004159</td>
      <td>0.016334</td>
      <td>0.035428</td>
      <td>0.018678</td>
      <td>0.003592</td>
      <td>0.009150</td>
      <td>0.006692</td>
      <td>0.042688</td>
    </tr>
    <tr>
      <th>9</th>
      <td>0.123435</td>
      <td>0.010154</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.033030</td>
      <td>0.125917</td>
      <td>0.111790</td>
      <td>0.004976</td>
      <td>0.003630</td>
      <td>0.046701</td>
      <td>...</td>
      <td>0.017093</td>
      <td>0.000653</td>
      <td>0.004302</td>
      <td>0.012159</td>
      <td>0.030150</td>
      <td>0.016790</td>
      <td>0.004592</td>
      <td>0.008234</td>
      <td>0.006062</td>
      <td>0.036497</td>
    </tr>
    <tr>
      <th>10</th>
      <td>0.119160</td>
      <td>0.010124</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.032334</td>
      <td>0.125917</td>
      <td>0.111801</td>
      <td>0.004981</td>
      <td>0.003630</td>
      <td>0.046557</td>
      <td>...</td>
      <td>0.016899</td>
      <td>0.000653</td>
      <td>0.004362</td>
      <td>0.012048</td>
      <td>0.030150</td>
      <td>0.016679</td>
      <td>0.004598</td>
      <td>0.007934</td>
      <td>0.006060</td>
      <td>0.036339</td>
    </tr>
    <tr>
      <th>11</th>
      <td>0.116464</td>
      <td>0.010103</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.031667</td>
      <td>0.125917</td>
      <td>0.111808</td>
      <td>0.004986</td>
      <td>0.003630</td>
      <td>0.046478</td>
      <td>...</td>
      <td>0.016835</td>
      <td>0.000653</td>
      <td>0.004425</td>
      <td>0.011984</td>
      <td>0.030150</td>
      <td>0.016595</td>
      <td>0.004602</td>
      <td>0.007689</td>
      <td>0.006057</td>
      <td>0.036223</td>
    </tr>
    <tr>
      <th>12</th>
      <td>0.114334</td>
      <td>0.010085</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.031067</td>
      <td>0.125917</td>
      <td>0.111815</td>
      <td>0.004991</td>
      <td>0.003630</td>
      <td>0.046428</td>
      <td>...</td>
      <td>0.016847</td>
      <td>0.000653</td>
      <td>0.004486</td>
      <td>0.011937</td>
      <td>0.030150</td>
      <td>0.016514</td>
      <td>0.004603</td>
      <td>0.007452</td>
      <td>0.006054</td>
      <td>0.036115</td>
    </tr>
    <tr>
      <th>13</th>
      <td>0.112345</td>
      <td>0.010067</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.030503</td>
      <td>0.125917</td>
      <td>0.111825</td>
      <td>0.004997</td>
      <td>0.003630</td>
      <td>0.046399</td>
      <td>...</td>
      <td>0.016938</td>
      <td>0.000653</td>
      <td>0.004540</td>
      <td>0.011898</td>
      <td>0.030150</td>
      <td>0.016423</td>
      <td>0.004597</td>
      <td>0.007184</td>
      <td>0.006051</td>
      <td>0.035997</td>
    </tr>
    <tr>
      <th>14</th>
      <td>0.110036</td>
      <td>0.010043</td>
      <td>0.000564</td>
      <td>0.001093</td>
      <td>0.029966</td>
      <td>0.125917</td>
      <td>0.111846</td>
      <td>0.005002</td>
      <td>0.003630</td>
      <td>0.046398</td>
      <td>...</td>
      <td>0.017176</td>
      <td>0.000653</td>
      <td>0.004575</td>
      <td>0.011857</td>
      <td>0.030150</td>
      <td>0.016293</td>
      <td>0.004580</td>
      <td>0.006819</td>
      <td>0.006048</td>
      <td>0.035833</td>
    </tr>
  </tbody>
</table>
<p>15 rows × 21 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer14</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">)</span>
<span class="n">infer14</span><span class="o">.</span><span class="n">run_inference_large_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [c2_A, tau_A, lam_A, Ah, c2_B, tau_B, lam_B, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='24000' class='' max='24000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [24000/24000 1:24:30<00:00 Sampling 4 chains, 2,652 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 4 chains for 2_000 tune and 4_000 draw iterations (8_000 + 16_000 draws total) took 5080 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
There were 2652 divergences after tuning. Increase `target_accept` or reparameterize.
Chain 0 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 1 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 2 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 3 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.309  0.176   0.065    0.509      0.088    0.067       4.0
Ah[0, 1]    0.000  0.016  -0.019    0.022      0.008    0.006       4.0
Ah[0, 2]    0.044  0.175  -0.247    0.221      0.087    0.067       4.0
Ah[0, 3]    0.085  0.024   0.058    0.111      0.012    0.009       4.0
Ah[0, 4]    0.050  0.100  -0.051    0.215      0.050    0.038       4.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.015  0.273  -0.208    0.478      0.136    0.104       4.0
Bh[20, 8]   0.060  0.369  -0.564    0.412      0.184    0.141       4.0
Bh[20, 9]  -0.467  0.190  -0.705   -0.208      0.095    0.073       4.0
Bh[20, 10]  0.331  0.187   0.110    0.608      0.094    0.072       4.0
Bh[20, 11] -0.236  0.143  -0.467   -0.098      0.071    0.055       5.0

            ess_tail  r_hat
Ah[0, 0]        13.0   3.09
Ah[0, 1]        14.0   3.45
Ah[0, 2]        13.0   3.21
Ah[0, 3]        13.0   3.91
Ah[0, 4]        11.0   3.52
...              ...    ...
Bh[20, 7]       11.0   3.30
Bh[20, 8]       11.0   3.38
Bh[20, 9]       12.0   3.73
Bh[20, 10]      13.0   3.39
Bh[20, 11]      22.0   3.10

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_large_xs_v1.nc
Data file: data_large_xs_v1.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_48_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_48_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer14</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_49_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_49_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_49_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_49_1.png" />
</div>
</div>
<hr class="docutils" />
<section id="Reducing-number-of-variables-(to-the-3-or-5-most-abundant)-to-reduce-the-computational-cost-and-simplify-the-analysis">
<h2>Reducing number of variables (to the 3 or 5 most abundant) to reduce the computational cost and simplify the analysis<a class="headerlink" href="#Reducing-number-of-variables-(to-the-3-or-5-most-abundant)-to-reduce-the-computational-cost-and-simplify-the-analysis" title="Link to this heading">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># We are going to develop a method that selects the top 4 or 5 abundances that change the most over time</span>


<span class="k">def</span> <span class="nf">select_top_variable_abundances</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">5</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Selects the top N abundances that change the most over time.</span>
<span class="sd">    Metric Used: Variance of abundances over time.</span>
<span class="sd">    Steps:</span>
<span class="sd">        Converts the data to a NumPy array if it is a DataFrame.</span>
<span class="sd">        Calculates the variance of each abundance across all time points.</span>
<span class="sd">        Sorts the variances in ascending order.</span>
<span class="sd">        Selects the indices of the top N abundances with the highest variance.</span>
<span class="sd">        Output: Indices of the top N most variable abundances.</span>
<span class="sd">        Benefit: Variance captures the overall variability of the abundances, providing a sense of how spread out the values are over time.</span>

<span class="sd">    Args:</span>
<span class="sd">    data (pd.DataFrame or np.ndarray): The abundance data.</span>
<span class="sd">    top_n (int): The number of top variable abundances to select.</span>

<span class="sd">    Returns:</span>
<span class="sd">    np.ndarray: The indices of the selected abundances.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">values</span>

    <span class="n">variances</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">top_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">variances</span><span class="p">)[</span><span class="o">-</span><span class="n">top_n</span><span class="p">:]</span>
    <span class="k">return</span> <span class="n">data</span><span class="p">[:,</span> <span class="n">top_indices</span><span class="p">]</span>


<span class="k">def</span> <span class="nf">top_abundance_changes</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">5</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Selects the top N abundances that change the most over time based on total changes.</span>
<span class="sd">    Metric Used: Sum of absolute changes in abundance between consecutive time points.</span>
<span class="sd">    Steps:</span>
<span class="sd">        Calculates the difference between consecutive time points to find changes.</span>
<span class="sd">        Takes the absolute value of these changes and sums them for each abundance.</span>
<span class="sd">        Sorts these summed changes in descending order.</span>
<span class="sd">        Selects the top N abundances with the highest summed changes.</span>
<span class="sd">        Output: A Series with the top N abundances that have the most change.</span>
<span class="sd">        Benefit: This method focuses on the magnitude of changes between consecutive time points, highlighting abundances that fluctuate the most over the entire period.</span>

<span class="sd">    Args:</span>
<span class="sd">    data (pd.DataFrame): The abundance data.</span>
<span class="sd">    top_n (int): The number of top variable abundances to select.</span>

<span class="sd">    Returns:</span>
<span class="sd">    pd.Series: The abundances with the most changes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Calculate the change in abundance for each genus</span>
    <span class="n">data_change</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">diff</span><span class="p">()</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">top_indices</span> <span class="o">=</span> <span class="n">data_change</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top_n</span><span class="p">)</span><span class="o">.</span><span class="n">index</span>
    <span class="k">return</span> <span class="n">data</span><span class="p">[</span><span class="n">top_indices</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now pick the top 5 most variable abundances for the MG (imputed_data) and the mp (imputed_MP_data) datasets</span>


<span class="n">top_variable_abundances_MG</span> <span class="o">=</span> <span class="n">top_abundance_changes</span><span class="p">(</span><span class="n">imputed_data</span><span class="p">)</span>

<span class="n">top_variable_abundances_MP</span> <span class="o">=</span> <span class="n">top_abundance_changes</span><span class="p">(</span><span class="n">imputed_MP_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Do the inference for these reduced datasets</span>

<span class="n">infer16</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">top_variable_abundances_MG</span><span class="p">,</span>
                    <span class="n">dataS</span><span class="o">=</span><span class="n">top_variable_abundances_MP</span><span class="p">)</span>
<span class="n">infer16</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 06:01<00:00 Sampling 2 chains, 2,036 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 366 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
           mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]  0.994  0.007   0.988    1.006      0.005    0.004       2.0
Ah[0, 1]  0.016  0.011   0.006    0.031      0.008    0.006       2.0
Ah[0, 2]  0.007  0.024  -0.026    0.031      0.017    0.014       3.0
Ah[0, 3] -0.034  0.014  -0.058   -0.021      0.010    0.008       3.0
Ah[0, 4]  0.023  0.012   0.013    0.043      0.008    0.006       3.0
Ah[1, 0]  0.138  0.152  -0.013    0.304      0.107    0.091       3.0
Ah[1, 1]  0.484  0.478   0.005    0.962      0.337    0.286       2.0
Ah[1, 2]  0.028  0.010   0.013    0.038      0.007    0.006       3.0
Ah[1, 3] -0.037  0.123  -0.291    0.063      0.084    0.070       2.0
Ah[1, 4] -0.007  0.013  -0.020    0.008      0.009    0.008       3.0
Ah[2, 0]  0.189  0.182   0.004    0.371      0.129    0.109       3.0
Ah[2, 1]  0.024  0.017   0.003    0.040      0.012    0.010       3.0
Ah[2, 2]  0.503  0.486   0.017    0.996      0.343    0.291       3.0
Ah[2, 3] -0.003  0.012  -0.022    0.009      0.008    0.007       3.0
Ah[2, 4]  0.084  0.085  -0.006    0.169      0.060    0.051       3.0
Ah[3, 0] -0.007  0.002  -0.010   -0.002      0.001    0.001       5.0
Ah[3, 1] -0.006  0.003  -0.009   -0.001      0.002    0.001       3.0
Ah[3, 2]  0.012  0.007  -0.003    0.018      0.003    0.003       5.0
Ah[3, 3]  1.000  0.004   0.992    1.009      0.001    0.001      18.0
Ah[3, 4]  0.012  0.004   0.005    0.022      0.001    0.001      24.0
Ah[4, 0]  0.123  0.124  -0.001    0.255      0.087    0.074       3.0
Ah[4, 1]  0.005  0.001   0.004    0.006      0.000    0.000       4.0
Ah[4, 2]  0.020  0.014   0.007    0.039      0.010    0.008       3.0
Ah[4, 3]  0.060  0.076  -0.015    0.153      0.053    0.045       3.0
Ah[4, 4]  0.501  0.493   0.004    0.994      0.348    0.295       2.0
Bh[0, 0] -0.024  0.298  -0.322    0.289      0.210    0.178       3.0
Bh[0, 1] -0.325  0.294  -0.648   -0.032      0.207    0.176       3.0
Bh[0, 2]  0.446  0.415   0.029    0.862      0.293    0.248       3.0
Bh[0, 3] -0.016  0.039  -0.095    0.046      0.025    0.021       3.0
Bh[0, 4] -0.260  0.124  -0.440   -0.140      0.086    0.072       3.0
Bh[1, 0] -0.047  0.353  -0.401    0.325      0.249    0.211       3.0
Bh[1, 1] -0.480  0.451  -1.087   -0.039      0.313    0.263       2.0
Bh[1, 2]  0.550  0.520   0.025    1.070      0.367    0.311       3.0
Bh[1, 3] -0.050  0.067  -0.208    0.020      0.040    0.032       4.0
Bh[1, 4] -0.166  0.129  -0.417    0.143      0.043    0.031       8.0
Bh[2, 0]  0.119  0.182  -0.063    0.312      0.128    0.109       3.0
Bh[2, 1] -0.445  0.441  -0.907   -0.004      0.311    0.264       3.0
Bh[2, 2]  0.101  0.071   0.026    0.173      0.050    0.043       3.0
Bh[2, 3] -0.043  0.072  -0.216    0.013      0.047    0.039       3.0
Bh[2, 4] -0.100  0.088  -0.242   -0.025      0.056    0.046       2.0
Bh[3, 0] -0.020  0.140  -0.159    0.126      0.098    0.083       3.0
Bh[3, 1]  0.019  0.058  -0.052    0.152      0.035    0.028       4.0
Bh[3, 2]  0.222  0.203   0.017    0.425      0.143    0.121       3.0
Bh[3, 3]  0.038  0.046  -0.007    0.098      0.032    0.026       3.0
Bh[3, 4] -0.306  0.242  -0.617   -0.069      0.169    0.143       2.0
Bh[4, 0] -0.002  0.096  -0.098    0.102      0.068    0.057       3.0
Bh[4, 1]  0.165  0.177  -0.010    0.385      0.124    0.104       3.0
Bh[4, 2]  0.141  0.120   0.016    0.262      0.085    0.072       3.0
Bh[4, 3]  0.075  0.085  -0.004    0.217      0.059    0.049       2.0
Bh[4, 4] -0.408  0.366  -0.810   -0.043      0.258    0.218       3.0

          ess_tail  r_hat
Ah[0, 0]      16.0   2.92
Ah[0, 1]      20.0   2.45
Ah[0, 2]      12.0   2.15
Ah[0, 3]      12.0   2.05
Ah[0, 4]      29.0   2.39
Ah[1, 0]       8.0   1.91
Ah[1, 1]      37.0   2.69
Ah[1, 2]      21.0   2.38
Ah[1, 3]      11.0   2.68
Ah[1, 4]       8.0   2.20
Ah[2, 0]      23.0   2.21
Ah[2, 1]      17.0   1.96
Ah[2, 2]      43.0   1.91
Ah[2, 3]      25.0   1.84
Ah[2, 4]      61.0   1.86
Ah[3, 0]      14.0   2.22
Ah[3, 1]      13.0   2.36
Ah[3, 2]      18.0   2.03
Ah[3, 3]      19.0   1.93
Ah[3, 4]      37.0   2.21
Ah[4, 0]      11.0   2.22
Ah[4, 1]      24.0   2.46
Ah[4, 2]      21.0   2.35
Ah[4, 3]       7.0   1.98
Ah[4, 4]      13.0   2.78
Bh[0, 0]      38.0   2.21
Bh[0, 1]      11.0   2.24
Bh[0, 2]      55.0   1.96
Bh[0, 3]      11.0   2.22
Bh[0, 4]      43.0   2.11
Bh[1, 0]      12.0   2.04
Bh[1, 1]      32.0   2.43
Bh[1, 2]      11.0   2.23
Bh[1, 3]      12.0   2.45
Bh[1, 4]      22.0   1.96
Bh[2, 0]      16.0   2.35
Bh[2, 1]      31.0   1.90
Bh[2, 2]      21.0   1.95
Bh[2, 3]      11.0   2.51
Bh[2, 4]      15.0   2.34
Bh[3, 0]      12.0   2.19
Bh[3, 1]      21.0   2.21
Bh[3, 2]      17.0   2.22
Bh[3, 3]      21.0   2.39
Bh[3, 4]      20.0   2.85
Bh[4, 0]      12.0   1.87
Bh[4, 1]      25.0   1.97
Bh[4, 2]      14.0   2.20
Bh[4, 3]      11.0   2.50
Bh[4, 4]      13.0   2.05
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (50) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_xs_v7.nc
Data file: data_xs_v7.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_53_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_53_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">infer16</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_54_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_54_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_54_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_54_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">imputed_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;imputed_data.csv&#39;</span><span class="p">)</span>
<span class="c1"># create a pandas dataframe of the original imputed data_set with days as the index</span>
<span class="n">imputed_data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;days&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">imputed_MP_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;MP_metabolite_imputed_data.csv&#39;</span><span class="p">)</span>
<span class="n">imputed_MP_data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># now use the VAR model with these as priors</span>
<span class="n">inferWithoutPrior</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">)</span>
<span class="n">inferWithoutPrior</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 23:54<00:00 Sampling 2 chains, 0 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 1439 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.133  0.047   0.086    0.180      0.033    0.028       3.0
Ah[0, 1]   -0.002  0.004  -0.006    0.003      0.003    0.003       2.0
Ah[0, 2]   -0.010  0.072  -0.082    0.062      0.051    0.043       3.0
Ah[0, 3]    0.045  0.008   0.037    0.053      0.006    0.005       3.0
Ah[0, 4]    0.125  0.007   0.118    0.132      0.005    0.004       2.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]  -0.159  0.108  -0.271   -0.050      0.076    0.065       3.0
Bh[20, 8]   0.107  0.379  -0.271    0.487      0.267    0.226       3.0
Bh[20, 9]   0.381  0.208   0.172    0.589      0.147    0.124       2.0
Bh[20, 10]  0.106  0.052   0.051    0.158      0.037    0.031       2.0
Bh[20, 11]  0.323  0.513  -0.190    0.837      0.362    0.307       3.0

            ess_tail  r_hat
Ah[0, 0]        11.0   2.31
Ah[0, 1]        25.0   2.81
Ah[0, 2]        14.0   2.16
Ah[0, 3]        11.0   2.21
Ah[0, 4]        11.0   2.40
...              ...    ...
Bh[20, 7]       11.0   2.02
Bh[20, 8]       14.0   2.23
Bh[20, 9]       18.0   2.46
Bh[20, 10]      18.0   2.94
Bh[20, 11]      11.0   2.16

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_xs_v8.nc
Data file: data_xs_v8.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_57_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_57_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[51]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithoutPrior</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_58_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_58_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_58_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_58_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statsmodels.tsa.api</span> <span class="kn">import</span> <span class="n">VAR</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># make VAR model with the imputed data set and statsmodels library</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">VAR</span><span class="p">(</span><span class="n">imputed_data</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">maxlags</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ml&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning: An unsupported index was provided and will be ignored when e.g. forecasting.
  self._init_dates(dates, freq)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">coefficients</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">params</span>
<span class="n">covariance_matrix</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">sigma_u</span>
<span class="n">intercepts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">results</span><span class="o">.</span><span class="n">params</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

<span class="c1"># remove the first row of the coefficients matrix</span>
<span class="n">coefficients</span> <span class="o">=</span> <span class="n">coefficients</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
<span class="c1"># transpose the coefficients matrix</span>
<span class="n">coefficients</span> <span class="o">=</span> <span class="n">coefficients</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">coefficients</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
                       L1.Acidimicrobium  L1.Acinetobacter  L1.Albidiferax  \
Acidimicrobium                  0.963205         -0.004673       -0.033426
Acinetobacter                  -0.116193          0.909450       -0.122479
Albidiferax                     0.036651         -0.005288        1.025312
Candidatus Microthrix          -0.075924         -0.049265       -0.105313
Chitinophaga pinensis          -0.115842         -0.047209        0.075176
Dechloromonas                   0.014449         -0.000909        0.004961
Haliscomenobacter              -0.007802         -0.004379       -0.045865
Intrasporangium                 0.039053          0.026513        0.021795
Leptospira                      0.165565          0.043335       -0.005578
Other                           0.014019          0.005023        0.029675
Xanthomonas                    -0.027268         -0.002598        0.014894
mean abundance &lt; 2%            -0.010830          0.006473        0.067992

                       L1.Candidatus Microthrix  L1.Chitinophaga pinensis  \
Acidimicrobium                        -0.007765                 -0.003890
Acinetobacter                         -0.053319                 -0.055020
Albidiferax                           -0.010315                 -0.017732
Candidatus Microthrix                  0.945760                 -0.062214
Chitinophaga pinensis                 -0.032266                  0.916255
Dechloromonas                         -0.001062                 -0.000513
Haliscomenobacter                     -0.017300                  0.005880
Intrasporangium                        0.009746                  0.018528
Leptospira                             0.028434                  0.080814
Other                                  0.005560                  0.001574
Xanthomonas                            0.000006                 -0.007796
mean abundance &lt; 2%                    0.007459                  0.002828

                       L1.Dechloromonas  L1.Haliscomenobacter  \
Acidimicrobium                -0.028836             -0.000381
Acinetobacter                 -0.120288             -0.065033
Albidiferax                   -0.010525             -0.007213
Candidatus Microthrix         -0.092386             -0.031745
Chitinophaga pinensis         -0.024099             -0.064282
Dechloromonas                  1.008220             -0.002000
Haliscomenobacter              0.027866              0.972498
Intrasporangium                0.062233              0.011920
Leptospira                     0.040301              0.083109
Other                         -0.006532              0.006801
Xanthomonas                   -0.022118              0.001205
mean abundance &lt; 2%            0.015356             -0.002470

                       L1.Intrasporangium  L1.Leptospira  L1.Other  \
Acidimicrobium                   0.003852      -0.018886 -0.071549
Acinetobacter                   -0.169053      -0.028418  0.082148
Albidiferax                     -0.009145       0.000898 -0.114294
Candidatus Microthrix            0.028549      -0.072750 -0.124845
Chitinophaga pinensis           -0.006870      -0.049551 -0.042289
Dechloromonas                    0.000399       0.001850  0.017999
Haliscomenobacter                0.001219      -0.028812  0.027804
Intrasporangium                  1.013150      -0.001406  0.032484
Leptospira                      -0.013502       1.039789  0.115065
Other                            0.001661       0.010241  0.980034
Xanthomonas                     -0.000700       0.000163 -0.023899
mean abundance &lt; 2%              0.019977       0.026820 -0.100714

                       L1.Xanthomonas  L1.mean abundance &lt; 2%
Acidimicrobium               0.053300               -0.003768
Acinetobacter               -0.351822               -0.037516
Albidiferax                  0.010249               -0.012573
Candidatus Microthrix        0.207496               -0.046151
Chitinophaga pinensis        0.168200               -0.053289
Dechloromonas                0.016044               -0.008142
Haliscomenobacter           -0.088561                0.000027
Intrasporangium              0.056446               -0.010126
Leptospira                  -0.055608               -0.013448
Other                        0.007062                0.004578
Xanthomonas                  0.994634                0.006090
mean abundance &lt; 2%          0.019977                1.019772
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">,</span>
                           <span class="n">coefficients</span><span class="o">=</span><span class="n">coefficients</span><span class="p">,</span> <span class="n">intercepts</span><span class="o">=</span><span class="n">intercepts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 09:36<00:00 Sampling 2 chains, 3,838 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 580 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.501  0.129   0.372    0.630      0.091    0.077       2.0
Ah[0, 1]   -0.077  0.110  -0.187    0.034      0.078    0.066       2.0
Ah[0, 2]    0.086  0.170  -0.084    0.255      0.120    0.101       3.0
Ah[0, 3]    0.068  0.040   0.028    0.108      0.028    0.024       2.0
Ah[0, 4]    0.014  0.096  -0.082    0.110      0.068    0.057       3.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]  -0.314  0.275  -0.589   -0.039      0.194    0.165       3.0
Bh[20, 8]   0.013  0.065  -0.052    0.078      0.046    0.039       3.0
Bh[20, 9]   0.486  0.002   0.484    0.488      0.002    0.001       3.0
Bh[20, 10]  0.218  0.025   0.193    0.243      0.018    0.015       3.0
Bh[20, 11]  0.352  0.256   0.096    0.608      0.181    0.153       2.0

            ess_tail  r_hat
Ah[0, 0]        13.0   2.42
Ah[0, 1]        25.0   2.44
Ah[0, 2]        13.0   2.27
Ah[0, 3]        11.0   2.60
Ah[0, 4]        22.0   1.91
...              ...    ...
Bh[20, 7]       14.0   2.18
Bh[20, 8]       11.0   2.23
Bh[20, 9]       24.0   2.24
Bh[20, 10]      22.0   1.97
Bh[20, 11]      31.0   2.85

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_xs_v9.nc
Data file: data_xs_v9.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_64_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_64_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_65_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_65_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_65_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_65_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[59]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior2</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">,</span>
                            <span class="n">coefficients</span><span class="o">=</span><span class="n">coefficients</span><span class="p">,</span> <span class="n">intercepts</span><span class="o">=</span><span class="n">intercepts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior2</span><span class="o">.</span><span class="n">run_inference_large_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [c2_A, tau_A, lam_A, Ah, c2_B, tau_B, lam_B, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='24000' class='' max='24000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [24000/24000 1:28:42<00:00 Sampling 4 chains, 0 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 4 chains for 2_000 tune and 4_000 draw iterations (8_000 + 16_000 draws total) took 5330 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Chain 0 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 1 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 2 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 3 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.254  0.212   0.074    0.604      0.106    0.081       4.0
Ah[0, 1]    0.031  0.050  -0.058    0.068      0.025    0.019       4.0
Ah[0, 2]    0.110  0.124  -0.044    0.298      0.062    0.047       4.0
Ah[0, 3]    0.129  0.034   0.088    0.167      0.017    0.013       4.0
Ah[0, 4]   -0.028  0.045  -0.100    0.016      0.023    0.017       4.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.168  0.429  -0.250    0.873      0.215    0.164       5.0
Bh[20, 8]  -0.056  0.191  -0.252    0.226      0.095    0.073       4.0
Bh[20, 9]   0.418  0.176   0.184    0.681      0.088    0.067       5.0
Bh[20, 10]  0.368  0.144   0.182    0.567      0.072    0.055       4.0
Bh[20, 11] -0.139  0.368  -0.575    0.387      0.184    0.141       5.0

            ess_tail  r_hat
Ah[0, 0]        11.0   3.68
Ah[0, 1]        11.0   3.68
Ah[0, 2]        16.0   3.30
Ah[0, 3]        11.0   3.31
Ah[0, 4]        11.0   3.48
...              ...    ...
Bh[20, 7]       26.0   3.00
Bh[20, 8]       14.0   3.39
Bh[20, 9]       22.0   2.87
Bh[20, 10]      12.0   3.98
Bh[20, 11]      18.0   3.14

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_large_xs_v2.nc
Data file: data_large_xs_v2.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_67_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_67_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[61]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior2</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_68_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_68_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_68_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_68_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[62]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">normalVAR</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[63]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">normalVAR</span><span class="o">.</span><span class="n">run_inference</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [x0, A, noise_chol]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 19:59<00:00 Sampling 2 chains, 0 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 1204 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
            mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
x0[0]      0.533  0.498  -0.385    1.472      0.018    0.013     758.0
x0[1]      1.589  0.944  -0.223    3.340      0.021    0.015    1999.0
x0[2]     -0.067  0.446  -0.892    0.768      0.014    0.010     967.0
x0[3]      1.557  0.901  -0.156    3.238      0.023    0.017    1492.0
x0[4]      2.456  0.744   1.042    3.832      0.024    0.017     960.0
...          ...    ...     ...      ...        ...      ...       ...
A[11, 7]   0.014  0.007   0.001    0.027      0.000    0.000    1173.0
A[11, 8]   0.021  0.006   0.009    0.032      0.000    0.000    1033.0
A[11, 9]  -0.110  0.024  -0.155   -0.065      0.000    0.000    2372.0
A[11, 10]  0.019  0.021  -0.018    0.060      0.000    0.000    2155.0
A[11, 11]  1.013  0.010   0.993    1.030      0.000    0.000    1379.0

           ess_tail  r_hat
x0[0]        1369.0    1.0
x0[1]        2340.0    1.0
x0[2]        1470.0    1.0
x0[3]        2057.0    1.0
x0[4]        1346.0    1.0
...             ...    ...
A[11, 7]     2202.0    1.0
A[11, 8]     1953.0    1.0
A[11, 9]     2596.0    1.0
A[11, 10]    2777.0    1.0
A[11, 11]    2322.0    1.0

[156 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (156) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_default.nc
Data file: data_default.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_70_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_70_6.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[65]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">posteriorVAR</span> <span class="o">=</span> <span class="n">az</span><span class="o">.</span><span class="n">from_netcdf</span><span class="p">(</span><span class="s1">&#39;model_posterior_default.nc&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[66]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">coefficients</span> <span class="o">=</span> <span class="n">posteriorVAR</span><span class="o">.</span><span class="n">posterior</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">coefficients_median</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">coefficients</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)))</span>
<span class="n">intercepts</span> <span class="o">=</span> <span class="n">posteriorVAR</span><span class="o">.</span><span class="n">posterior</span><span class="p">[</span><span class="s1">&#39;x0&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">intercepts_median</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">intercepts</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[67]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior2</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">imputed_data</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">,</span>
                            <span class="n">coefficients</span><span class="o">=</span><span class="n">coefficients_median</span><span class="p">,</span> <span class="n">intercepts</span><span class="o">=</span><span class="n">intercepts_median</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[68]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior2</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 24:29<00:00 Sampling 2 chains, 2,000 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 1473 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.037  0.087  -0.050    0.124      0.061    0.052       2.0
Ah[0, 1]    0.032  0.034  -0.002    0.066      0.024    0.020       3.0
Ah[0, 2]    0.137  0.059   0.078    0.197      0.042    0.036       3.0
Ah[0, 3]    0.205  0.069   0.137    0.274      0.048    0.041       3.0
Ah[0, 4]   -0.022  0.016  -0.039   -0.006      0.011    0.010       3.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.002  0.374  -0.372    0.376      0.264    0.223       3.0
Bh[20, 8]   0.237  0.505  -0.267    0.742      0.356    0.302       3.0
Bh[20, 9]   0.186  0.109   0.076    0.296      0.077    0.065       3.0
Bh[20, 10] -0.316  0.240  -0.556   -0.077      0.169    0.143       3.0
Bh[20, 11] -0.051  0.061  -0.112    0.011      0.043    0.037       2.0

            ess_tail  r_hat
Ah[0, 0]        20.0   2.35
Ah[0, 1]        26.0   2.12
Ah[0, 2]        22.0   2.18
Ah[0, 3]        17.0   2.01
Ah[0, 4]        15.0   1.92
...              ...    ...
Bh[20, 7]       21.0   2.25
Bh[20, 8]       18.0   2.14
Bh[20, 9]       11.0   1.92
Bh[20, 10]      12.0   2.21
Bh[20, 11]      13.0   2.69

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_xs_v10.nc
Data file: data_xs_v10.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_74_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_74_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[69]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferWithPrior2</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_75_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_75_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_75_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_75_1.png" />
</div>
</div>
<p>So, these inferences show that there is no easy correlation to infer between the species relative abundances and the metabolite concentrations. This could be due to us trying to infer the causal inference between the relative abundance, and not the change in relative abundance of the species. In the next step, we will try to infer the causal relationships between the change in relative abundance of the species and the metabolite concentrations.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[70]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Let&#39;s calculate the change in relative abundance for each genus over time from the imputed_data dataset</span>

<span class="c1"># Calculate the change in relative abundance for each genus</span>
<span class="n">data_change</span> <span class="o">=</span> <span class="n">imputed_data</span><span class="o">.</span><span class="n">diff</span><span class="p">()</span>

<span class="c1"># Make the first row as 0.0 replacing NaN</span>
<span class="n">data_change</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[71]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChange</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data_change</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">)</span>
<span class="n">inferChange</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='6000' class='' max='6000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [6000/6000 29:12<00:00 Sampling 2 chains, 0 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 1_000 tune and 2_000 draw iterations (2_000 + 4_000 draws total) took 1758 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.535  0.348   0.133    0.922      0.244    0.207       3.0
Ah[0, 1]    0.004  0.005  -0.007    0.011      0.003    0.002       4.0
Ah[0, 2]   -0.199  0.121  -0.375   -0.043      0.082    0.068       3.0
Ah[0, 3]    0.019  0.007   0.007    0.034      0.002    0.001      16.0
Ah[0, 4]   -0.060  0.054  -0.145    0.009      0.036    0.030       3.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]  -0.044  0.678  -1.437    1.555      0.195    0.323      14.0
Bh[20, 8]  -0.058  0.237  -0.497    0.320      0.134    0.106       3.0
Bh[20, 9]  -0.167  0.751  -1.615    1.063      0.177    0.128      18.0
Bh[20, 10] -0.028  0.628  -1.332    0.847      0.373    0.297       4.0
Bh[20, 11] -0.097  0.701  -1.319    1.071      0.410    0.325       3.0

            ess_tail  r_hat
Ah[0, 0]        11.0   1.83
Ah[0, 1]        14.0   1.51
Ah[0, 2]        16.0   1.96
Ah[0, 3]        18.0   1.06
Ah[0, 4]        29.0   2.05
...              ...    ...
Bh[20, 7]       11.0   1.56
Bh[20, 8]       24.0   1.69
Bh[20, 9]       26.0   1.08
Bh[20, 10]      18.0   1.69
Bh[20, 11]      14.0   1.68

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_xs_v11.nc
Data file: data_xs_v11.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_78_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_78_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[72]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChange</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_79_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_79_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_79_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_79_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[73]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">statsmodels.tsa.api</span> <span class="kn">import</span> <span class="n">VAR</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[74]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># make VAR model with the imputed data set and statsmodels library</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">VAR</span><span class="p">(</span><span class="n">data_change</span><span class="p">)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">maxlags</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ml&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning: An unsupported index was provided and will be ignored when e.g. forecasting.
  self._init_dates(dates, freq)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[75]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">coefficients</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">params</span>
<span class="n">covariance_matrix</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">sigma_u</span>
<span class="n">intercepts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">results</span><span class="o">.</span><span class="n">params</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

<span class="c1"># remove the first row of the coefficients matrix</span>
<span class="n">coefficients</span> <span class="o">=</span> <span class="n">coefficients</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
<span class="c1"># transpose the coefficients matrix</span>
<span class="n">coefficients</span> <span class="o">=</span> <span class="n">coefficients</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[76]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data_change</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">,</span>
                                <span class="n">coefficients</span><span class="o">=</span><span class="n">coefficients</span><span class="p">,</span> <span class="n">intercepts</span><span class="o">=</span><span class="n">intercepts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[77]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">(</span><span class="n">tune</span><span class="o">=</span><span class="mi">5000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='14000' class='' max='14000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [14000/14000 1:04:29<00:00 Sampling 2 chains, 0 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 2 chains for 5_000 tune and 2_000 draw iterations (10_000 + 4_000 draws total) took 3873 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.900  0.022   0.861    0.941      0.002    0.001     200.0
Ah[0, 1]    0.011  0.004   0.002    0.018      0.000    0.000     126.0
Ah[0, 2]   -0.086  0.023  -0.127   -0.043      0.002    0.001     148.0
Ah[0, 3]    0.023  0.007   0.009    0.036      0.001    0.001      83.0
Ah[0, 4]    0.019  0.010   0.002    0.038      0.001    0.001     189.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.080  0.273  -0.438    0.584      0.160    0.127       3.0
Bh[20, 8]   0.142  0.246  -0.212    0.596      0.130    0.100       4.0
Bh[20, 9]  -0.073  1.003  -1.921    1.855      0.144    0.103      49.0
Bh[20, 10]  0.122  0.508  -0.704    1.049      0.185    0.136       8.0
Bh[20, 11]  0.207  0.380  -0.427    0.898      0.189    0.145       5.0

            ess_tail  r_hat
Ah[0, 0]       326.0   1.00
Ah[0, 1]       242.0   1.02
Ah[0, 2]       392.0   1.01
Ah[0, 3]       155.0   1.04
Ah[0, 4]       504.0   1.01
...              ...    ...
Bh[20, 7]       23.0   1.79
Bh[20, 8]       40.0   1.39
Bh[20, 9]      118.0   1.03
Bh[20, 10]      49.0   1.24
Bh[20, 11]      58.0   1.34

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_xs_v12.nc
Data file: data_xs_v12.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_84_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_84_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[78]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_85_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_85_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_85_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_85_1.png" />
</div>
</div>
<p>We can also increase the increase the burn-in and number of samples to get more accurate results.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[79]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior2</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data_change</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">,</span>
                                 <span class="n">coefficients</span><span class="o">=</span><span class="n">coefficients</span><span class="p">,</span> <span class="n">intercepts</span><span class="o">=</span><span class="n">intercepts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[80]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior2</span><span class="o">.</span><span class="n">run_inference_xs</span><span class="p">(</span><span class="n">tune</span><span class="o">=</span><span class="mi">8000</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="mi">4000</span><span class="p">,</span> <span class="n">cores</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [X0h, S0h, Ah, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='48000' class='' max='48000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [48000/48000 2:51:05<00:00 Sampling 4 chains, 0 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 4 chains for 8_000 tune and 4_000 draw iterations (32_000 + 16_000 draws total) took 10274 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Chain 0 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 1 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 2 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 3 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.900  0.021   0.859    0.938      0.001    0.001     614.0
Ah[0, 1]    0.010  0.005   0.001    0.019      0.001    0.000      85.0
Ah[0, 2]   -0.087  0.022  -0.128   -0.046      0.001    0.001     693.0
Ah[0, 3]    0.023  0.008   0.008    0.039      0.001    0.001      67.0
Ah[0, 4]    0.020  0.011  -0.001    0.039      0.001    0.000     236.0
...           ...    ...     ...      ...        ...      ...       ...
Bh[20, 7]   0.038  0.262  -0.437    0.618      0.043    0.039      37.0
Bh[20, 8]  -0.013  0.260  -0.547    0.469      0.081    0.059      10.0
Bh[20, 9]  -0.108  1.015  -1.993    1.816      0.122    0.087      70.0
Bh[20, 10] -0.084  0.596  -1.135    1.214      0.197    0.144       9.0
Bh[20, 11] -0.030  0.406  -0.795    0.682      0.132    0.096      10.0

            ess_tail  r_hat
Ah[0, 0]      1050.0   1.00
Ah[0, 1]       474.0   1.05
Ah[0, 2]      1557.0   1.00
Ah[0, 3]       236.0   1.06
Ah[0, 4]       658.0   1.02
...              ...    ...
Bh[20, 7]       49.0   1.24
Bh[20, 8]       31.0   1.30
Bh[20, 9]      131.0   1.06
Bh[20, 10]      23.0   1.37
Bh[20, 11]      28.0   1.33

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_xs_v13.nc
Data file: data_xs_v13.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_88_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_88_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[81]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior2</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_89_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_89_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_89_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_89_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[82]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior3</span> <span class="o">=</span> <span class="n">infer_VAR</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data_change</span><span class="p">,</span> <span class="n">dataS</span><span class="o">=</span><span class="n">imputed_MP_data</span><span class="p">,</span>
                                 <span class="n">coefficients</span><span class="o">=</span><span class="n">coefficients</span><span class="p">,</span> <span class="n">intercepts</span><span class="o">=</span><span class="n">intercepts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[83]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior3</span><span class="o">.</span><span class="n">run_inference_large_xs</span><span class="p">(</span><span class="n">tune</span><span class="o">=</span><span class="mi">5000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [c2_A, tau_A, lam_A, Ah, c2_B, tau_B, lam_B, Bh, sigma]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
    <style>
        /* Turns off some styling */
        progress {
            /* gets rid of default border in Firefox and Opera. */
            border: none;
            /* Needs to be in here for Safari polyfill so background images work as expected. */
            background-size: auto;
        }
        .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
            background: #F44336;
        }
    </style>
  <progress value='36000' class='' max='36000' style='width:300px; height:20px; vertical-align: middle;'></progress>
  100.00% [36000/36000 2:36:11<00:00 Sampling 4 chains, 9,795 divergences]
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Sampling 4 chains for 5_000 tune and 4_000 draw iterations (20_000 + 16_000 draws total) took 9380 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
There were 9795 divergences after tuning. Increase `target_accept` or reparameterize.
Chain 2 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
Chain 3 reached the maximum tree depth. Increase `max_treedepth`, increase `target_accept` or reparameterize.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
             mean   sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
Ah[0, 0]    0.912  0.0   0.912    0.912        0.0      0.0      14.0
Ah[0, 1]    0.006  0.0   0.006    0.006        0.0      0.0       9.0
Ah[0, 2]   -0.084  0.0  -0.084   -0.084        0.0      0.0      21.0
Ah[0, 3]    0.014  0.0   0.014    0.014        0.0      0.0      19.0
Ah[0, 4]    0.016  0.0   0.016    0.016        0.0      0.0      19.0
...           ...  ...     ...      ...        ...      ...       ...
Bh[20, 7]  -0.000  0.0  -0.000    0.000        0.0      0.0      68.0
Bh[20, 8]   0.000  0.0  -0.000    0.000        0.0      0.0      71.0
Bh[20, 9]   0.000  0.0  -0.000    0.000        0.0      0.0     357.0
Bh[20, 10]  0.000  0.0  -0.000    0.000        0.0      0.0      32.0
Bh[20, 11]  0.000  0.0  -0.000    0.000        0.0      0.0      30.0

            ess_tail  r_hat
Ah[0, 0]        41.0   1.21
Ah[0, 1]        33.0   1.77
Ah[0, 2]        17.0   1.68
Ah[0, 3]        78.0   1.57
Ah[0, 4]        79.0   1.52
...              ...    ...
Bh[20, 7]       68.0   1.64
Bh[20, 8]       95.0   1.61
Bh[20, 9]       87.0   1.61
Bh[20, 10]      30.0   1.57
Bh[20, 11]      56.0   1.45

[396 rows x 9 columns]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\arviz\plots\plot_utils.py:271: UserWarning: rcParams[&#39;plot.max_subplots&#39;] (40) is smaller than the number of variables to plot (396) in plot_posterior, generating only 40 plots
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Results saved as:
NetCDF file: model_posterior_large_xs_v3.nc
Data file: data_large_xs_v3.npz
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_91_6.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_91_6.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[84]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inferChageWithPrior3</span><span class="o">.</span><span class="n">posterior_analysis</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_92_0.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_92_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_92_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples-Herold-sVAR_92_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[85]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now we are going to check for stationarity in the data using adf and kpss tests from stasmodels library</span>
<span class="c1"># We will use the imputed data set</span>

<span class="kn">import</span> <span class="nn">statsmodels.api</span> <span class="k">as</span> <span class="nn">sm</span>
<span class="kn">from</span> <span class="nn">statsmodels.tsa.stattools</span> <span class="kn">import</span> <span class="n">adfuller</span><span class="p">,</span> <span class="n">kpss</span>


<span class="k">def</span> <span class="nf">adf_test</span><span class="p">(</span><span class="n">timeseries</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Results of Dickey-Fuller Test:&quot;</span><span class="p">)</span>
    <span class="n">dftest</span> <span class="o">=</span> <span class="n">adfuller</span><span class="p">(</span><span class="n">timeseries</span><span class="p">,</span> <span class="n">autolag</span><span class="o">=</span><span class="s2">&quot;AIC&quot;</span><span class="p">)</span>
    <span class="n">dfoutput</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
        <span class="n">dftest</span><span class="p">[:</span><span class="mi">4</span><span class="p">],</span>
        <span class="n">index</span><span class="o">=</span><span class="p">[</span>
            <span class="s2">&quot;Test Statistic&quot;</span><span class="p">,</span>
            <span class="s2">&quot;p-value&quot;</span><span class="p">,</span>
            <span class="s2">&quot;#Lags Used&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Number of Observations Used&quot;</span><span class="p">,</span>
        <span class="p">],</span>
    <span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">dftest</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">dfoutput</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Critical Value (</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">dfoutput</span><span class="p">)</span>
    <span class="c1"># if p-value is less than 0.05, we reject the null hypothesis and say that the data is stationary</span>
    <span class="c1"># print if the data is stationary or not</span>
    <span class="k">if</span> <span class="n">dftest</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mf">0.05</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ADF Test: Data is stationary&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ADF Test: Data is not stationary&quot;</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">kpss_test</span><span class="p">(</span><span class="n">timeseries</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Results of KPSS Test:&quot;</span><span class="p">)</span>
    <span class="n">kpsstest</span> <span class="o">=</span> <span class="n">kpss</span><span class="p">(</span><span class="n">timeseries</span><span class="p">,</span> <span class="n">regression</span><span class="o">=</span><span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="n">nlags</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">)</span>
    <span class="n">kpss_output</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
        <span class="n">kpsstest</span><span class="p">[:</span><span class="mi">3</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Test Statistic&quot;</span><span class="p">,</span> <span class="s2">&quot;p-value&quot;</span><span class="p">,</span> <span class="s2">&quot;Lags Used&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">kpsstest</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">kpss_output</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Critical Value (</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">kpss_output</span><span class="p">)</span>
    <span class="c1"># if p-value is greater than 0.05, we reject the null hypothesis and say that the data is stationary</span>
    <span class="c1"># print if the data is stationary or not</span>
    <span class="k">if</span> <span class="n">kpsstest</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mf">0.05</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;KPSS test: Data is stationary&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;KPSS test: Data is not stationary&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[86]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check for stationarity in the data imputed_dataset (without the &#39;days&#39; column)</span>
<span class="k">for</span> <span class="n">genus</span> <span class="ow">in</span> <span class="n">data_change</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">genus</span><span class="p">)</span>
    <span class="n">adf_test</span><span class="p">(</span><span class="n">data_change</span><span class="p">[</span><span class="n">genus</span><span class="p">])</span>
    <span class="n">kpss_test</span><span class="p">(</span><span class="n">data_change</span><span class="p">[</span><span class="n">genus</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Acidimicrobium
Results of Dickey-Fuller Test:
Test Statistic                  -4.877743
p-value                          0.000039
#Lags Used                      16.000000
Number of Observations Used    393.000000
Critical Value (1%)             -3.447099
Critical Value (5%)             -2.868923
Critical Value (10%)            -2.570703
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.032404
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Acinetobacter
Results of Dickey-Fuller Test:
Test Statistic                  -4.598946
p-value                          0.000130
#Lags Used                      17.000000
Number of Observations Used    392.000000
Critical Value (1%)             -3.447142
Critical Value (5%)             -2.868941
Critical Value (10%)            -2.570713
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.173686
p-value                   0.100000
Lags Used                10.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Albidiferax
Results of Dickey-Fuller Test:
Test Statistic                  -3.829655
p-value                          0.002618
#Lags Used                      11.000000
Number of Observations Used    398.000000
Critical Value (1%)             -3.446888
Critical Value (5%)             -2.868829
Critical Value (10%)            -2.570653
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.043204
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Candidatus Microthrix
Results of Dickey-Fuller Test:
Test Statistic                -7.937074e+00
p-value                        3.391025e-12
#Lags Used                     6.000000e+00
Number of Observations Used    4.030000e+02
Critical Value (1%)           -3.446681e+00
Critical Value (5%)           -2.868739e+00
Critical Value (10%)          -2.570605e+00
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.094575
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Chitinophaga pinensis
Results of Dickey-Fuller Test:
Test Statistic                  -3.816479
p-value                          0.002743
#Lags Used                      16.000000
Number of Observations Used    393.000000
Critical Value (1%)             -3.447099
Critical Value (5%)             -2.868923
Critical Value (10%)            -2.570703
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.059832
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Dechloromonas
Results of Dickey-Fuller Test:
Test Statistic                  -3.320016
p-value                          0.014007
#Lags Used                      15.000000
Number of Observations Used    394.000000
Critical Value (1%)             -3.447057
Critical Value (5%)             -2.868904
Critical Value (10%)            -2.570693
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.106131
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Haliscomenobacter
Results of Dickey-Fuller Test:
Test Statistic                  -3.790187
p-value                          0.003009
#Lags Used                      17.000000
Number of Observations Used    392.000000
Critical Value (1%)             -3.447142
Critical Value (5%)             -2.868941
Critical Value (10%)            -2.570713
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.099982
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Intrasporangium
Results of Dickey-Fuller Test:
Test Statistic                  -4.249279
p-value                          0.000543
#Lags Used                      18.000000
Number of Observations Used    391.000000
Critical Value (1%)             -3.447186
Critical Value (5%)             -2.868960
Critical Value (10%)            -2.570723
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.197429
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Leptospira
Results of Dickey-Fuller Test:
Test Statistic                  -3.825629
p-value                          0.002656
#Lags Used                      17.000000
Number of Observations Used    392.000000
Critical Value (1%)             -3.447142
Critical Value (5%)             -2.868941
Critical Value (10%)            -2.570713
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.039661
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Other
Results of Dickey-Fuller Test:
Test Statistic                -7.502466e+00
p-value                        4.218178e-11
#Lags Used                     6.000000e+00
Number of Observations Used    4.030000e+02
Critical Value (1%)           -3.446681e+00
Critical Value (5%)           -2.868739e+00
Critical Value (10%)          -2.570605e+00
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.075135
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


Xanthomonas
Results of Dickey-Fuller Test:
Test Statistic                  -3.684859
p-value                          0.004331
#Lags Used                      18.000000
Number of Observations Used    391.000000
Critical Value (1%)             -3.447186
Critical Value (5%)             -2.868960
Critical Value (10%)            -2.570723
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.1455
p-value                   0.1000
Lags Used                11.0000
Critical Value (10%)      0.3470
Critical Value (5%)       0.4630
Critical Value (2.5%)     0.5740
Critical Value (1%)       0.7390
dtype: float64
KPSS test: Data is stationary


mean abundance &lt; 2%
Results of Dickey-Fuller Test:
Test Statistic                  -4.761013
p-value                          0.000065
#Lags Used                       7.000000
Number of Observations Used    402.000000
Critical Value (1%)             -3.446722
Critical Value (5%)             -2.868757
Critical Value (10%)            -2.570614
dtype: float64
ADF Test: Data is stationary
Results of KPSS Test:
Test Statistic            0.189282
p-value                   0.100000
Lags Used                11.000000
Critical Value (10%)      0.347000
Critical Value (5%)       0.463000
Critical Value (2.5%)     0.574000
Critical Value (1%)       0.739000
dtype: float64
KPSS test: Data is stationary


</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\User\AppData\Local\Temp\ipykernel_41148\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_41148\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_41148\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_41148\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_41148\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_41148\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_41148\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_41148\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_41148\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_41148\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_41148\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
C:\Users\User\AppData\Local\Temp\ipykernel_41148\3039934652.py:33: InterpolationWarning: The test statistic is outside of the range of p-values available in the
look-up table. The actual p-value is greater than the p-value returned.

  kpsstest = kpss(timeseries, regression=&#34;c&#34;, nlags=&#34;auto&#34;)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[87]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[89]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import the matrix A from the model</span>

<span class="n">posterior_samples</span> <span class="o">=</span> <span class="n">az</span><span class="o">.</span><span class="n">from_netcdf</span><span class="p">(</span><span class="s1">&#39;model_posterior_large_xs_v3.nc&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">posterior</span><span class="p">[</span><span class="s1">&#39;Ah&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[90]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute the median of the posterior samples</span>
<span class="n">median_Ah</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">posterior_samples</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[91]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to check stability of a single matrix</span>
<span class="k">def</span> <span class="nf">is_stable</span><span class="p">(</span><span class="n">matrix</span><span class="p">):</span>
    <span class="n">eigenvalues</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">eigvals</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
    <span class="c1"># return all(np.sqrt(np.real(eigenvalue)**2 + np.imag(eigenvalue)**2) &lt; 1 for eigenvalue in eigenvalues)</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">eigenvalues</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">)</span>


<span class="c1"># Check stability of the median matrix</span>
<span class="n">stability_median_Ah</span> <span class="o">=</span> <span class="n">is_stable</span><span class="p">(</span><span class="n">median_Ah</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Median Ah matrix stability:&quot;</span><span class="p">,</span> <span class="n">stability_median_Ah</span><span class="p">)</span>
<span class="k">if</span> <span class="n">stability_median_Ah</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The median Ah matrix is stable.&quot;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The median Ah matrix is not stable.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Median Ah matrix stability: True
The median Ah matrix is stable.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[92]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to check stability of a single matrix</span>
<span class="k">def</span> <span class="nf">is_stable</span><span class="p">(</span><span class="n">matrix</span><span class="p">):</span>
    <span class="n">eigenvalues</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">eigvals</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">eigenvalues</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">)</span>


<span class="c1"># Check stability for each sampled matrix</span>
<span class="n">num_chains</span><span class="p">,</span> <span class="n">num_samples</span><span class="p">,</span> <span class="n">nX</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">posterior_samples</span><span class="o">.</span><span class="n">shape</span>
<span class="n">stability_results</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">num_chains</span><span class="p">,</span> <span class="n">num_samples</span><span class="p">))</span>

<span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_chains</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_samples</span><span class="p">):</span>
        <span class="n">stability_results</span><span class="p">[</span><span class="n">chain</span><span class="p">,</span> <span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">is_stable</span><span class="p">(</span>
            <span class="n">posterior_samples</span><span class="p">[</span><span class="n">chain</span><span class="p">,</span> <span class="n">sample</span><span class="p">])</span>

<span class="c1"># Percentage of stable samples</span>
<span class="n">stable_percentage</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">stability_results</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Percentage of stable samples: </span><span class="si">{</span><span class="n">stable_percentage</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>

<span class="k">if</span> <span class="n">stable_percentage</span> <span class="o">&lt;</span> <span class="mi">100</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: Not all samples are stable.&quot;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;All samples are stable.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Percentage of stable samples: 100.00%
All samples are stable.
</pre></div></div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">MIMIC</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../readme.html">Modelling and Inference of MICrobiomes Project (MIMIC)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">mimic</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../../examples.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../MVAR/examples-sim-MVAR.html">MVAR Simulation for Microbiota and Metabolites</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../MVAR/examples-infer-MVAR.html">MVAR Simulation for Microbiota and Metabolites</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gLV/examples-bayes-gLV.html">Used Bayesian inference to infer the parameters of a (linearised) gLV model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gLV/examples-sim-gLV.html">Simulate some time course data from the gLV</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gLV/examples-ridge-gLV.html">Simulate some time course data and perform ridge regression as in Stein et al. 2013</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gLV/examples-Rutter-Dekker.html">Repeat Rutter &amp; Dekker et al 2024 analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gLV/examples-Rutter-Dekker.html#Perform-Bayesian-inference-with-shrinkage">Perform Bayesian inference with shrinkage</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gLV/examples-Stein.html">Repeat Stein et al. 2013 analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../GP/examples-impute-GP.html">Imputing Missing Values using Gaussian Process</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../GP/examples-impute-GP_Stein.html">Using the Data Imputator on a Real Public Dataset - Stein et al. (2013)</a></li>
<li class="toctree-l2"><a class="reference internal" href="examples-Herold-VAR.html">VAR inference on a Real Public Dataset, Herold et al. (2020)</a></li>
<li class="toctree-l2"><a class="reference internal" href="examples-Herold-VAR.html#Separate-the-data-into-MG-and-MT-datasets">Separate the data into MG and MT datasets</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Sparse VAR on a Real Public Dataset, Herold et al. (2020)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Separate-the-data-into-MG-and-MT-datasets">Separate the data into MG and MT datasets</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Reducing-number-of-variables-(to-the-3-or-5-most-abundant)-to-reduce-the-computational-cost-and-simplify-the-analysis">Reducing number of variables (to the 3 or 5 most abundant) to reduce the computational cost and simplify the analysis</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="examples_impute_data.html">Data imputation on a Real Public Dataset, Herold et al. (2020)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../VAR/examples-bayes-VAR.html">Bayesian Inference for Vector AutoRegression (VAR) Models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../VAR/examples-sim-VAR.html">Vector AutoRegression (VAR) Simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gMLV/examples-sim-gMLV.html">Simulate some time course data from the gMLV</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gMLV/examples-ridge-lasso-gMLV.html">Run ridge regression and lasso for the gMLV model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../contributing.html">Contributing to MIMIC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../history.html">History</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../../examples.html">Examples</a><ul>
      <li>Previous: <a href="examples-Herold-VAR.html" title="previous chapter">VAR inference on a Real Public Dataset, Herold et al. (2020)</a></li>
      <li>Next: <a href="examples_impute_data.html" title="next chapter">Data imputation on a Real Public Dataset, Herold et al. (2020)</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2024, Pedro Fontanarrosa.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.4.7</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
      |
      <a href="../../../_sources/notebooks/MultiModel/Herold/examples-Herold-sVAR.ipynb.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>