<!DOCTYPE html>

<html lang="en" data-content_root="../../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Data imputation on a Real Public Dataset, Herold et al. (2020) &#8212; MIMIC 0.1.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="../../../_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="../../../_static/alabaster.css?v=27fed22d" />
    <link rel="stylesheet" type="text/css" href="../../../_static/nbsphinx-code-cells.css?v=2aa19091" />
    <script src="../../../_static/documentation_options.js?v=01f34227"></script>
    <script src="../../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Bayesian Inference for Vector AutoRegression (VAR) Models" href="../../VAR/examples-bayes-VAR.html" />
    <link rel="prev" title="Sparse VAR on a Real Public Dataset, Herold et al. (2020)" href="examples-Herold-sVAR.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="Data-imputation-on-a-Real-Public-Dataset,-Herold-et-al.-(2020)">
<h1>Data imputation on a Real Public Dataset, Herold et al. (2020)<a class="headerlink" href="#Data-imputation-on-a-Real-Public-Dataset,-Herold-et-al.-(2020)" title="Link to this heading">¶</a></h1>
<p>The article “Integration of time-series meta-omics data reveals how microbial ecosystems respond to disturbance” explores the dynamic nature of microbial communities in biological wastewater treatment plants (BWWTPs) by combining metagenomics (MG), metatranscriptomics (MT), and other meta-omics approaches. The study provides a detailed analysis of these communities’ resilience and resistance to environmental disturbances. We are going to use the data from this study to perform data imputation
and analysis. Specifically, we will focus on the following figures:</p>
<section id="Figure-3---Community-Structure-and-Function-Dynamics:">
<h2>Figure 3 - Community Structure and Function Dynamics:<a class="headerlink" href="#Figure-3---Community-Structure-and-Function-Dynamics:" title="Link to this heading">¶</a></h2>
<p>Figure 3 presents the community structure and function dynamics over time, focusing on the relative abundance and expression levels of microbial populations. The data are derived from recovered populations represented by reconstructed metagenome-assembled genomes (rMAGs), which show genetic composition and activity through MG and MT depth coverage, with mapping percentages of 26% ± 3% and 27% ± 3%, respectively. The relative abundance data are categorized by genus-level taxonomic assignments,
and less abundant genera are grouped under “Other” if their mean abundance is below 2%.</p>
<p>This figure also includes ordination plots based on Bray–Curtis dissimilarity of rMAGs’ relative abundances, constrained by abiotic factors such as metabolite levels, metabolite ratios, and physicochemical parameters (shown as black arrows). The results highlight seasonal shifts and environmental influences on microbial community dynamics, with time-course sampling illustrated through connecting points and color coding by the month of sampling.</p>
</section>
<section id="Figure-4---Metabolite-Levels-and-Environmental-Dynamics:">
<h2>Figure 4 - Metabolite Levels and Environmental Dynamics:<a class="headerlink" href="#Figure-4---Metabolite-Levels-and-Environmental-Dynamics:" title="Link to this heading">¶</a></h2>
<p>Figure 4 presents time-series data on metabolite levels and physicochemical parameters, highlighting fluctuations in metabolite classes, types, and measurement subtypes. It uses a heatmap to display Z-score transformed intensities of metabolites and other parameters, showcasing intracellular and extracellular metabolite levels, ratios, and several environmental parameters collected manually or through continuous monitoring at a wastewater treatment plant. This data reflects how environmental and
metabolite dynamics interact, impacting microbial activity and metabolism over time (Herold et al., 2020).</p>
<p>By focusing on these figures, our VAR (Vector Autoregression) inference and Gaussian Process imputer (GP Imputer) can leverage time-series trends and correlations between microbial populations and their environment within BWWTP systems.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>


<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>


<span class="kn">from</span><span class="w"> </span><span class="nn">mimic.data_imputation.impute_GP</span><span class="w"> </span><span class="kn">import</span> <span class="n">GPImputer</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">mimic.model_infer.infer_VAR_bayes</span><span class="w"> </span><span class="kn">import</span> <span class="o">*</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING:tensorflow:From c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\tf_keras\src\losses.py:2976: The name tf.losses.sparse_softmax_cross_entropy is deprecated. Please use tf.compat.v1.losses.sparse_softmax_cross_entropy instead.

WARNING:tensorflow:From c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\tensorflow_probability\python\internal\backend\numpy\_utils.py:48: The name tf.logging.TaskLevelStatusMessage is deprecated. Please use tf.compat.v1.logging.TaskLevelStatusMessage instead.

WARNING:tensorflow:From c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\tensorflow_probability\python\internal\backend\numpy\_utils.py:48: The name tf.control_flow_v2_enabled is deprecated. Please use tf.compat.v1.control_flow_v2_enabled instead.

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
WARNING (pytensor.tensor.blas): Using NumPy C-API based implementation for BLAS functions.
</pre></div></div>
</div>
<section id="Imputing-Microbial-Community-Data-(Figure-3)-using-Gaussian-Process-Imputer">
<h3>Imputing Microbial Community Data (Figure 3) using Gaussian Process Imputer<a class="headerlink" href="#Imputing-Microbial-Community-Data-(Figure-3)-using-Gaussian-Process-Imputer" title="Link to this heading">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Source Data/fig3ab.tsv&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<section id="Separate-the-data-into-MG-and-MT-datasets">
<h4>Separate the data into MG and MT datasets<a class="headerlink" href="#Separate-the-data-into-MG-and-MT-datasets" title="Link to this heading">¶</a></h4>
<p>The data is separated into two datasets, MG and MT. The MG dataset contains the meta-genomics data, while the MT dataset contains the meta-transcriptomics. We will be using the MG dataset in this example.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># separate the data into MG and MT datasets</span>

<span class="n">dataMG</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;MG&#39;</span><span class="p">]</span>

<span class="n">dataMG</span><span class="o">.</span><span class="n">info</span><span class="p">()</span>

<span class="n">dataMT</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;MT&#39;</span><span class="p">]</span>

<span class="n">dataMT</span><span class="o">.</span><span class="n">info</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>We will proceed with some data wrangling to prepare the data for the imputation process.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Data cleaning</span>
<span class="c1"># Rename Nan in the genus column to &#39;Other&#39;</span>

<span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="s1">&#39;Other&#39;</span><span class="p">)</span>

<span class="c1"># Drop the &#39;type&#39; column</span>
<span class="n">dataMG</span> <span class="o">=</span> <span class="n">dataMG</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">])</span>

<span class="c1"># Change relative Date columns to days</span>
<span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_datetime</span><span class="p">(</span><span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">days</span>

<span class="c1"># Change the &#39;genus&#39; column to a categorical column</span>
<span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>


<span class="n">dataMG</span><span class="o">.</span><span class="n">info</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the data, relative abundance (value) vs. days (date) for each genus</span>


<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataMG</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Microbial Growth Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Pivot the data so each &#39;genus&#39; is a column</span>
<span class="n">dataMG_pivot</span> <span class="o">=</span> <span class="n">dataMG</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Create a stacked area chart</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">stackplot</span><span class="p">(</span><span class="n">dataMG_pivot</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">dataMG_pivot</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">dataMG_pivot</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Microbial Growth Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Do the same data cleaning for the MT data set</span>

<span class="c1"># Rename Nan in the genus column to &#39;Other&#39;</span>

<span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="s1">&#39;Other&#39;</span><span class="p">)</span>

<span class="c1"># Drop the &#39;type&#39; column</span>
<span class="n">dataMT</span> <span class="o">=</span> <span class="n">dataMT</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">])</span>

<span class="c1"># Change relative Date columns to days</span>
<span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_datetime</span><span class="p">(</span><span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">days</span>

<span class="c1"># Change the &#39;genus&#39; column to a categorical column</span>
<span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataMT</span><span class="p">[</span><span class="s1">&#39;genus&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the data, relative abundance (value) vs. days (date) for each genus</span>


<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataMT</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Microbial Meta Transcriptomics Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Pivot the data so each &#39;genus&#39; is a column</span>
<span class="n">dataMT_pivot</span> <span class="o">=</span> <span class="n">dataMT</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Pivot the MG data so each &#39;genus&#39; is a column</span>
<span class="n">dataMG_pivot</span> <span class="o">=</span> <span class="n">dataMG</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;val&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="c1"># Create a DataFrame with a complete range of days</span>
<span class="n">all_days</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span><span class="s1">&#39;days&#39;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">dataMG</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)})</span>

<span class="c1"># Reindex the pivoted DataFrame to include all days, filling missing values with None</span>
<span class="n">complete_dataMG</span> <span class="o">=</span> <span class="n">dataMG_pivot</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">all_days</span><span class="p">[</span><span class="s1">&#39;days&#39;</span><span class="p">])</span>

<span class="n">complete_dataMG</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">complete_dataMG</span><span class="o">.</span><span class="n">columns</span>
</pre></div>
</div>
</div>
<p>If we pick one column, say ‘Acinetobacter’, we can see that there are missing values in the dataset. We will use the GP imputator to fill in the missing values.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">complete_dataMG</span><span class="p">[</span><span class="s1">&#39;Acinetobacter&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reset the index</span>
<span class="n">complete_dataMG_reset</span> <span class="o">=</span> <span class="n">complete_dataMG</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

<span class="c1"># Make index as float</span>
<span class="n">complete_dataMG_reset</span><span class="p">[</span><span class="s1">&#39;days&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">complete_dataMG_reset</span><span class="p">[</span><span class="s1">&#39;days&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now &#39;days&#39; is a regular column and you can use it as a feature</span>
<span class="n">new_dataset</span> <span class="o">=</span> <span class="n">GPImputer</span><span class="p">()</span><span class="o">.</span><span class="n">impute_missing_values</span><span class="p">(</span>
    <span class="n">complete_dataMG_reset</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;days&#39;</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;Acidimicrobium&#39;</span><span class="p">,</span>
                                      <span class="s1">&#39;Acinetobacter&#39;</span><span class="p">,</span> <span class="s1">&#39;Albidiferax&#39;</span><span class="p">],</span> <span class="s1">&#39;Acinetobacter&#39;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">new_dataset</span><span class="p">[</span><span class="s1">&#39;Acinetobacter&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>We can see now that the data imputator calculated the missing values for the ‘Acinetobacter’ column. However, as we can see fro the graph, the imputed values are not great. This can be caused by multiple factors, so we are going to explore different methods to better impute the missing values.</p>
<p>We are first going to try to do imputation using VGP imputation. This is a variational Gaussian process imputation method that is more robust to missing data and uses multiple imputs to calculate the missing values. However, as we will see, this method creates an imputation that is heavily biased by other imputs, so it doesn’t work in our case to determine the missing values of the ‘Acinetobacter’ column.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now &#39;days&#39; is a regular column and you can use it as a feature</span>
<span class="n">new_dataset</span> <span class="o">=</span> <span class="n">GPImputer</span><span class="p">()</span><span class="o">.</span><span class="n">impute_missing_values</span><span class="p">(</span>
    <span class="n">complete_dataMG_reset</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;days&#39;</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;Acidimicrobium&#39;</span><span class="p">,</span> <span class="s1">&#39;Acinetobacter&#39;</span><span class="p">,</span> <span class="s1">&#39;Albidiferax&#39;</span><span class="p">,</span>
                                      <span class="s1">&#39;Candidatus Microthrix&#39;</span><span class="p">,</span> <span class="s1">&#39;Chitinophaga pinensis&#39;</span><span class="p">,</span>
                                      <span class="s1">&#39;Dechloromonas&#39;</span><span class="p">,</span> <span class="s1">&#39;Haliscomenobacter&#39;</span><span class="p">,</span> <span class="s1">&#39;Intrasporangium&#39;</span><span class="p">,</span>
                                      <span class="s1">&#39;Leptospira&#39;</span><span class="p">,</span> <span class="s1">&#39;Other&#39;</span><span class="p">,</span> <span class="s1">&#39;Xanthomonas&#39;</span><span class="p">,</span> <span class="s1">&#39;mean abundance &lt; 2%&#39;</span><span class="p">],</span> <span class="s1">&#39;Acinetobacter&#39;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>If we, isntead, do a single imputation using only a GPR model with a single imput (‘Acinetobacter’), we can see that the imputation is much better. This is because the imputation is not biased by other imputs, and the model can better predict the missing values.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now &#39;days&#39; is a regular column and you can use it as a feature</span>
<span class="n">new_dataset2</span> <span class="o">=</span> <span class="n">GPImputer</span><span class="p">()</span><span class="o">.</span><span class="n">impute_missing_values</span><span class="p">(</span>
    <span class="n">complete_dataMG_reset</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;days&#39;</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;Acinetobacter&#39;</span><span class="p">],</span> <span class="s1">&#39;Acinetobacter&#39;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Save the dataset (new_dataset2) to a csv file</span>
<span class="n">new_dataset2</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;imputed_data.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>So we will proceed to do a single imput GPR model for all the columns in the dataset. We will then use the VAR inference method to infer the causal relationships between the variables in the time series.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># do the imputation for each column (genus) in the dataset</span>
<span class="c1"># Copy the original dataset</span>
<span class="n">new_dataset_complete</span> <span class="o">=</span> <span class="n">complete_dataMG_reset</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="k">for</span> <span class="n">genus</span> <span class="ow">in</span> <span class="n">complete_dataMG</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="c1"># Impute the missing values for the current column</span>
    <span class="n">imputed_data</span> <span class="o">=</span> <span class="n">GPImputer</span><span class="p">()</span><span class="o">.</span><span class="n">impute_missing_values</span><span class="p">(</span>
        <span class="n">complete_dataMG_reset</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;days&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">genus</span><span class="p">],</span> <span class="n">genus</span>
    <span class="p">)</span>

    <span class="c1"># Update the column in the new dataset with the imputed data</span>
    <span class="n">new_dataset_complete</span><span class="p">[</span><span class="n">genus</span><span class="p">]</span> <span class="o">=</span> <span class="n">imputed_data</span><span class="p">[</span><span class="n">genus</span><span class="p">]</span>

<span class="c1"># Save the dataset (new_dataset_complete) to a csv file</span>
<span class="n">new_dataset_complete</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;imputed_data.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="Imputing-Metabolite-Data-(Figure-4)-using-Gaussian-Process-Imputer">
<h3>Imputing Metabolite Data (Figure 4) using Gaussian Process Imputer<a class="headerlink" href="#Imputing-Metabolite-Data-(Figure-4)-using-Gaussian-Process-Imputer" title="Link to this heading">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metabolites</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Source Data/fig4.tsv&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>This dataset contains several columns, each representing different attributes related to microbial activity and metabolite concentrations. Here is a detailed explanation of each column:</p>
<ul class="simple">
<li><p><strong>date</strong>: The date when the sample was collected.</p></li>
<li><p><strong>L1</strong>: The functional category or pathway associated with the data point.</p></li>
<li><p><strong>mp_tot</strong>: Total metaproteomic count for that date and pathway.</p></li>
<li><p><strong>mt_tot</strong>: Total metatranscriptomic count for that date and pathway.</p></li>
<li><p><strong>L1sum_MP</strong>: Sum of metaproteomic counts for that specific functional category (L1).</p></li>
<li><p><strong>L1sum_MT</strong>: Sum of metatranscriptomic counts for that specific functional category (L1).</p></li>
<li><p><strong>L1rel_MP</strong>: Relative abundance of metaproteomic counts within the functional category.</p></li>
<li><p><strong>L1rel_MT</strong>: Relative abundance of metatranscriptomic counts within the functional category.</p></li>
<li><p><strong>MP_std_L1</strong>: Standard deviation of metaproteomic counts for that functional category.</p></li>
<li><p><strong>MT_std_L1</strong>: Standard deviation of metatranscriptomic counts for that functional category.</p></li>
<li><p><strong>cor</strong>: Correlation value between metaproteomic and metatranscriptomic data for that pathway.</p></li>
<li><p><strong>newL1</strong>: A reformatted version of the functional category name, often including a correlation coefficient (r value) for easier readability in the figure.</p></li>
</ul>
<p>We will be using the <code class="docutils literal notranslate"><span class="pre">L1rel_MP</span></code> and <code class="docutils literal notranslate"><span class="pre">L1rel_MT</span></code> columns in this example to use it with the VAR inference method to infer the causal relationships between the variables in the time series with the meta transcriptomics (MT) and meta genomics (MG) data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Letr&#39;s create 2 dataframes, one for the L1rel_MP and the other for the L1rel_MT data, using the date, L1 and L2tel values</span>

<span class="n">dataL1_MP</span> <span class="o">=</span> <span class="n">metabolites</span><span class="p">[[</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="s1">&#39;L1rel_MP&#39;</span><span class="p">]]</span>

<span class="n">dataL1_MT</span> <span class="o">=</span> <span class="n">metabolites</span><span class="p">[[</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="s1">&#39;L1rel_MT&#39;</span><span class="p">]]</span>
<span class="n">dataL1_MP</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>date</th>
      <th>L1</th>
      <th>L1rel_MP</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2011-03-21</td>
      <td>01_Fermentation</td>
      <td>0.096899</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2011-03-21</td>
      <td>02_Homoacetogenesis</td>
      <td>0.009890</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2011-03-21</td>
      <td>03_Superpathway of thiosulfate metabolism (Des...</td>
      <td>0.000579</td>
    </tr>
    <tr>
      <th>3</th>
      <td>2011-03-21</td>
      <td>04_Utililization of sugar, conversion of pento...</td>
      <td>0.000535</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2011-03-21</td>
      <td>05_Fatty acid oxidation</td>
      <td>0.025483</td>
    </tr>
    <tr>
      <th>5</th>
      <td>2011-03-21</td>
      <td>06_Amino acid utilization biosynthesis metabolism</td>
      <td>0.122784</td>
    </tr>
    <tr>
      <th>6</th>
      <td>2011-03-21</td>
      <td>07_Nucleic acid metabolism</td>
      <td>0.119531</td>
    </tr>
    <tr>
      <th>7</th>
      <td>2011-03-21</td>
      <td>08_Hydrocarbon degradation</td>
      <td>0.005524</td>
    </tr>
    <tr>
      <th>8</th>
      <td>2011-03-21</td>
      <td>09_Carbohydrate Active enzyme - CAZy</td>
      <td>0.003921</td>
    </tr>
    <tr>
      <th>9</th>
      <td>2011-03-21</td>
      <td>10_TCA cycle</td>
      <td>0.045398</td>
    </tr>
    <tr>
      <th>10</th>
      <td>2011-03-21</td>
      <td>11_Nitrogen cycle</td>
      <td>0.026063</td>
    </tr>
    <tr>
      <th>11</th>
      <td>2011-03-21</td>
      <td>12_Transporters</td>
      <td>0.015593</td>
    </tr>
    <tr>
      <th>12</th>
      <td>2011-03-21</td>
      <td>13_Hydrogen metabolism</td>
      <td>0.000846</td>
    </tr>
    <tr>
      <th>13</th>
      <td>2011-03-21</td>
      <td>14_Methanogenesis</td>
      <td>0.003520</td>
    </tr>
    <tr>
      <th>14</th>
      <td>2011-03-21</td>
      <td>15_Methylotrophy</td>
      <td>0.009267</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Now let&#39;s visualize the data</span>

<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;L1rel_MP&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataL1_MP</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Metaproteomic Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_33_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_33_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Same for the metatranscriptomic data</span>

<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;L1rel_MT&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataL1_MT</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Metatranscriptomic Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_34_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_34_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># And do the same as we did for the MG and MT data sets</span>

<span class="c1"># Rename Nan in the L1 column to &#39;Other&#39;</span>

<span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="s1">&#39;Other&#39;</span><span class="p">)</span>

<span class="c1"># Change relative Date columns to days</span>
<span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_datetime</span><span class="p">(</span><span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">days</span>

<span class="c1"># Change the &#39;L1&#39; column to a categorical column</span>
<span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataL1_MP</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\User\AppData\Local\Temp\ipykernel_32172\300627487.py:5: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MP[&#39;L1&#39;] = dataL1_MP[&#39;L1&#39;].replace(np.nan, &#39;Other&#39;)
C:\Users\User\AppData\Local\Temp\ipykernel_32172\300627487.py:8: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MP[&#39;date&#39;] = pd.to_datetime(dataL1_MP[&#39;date&#39;])
C:\Users\User\AppData\Local\Temp\ipykernel_32172\300627487.py:9: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MP[&#39;date&#39;] = (dataL1_MP[&#39;date&#39;] - dataL1_MP[&#39;date&#39;].min()).dt.days
C:\Users\User\AppData\Local\Temp\ipykernel_32172\300627487.py:12: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MP[&#39;L1&#39;] = dataL1_MP[&#39;L1&#39;].astype(&#39;category&#39;)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># do the same for the L1rel_MT data set</span>

<span class="c1"># Rename Nan in the L1 column to &#39;Other&#39;</span>

<span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="s1">&#39;Other&#39;</span><span class="p">)</span>

<span class="c1"># Change relative Date columns to days</span>
<span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_datetime</span><span class="p">(</span><span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">])</span>
<span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span><span class="o">.</span><span class="n">dt</span><span class="o">.</span><span class="n">days</span>

<span class="c1"># Change the &#39;L1&#39; column to a categorical column</span>
<span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataL1_MT</span><span class="p">[</span><span class="s1">&#39;L1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>

<span class="c1"># Plot the data, relative abundance (value) vs. days (date) for each L1</span>


<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;L1rel_MP&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">dataL1_MP</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Metaproteomic Data&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Days&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Abundance&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\User\AppData\Local\Temp\ipykernel_32172\4211173979.py:5: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MT[&#39;L1&#39;] = dataL1_MT[&#39;L1&#39;].replace(np.nan, &#39;Other&#39;)
C:\Users\User\AppData\Local\Temp\ipykernel_32172\4211173979.py:8: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MT[&#39;date&#39;] = pd.to_datetime(dataL1_MT[&#39;date&#39;])
C:\Users\User\AppData\Local\Temp\ipykernel_32172\4211173979.py:9: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MT[&#39;date&#39;] = (dataL1_MT[&#39;date&#39;] - dataL1_MT[&#39;date&#39;].min()).dt.days
C:\Users\User\AppData\Local\Temp\ipykernel_32172\4211173979.py:12: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataL1_MT[&#39;L1&#39;] = dataL1_MT[&#39;L1&#39;].astype(&#39;category&#39;)
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
c:\ProgramData\anaconda3\envs\MIMIC\Lib\site-packages\seaborn\_oldcore.py:1057: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  grouped_data = data.groupby(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_36_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_36_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># And now pivot the data so each &#39;L1&#39; is a column for each data set</span>

<span class="n">dataL1_MP_pivot</span> <span class="o">=</span> <span class="n">dataL1_MP</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;L1rel_MP&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="n">dataL1_MT_pivot</span> <span class="o">=</span> <span class="n">dataL1_MT</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="s1">&#39;date&#39;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;L1&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;L1rel_MT&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Step 1: Determine the minimum and maximum index</span>
<span class="n">min_index</span> <span class="o">=</span> <span class="n">dataL1_MP_pivot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<span class="n">max_index</span> <span class="o">=</span> <span class="n">dataL1_MP_pivot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>

<span class="c1"># Step 2: Create a new index range from min_index to max_index</span>
<span class="n">new_index</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">min_index</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">max_index</span><span class="p">)</span> <span class="o">+</span>
                  <span class="mi">1</span><span class="p">)</span>  <span class="c1"># Ensure it&#39;s an integer range</span>

<span class="c1"># Step 3: Re-index the DataFrame</span>
<span class="n">dataL1_MP_complete</span> <span class="o">=</span> <span class="n">dataL1_MP_pivot</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">new_index</span><span class="p">)</span>

<span class="n">dataL1_MP_complete</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>L1</th>
      <th>01_Fermentation</th>
      <th>02_Homoacetogenesis</th>
      <th>03_Superpathway of thiosulfate metabolism (Desulfovibrio sulfodismutans)</th>
      <th>04_Utililization of sugar, conversion of pentose to EMP pathway intermediates</th>
      <th>05_Fatty acid oxidation</th>
      <th>06_Amino acid utilization biosynthesis metabolism</th>
      <th>07_Nucleic acid metabolism</th>
      <th>08_Hydrocarbon degradation</th>
      <th>09_Carbohydrate Active enzyme - CAZy</th>
      <th>10_TCA cycle</th>
      <th>...</th>
      <th>12_Transporters</th>
      <th>13_Hydrogen metabolism</th>
      <th>14_Methanogenesis</th>
      <th>15_Methylotrophy</th>
      <th>16_Embden Meyerhof-Parnos (EMP)</th>
      <th>17_Gluconeogenesis</th>
      <th>18_Sulfur compounds metabolism</th>
      <th>19_Saccharide and derivated synthesis</th>
      <th>20_Hydrolysis of polymers</th>
      <th>21_Cellular response to stress</th>
    </tr>
    <tr>
      <th>date</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.096899</td>
      <td>0.009890</td>
      <td>0.000579</td>
      <td>0.000535</td>
      <td>0.025483</td>
      <td>0.122784</td>
      <td>0.119531</td>
      <td>0.005524</td>
      <td>0.003921</td>
      <td>0.045398</td>
      <td>...</td>
      <td>0.015593</td>
      <td>0.000846</td>
      <td>0.003520</td>
      <td>0.009267</td>
      <td>0.030473</td>
      <td>0.016261</td>
      <td>0.005346</td>
      <td>0.009267</td>
      <td>0.006193</td>
      <td>0.037334</td>
    </tr>
    <tr>
      <th>1</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>5</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>6</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>7</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>8</th>
      <td>0.133734</td>
      <td>0.011645</td>
      <td>0.000340</td>
      <td>0.000870</td>
      <td>0.033575</td>
      <td>0.129235</td>
      <td>0.108779</td>
      <td>0.004688</td>
      <td>0.003063</td>
      <td>0.052367</td>
      <td>...</td>
      <td>0.018224</td>
      <td>0.001059</td>
      <td>0.004159</td>
      <td>0.016334</td>
      <td>0.035428</td>
      <td>0.018678</td>
      <td>0.003592</td>
      <td>0.009150</td>
      <td>0.006692</td>
      <td>0.042688</td>
    </tr>
    <tr>
      <th>9</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>10</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>11</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>12</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>13</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>14</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>15 rows × 21 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reset the index</span>
<span class="n">complete_dataMP_reset</span> <span class="o">=</span> <span class="n">dataL1_MP_complete</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

<span class="c1"># Make index as float</span>
<span class="n">complete_dataMP_reset</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">complete_dataMP_reset</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>We are going to use the GP imputer to fill in the missing values in the dataset. We will then use the VAR inference method to infer the causal relationships between the variables in the time series. Note how we are only using the matern (M32, M12, and M52) and radial basis function (RBF) kernels in the GP imputer. This is because these kernels are the most suitable for time series data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># do the imputation for each column (metabolite) in the dataset</span>
<span class="c1"># Copy the original dataset</span>
<span class="n">new_dataset_complete</span> <span class="o">=</span> <span class="n">complete_dataMP_reset</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="k">for</span> <span class="n">metabolite</span> <span class="ow">in</span> <span class="n">dataL1_MP_complete</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="c1"># Impute the missing values for the current column</span>
    <span class="n">imputed_data</span> <span class="o">=</span> <span class="n">GPImputer</span><span class="p">()</span><span class="o">.</span><span class="n">impute_missing_values</span><span class="p">(</span>
        <span class="n">complete_dataMP_reset</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">metabolite</span><span class="p">],</span> <span class="n">metabolite</span><span class="p">,</span> <span class="n">kernel</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;M32&#39;</span><span class="p">,</span> <span class="s1">&#39;M12&#39;</span><span class="p">,</span> <span class="s1">&#39;M52&#39;</span><span class="p">,</span> <span class="s1">&#39;RBF&#39;</span><span class="p">]</span>
    <span class="p">)</span>

    <span class="c1"># Update the column in the new dataset with the imputed data</span>
    <span class="n">new_dataset_complete</span><span class="p">[</span><span class="n">metabolite</span><span class="p">]</span> <span class="o">=</span> <span class="n">imputed_data</span><span class="p">[</span><span class="n">metabolite</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern52&#39;&gt;
BIC: 305.14796205640255
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_1.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern52&#39;&gt;
BIC: 539.9907420516718
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_3.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_3.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern12&#39;&gt;
BIC: 593.6954527249973
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_5.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_5.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern52&#39;&gt;
BIC: 606.429158574828
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_7.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_7.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern12&#39;&gt;
BIC: 389.2225921347663
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_9.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_9.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern52&#39;&gt;
BIC: 321.45473434185317
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_11.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_11.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern32&#39;&gt;
BIC: 292.9620322625907
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_13.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_13.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern12&#39;&gt;
BIC: 558.9765129518931
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_15.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_15.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern32&#39;&gt;
BIC: 581.8255133135826
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_17.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_17.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern32&#39;&gt;
BIC: 384.214449988753
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_19.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_19.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern32&#39;&gt;
BIC: 448.7262491062683
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_21.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_21.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern32&#39;&gt;
BIC: 475.6907363223153
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_23.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_23.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern32&#39;&gt;
BIC: 595.2739514099666
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_25.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_25.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern12&#39;&gt;
BIC: 523.8038498902097
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_27.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_27.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern52&#39;&gt;
BIC: 492.72580947714255
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_29.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_29.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern32&#39;&gt;
BIC: 404.635080730377
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_31.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_31.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern32&#39;&gt;
BIC: 488.89321087777034
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_33.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_33.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern12&#39;&gt;
BIC: 443.5037096251963
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_35.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_35.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern32&#39;&gt;
BIC: 466.12061932819296
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_37.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_37.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern12&#39;&gt;
BIC: 520.4635688670087
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_39.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_39.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Model: GPR
Best kernel: &lt;class &#39;gpflow.kernels.stationaries.Matern32&#39;&gt;
BIC: 405.9096561312746
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_41.png" src="../../../_images/notebooks_MultiModel_Herold_examples_impute_data_41_41.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Save the dataset (new_dataset_complete) to a csv file</span>
<span class="n">new_dataset_complete</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;MP_metabolite_imputed_data.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Let&#39;s do the same for the MT dataset</span>

<span class="c1"># Step 1: Determine the minimum and maximum index</span>
<span class="n">min_index</span> <span class="o">=</span> <span class="n">dataL1_MT_pivot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<span class="n">max_index</span> <span class="o">=</span> <span class="n">dataL1_MT_pivot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>

<span class="c1"># Step 2: Create a new index range from min_index to max_index</span>
<span class="n">new_index</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">min_index</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">max_index</span><span class="p">)</span> <span class="o">+</span>
                  <span class="mi">1</span><span class="p">)</span>  <span class="c1"># Ensure it&#39;s an integer range</span>

<span class="c1"># Step 3: Re-index the DataFrame</span>
<span class="n">dataL1_MT_complete</span> <span class="o">=</span> <span class="n">dataL1_MT_pivot</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">new_index</span><span class="p">)</span>

<span class="n">dataL1_MT_complete</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Reset the index</span>
<span class="n">complete_dataMT_reset</span> <span class="o">=</span> <span class="n">dataL1_MT_complete</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

<span class="c1"># Make index as float</span>
<span class="n">complete_dataMT_reset</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">complete_dataMT_reset</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># do the imputation for each column (metabolite) in the dataset</span>
<span class="c1"># Copy the original dataset</span>
<span class="n">new_dataset_complete</span> <span class="o">=</span> <span class="n">complete_dataMT_reset</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="k">for</span> <span class="n">metabolite</span> <span class="ow">in</span> <span class="n">dataL1_MT_complete</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="c1"># Impute the missing values for the current column</span>
    <span class="n">imputed_data</span> <span class="o">=</span> <span class="n">GPImputer</span><span class="p">()</span><span class="o">.</span><span class="n">impute_missing_values</span><span class="p">(</span>
        <span class="n">complete_dataMT_reset</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">metabolite</span><span class="p">],</span> <span class="n">metabolite</span><span class="p">,</span> <span class="n">kernel</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;M32&#39;</span><span class="p">,</span> <span class="s1">&#39;M12&#39;</span><span class="p">,</span> <span class="s1">&#39;M52&#39;</span><span class="p">,</span> <span class="s1">&#39;RBF&#39;</span><span class="p">]</span>
    <span class="p">)</span>

    <span class="c1"># Update the column in the new dataset with the imputed data</span>
    <span class="n">new_dataset_complete</span><span class="p">[</span><span class="n">metabolite</span><span class="p">]</span> <span class="o">=</span> <span class="n">imputed_data</span><span class="p">[</span><span class="n">metabolite</span><span class="p">]</span>

<span class="c1"># Save the dataset (new_dataset_complete) to a csv file</span>
<span class="n">new_dataset_complete</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;MT_metabolite_imputed_data.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">MIMIC</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../readme.html">Modelling and Inference of MICrobiomes Project (MIMIC)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">mimic</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../../examples.html">Examples</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../../../examples.html#jupyter-notebook-examples-by-model">Jupyter Notebook Examples by Model</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../../../examples.html#consumer-resource-model-crm">Consumer Resource Model (CRM)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples.html#gaussian-processes-gp">Gaussian Processes (GP)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples.html#multivariate-autoregressive-model-mvar">Multivariate Autoregressive Model (MVAR)</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="../../../examples.html#multi-model-analysis">Multi-Model Analysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples.html#vector-autoregression-var">Vector Autoregression (VAR)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples.html#generalized-lotka-volterra-glv">Generalized Lotka-Volterra (gLV)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples.html#generalized-metabolic-lotka-volterra-gmlv">Generalized Metabolic Lotka-Volterra (gMLV)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../contributing.html">Contributing to MIMIC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../history.html">History</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../../examples.html">Examples</a><ul>
      <li>Previous: <a href="examples-Herold-sVAR.html" title="previous chapter">Sparse VAR on a Real Public Dataset, Herold et al. (2020)</a></li>
      <li>Next: <a href="../../VAR/examples-bayes-VAR.html" title="next chapter">Bayesian Inference for Vector AutoRegression (VAR) Models</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2024, Pedro Fontanarrosa.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="../../../_sources/notebooks/MultiModel/Herold/examples_impute_data.ipynb.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>